// @atlaskit/media-picker/dist/es2015/popup/components/views/editor/errorView/errorView.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Component } from 'react';
// import Button from '@atlaskit/button';
import { EscHelper } from '../escHelper';
import { errorHintRetry, errorHintCritical, errorButtonRetry, errorButtonCancel, errorButtonClose, } from '../phrases';
import { CenterView } from '../styles';
import { ErrorPopup, ErrorIconWrapper, ErrorMessage, ErrorHint, ErrorButton, } from './styles';
import { errorIcon } from '../../../../../icons';
var ErrorView = /** @class */ (function (_super) {
    tslib_1.__extends(ErrorView, _super);
    function ErrorView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ErrorView.prototype.componentDidMount = function () {
        this.escHelper = new EscHelper(this.props.onCancel);
    };
    ErrorView.prototype.componentWillUnmount = function () {
        this.escHelper.teardown();
    };
    ErrorView.prototype.render = function () {
        return (React.createElement(CenterView, null,
            React.createElement(ErrorPopup, null,
                React.createElement(ErrorIconWrapper, null, errorIcon),
                React.createElement(ErrorMessage, null, this.props.message),
                React.createElement(ErrorHint, null, this.renderHint()),
                this.renderTryAgainButton(),
                this.renderCancelButton())));
    };
    ErrorView.prototype.renderHint = function () {
        var onRetry = this.props.onRetry;
        if (onRetry) {
            return React.createElement("span", null, errorHintRetry);
        }
        return React.createElement("span", null, errorHintCritical);
    };
    ErrorView.prototype.renderTryAgainButton = function () {
        var onRetry = this.props.onRetry;
        if (onRetry) {
            return (React.createElement(ErrorButton, { appearance: "primary", onClick: onRetry }, errorButtonRetry));
        }
        return null;
    };
    ErrorView.prototype.renderCancelButton = function () {
        var _a = this.props, onCancel = _a.onCancel, onRetry = _a.onRetry;
        return (React.createElement(ErrorButton, { appearance: "subtle", onClick: onCancel }, onRetry ? errorButtonCancel : errorButtonClose));
    };
    return ErrorView;
}(Component));
export { ErrorView };
