// @atlaskit/media-card/dist/es2015/root/withDataURI.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { isRetina } from '../utils/isRetina';
import { isLinkDetails } from '../utils/isLinkDetails';
import { isValidPercentageUnit } from '../utils/isValidPercentageUnit';
import { containsPixelUnit } from '../utils/containsPixelUnit';
import { getCardMinHeight } from '../utils/cardDimensions';
import { getElementDimension, } from '../utils/getElementDimension';
import { defaultImageCardDimensions } from '../utils';
// return type is "any" to avoid TS attempting to infer the return type
// if TS attempts to infer the return type it can NOT publish .d.ts files because WithDataURIImpl isn't exported
export function withDataURI(Component) {
    var WithDataURIImpl = /** @class */ (function (_super) {
        tslib_1.__extends(WithDataURIImpl, _super);
        function WithDataURIImpl() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.state = {};
            _this.setDataURI = function (dataURI) { return _this.setState({ dataURI: dataURI }); };
            _this.clearDataURI = function () { return _this.setState({ dataURI: undefined }); };
            return _this;
        }
        WithDataURIImpl.prototype.componentDidMount = function () {
            this.updateDataURI(this.props);
        };
        WithDataURIImpl.prototype.componentWillReceiveProps = function (nextProps) {
            var _a = this.props, currentDataURIService = _a.dataURIService, currentMetadata = _a.metadata;
            var nextDataURIService = nextProps.dataURIService, nextMetadata = nextProps.metadata;
            if (nextDataURIService !== currentDataURIService ||
                nextMetadata !== currentMetadata) {
                this.updateDataURI(nextProps);
            }
        };
        WithDataURIImpl.prototype.isSmall = function () {
            return this.props.appearance === 'small';
        };
        // No mather if the integrator passed pixels or percentages, this will
        // always return a pixels value that the /image endpoint can use
        WithDataURIImpl.prototype.dataURIDimension = function (dimension) {
            var retinaFactor = isRetina() ? 2 : 1;
            var dimensionValue = (this.props.dimensions && this.props.dimensions[dimension]) || '';
            if (this.isSmall()) {
                return getCardMinHeight('small') * retinaFactor;
            }
            if (isValidPercentageUnit(dimensionValue)) {
                return getElementDimension(this, dimension) * retinaFactor;
            }
            if (typeof dimensionValue === 'number') {
                return dimensionValue * retinaFactor;
            }
            if (containsPixelUnit("" + dimensionValue)) {
                return parseInt("" + dimensionValue, 10) * retinaFactor;
            }
            return defaultImageCardDimensions[dimension] * retinaFactor;
        };
        WithDataURIImpl.prototype.updateDataURI = function (props) {
            var dataURIService = props.dataURIService, metadata = props.metadata, resizeMode = props.resizeMode, appearance = props.appearance;
            var _a = this, setDataURI = _a.setDataURI, clearDataURI = _a.clearDataURI;
            // clear the dataURI if we're updating to undefined metadata or we're updating to a link
            if (!dataURIService || !metadata || isLinkDetails(metadata)) {
                this.clearDataURI();
                return;
            }
            var width = this.dataURIDimension('width');
            var height = this.dataURIDimension('height');
            var allowAnimated = appearance !== 'small';
            dataURIService
                .fetchImageDataUri({ type: 'file', details: metadata }, {
                width: width,
                height: height,
                mode: resizeMode,
                allowAnimated: allowAnimated,
            })
                .then(setDataURI, clearDataURI);
        };
        WithDataURIImpl.prototype.render = function () {
            var dataURI = this.state.dataURI || this.props.preview;
            var props = tslib_1.__assign({}, this.props);
            delete props.dataURIService;
            var otherProps = tslib_1.__assign({ dataURI: dataURI }, props);
            return React.createElement(Component, tslib_1.__assign({}, otherProps));
        };
        return WithDataURIImpl;
    }(React.Component));
    return WithDataURIImpl;
}
