// @atlaskit/field-base/dist/cjs/components/ValidationElement.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WarningIcon = undefined;

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  align-items: center;\n  color: ', ';\n  display: flex;\n  flex-shrink: 0;\n'], ['\n  align-items: center;\n  color: ', ';\n  display: flex;\n  flex-shrink: 0;\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  height: 20px;\n  margin-left: 10px;\n'], ['\n  height: 20px;\n  margin-left: 10px;\n']);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _theme = require('@atlaskit/theme');

var _spinner = require('@atlaskit/spinner');

var _spinner2 = _interopRequireDefault(_spinner);

var _warning = require('@atlaskit/icon/glyph/warning');

var _warning2 = _interopRequireDefault(_warning);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// exported for testing
var WarningIcon = exports.WarningIcon = _styledComponents2.default.div(_templateObject, _theme.colors.yellow);

// Spinner needs set height to avoid height jumping
// Also needs a margin so there is space between it and preceding text
var SpinnerParent = _styledComponents2.default.div(_templateObject2);

var ValidationElement = function (_Component) {
  (0, _inherits3.default)(ValidationElement, _Component);

  function ValidationElement() {
    (0, _classCallCheck3.default)(this, ValidationElement);
    return (0, _possibleConstructorReturn3.default)(this, (ValidationElement.__proto__ || (0, _getPrototypeOf2.default)(ValidationElement)).apply(this, arguments));
  }

  (0, _createClass3.default)(ValidationElement, [{
    key: 'render',
    value: function render() {
      if (!this.props.isDisabled && this.props.isInvalid) {
        return _react2.default.createElement(
          WarningIcon,
          null,
          _react2.default.createElement(_warning2.default, { label: 'warning' })
        );
      }

      return this.props.isLoading ? _react2.default.createElement(
        SpinnerParent,
        null,
        _react2.default.createElement(_spinner2.default, { size: 'small' })
      ) : null;
    }
  }]);
  return ValidationElement;
}(_react.Component);

ValidationElement.defaultProps = {
  isDisabled: false,
  isInvalid: false,
  isLoading: false
};
exports.default = ValidationElement;
