// @atlaskit/renderer/dist/es5/react/marks/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var code_1 = require("./code");
exports.Code = code_1.default;
var em_1 = require("./em");
exports.Em = em_1.default;
var link_1 = require("./link");
exports.Link = link_1.default;
var strike_1 = require("./strike");
exports.Strike = strike_1.default;
var strong_1 = require("./strong");
exports.Strong = strong_1.default;
var subsup_1 = require("./subsup");
exports.Subsup = subsup_1.default;
var textColor_1 = require("./textColor");
exports.TextColor = textColor_1.default;
var underline_1 = require("./underline");
exports.Underline = underline_1.default;
var action_1 = require("./action");
exports.Action = action_1.default;
// Stage0
var confluence_inline_comment_1 = require("./confluence-inline-comment");
exports.markToReact = {
    code: code_1.default,
    em: em_1.default,
    link: link_1.default,
    strike: strike_1.default,
    strong: strong_1.default,
    subsup: subsup_1.default,
    textColor: textColor_1.default,
    underline: underline_1.default,
    action: action_1.default,
    // Stage0
    confluenceInlineComment: confluence_inline_comment_1.default,
};
exports.toReact = function (mark) {
    return exports.markToReact[mark.type.name];
};
