// @atlaskit/task-decision/dist/es5/components/Participants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var react_1 = require("react");
var avatar_1 = require("@atlaskit/avatar");
var theme_1 = require("@atlaskit/theme");
var Partipants = /** @class */ (function (_super) {
    tslib_1.__extends(Partipants, _super);
    function Partipants() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Partipants.prototype.getAvatarData = function () {
        return this.props.participants.map(function (p) { return ({
            src: p.avatarUrl,
            name: p.displayName,
            size: 'small',
        }); });
    };
    Partipants.prototype.render = function () {
        return (React.createElement(avatar_1.AvatarGroup, { appearance: "stack", borderColor: theme_1.colors.N20, maxCount: 4, size: "small", data: this.getAvatarData(), boundariesElement: "scrollParent" }));
    };
    return Partipants;
}(react_1.PureComponent));
exports.default = Partipants;
