// @atlaskit/task-decision/dist/es5/components/ResourcedItemList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var react_1 = require("react");
var styled_components_1 = require("styled-components");
var spinner_1 = require("@atlaskit/spinner");
var constants_1 = require("../constants");
var TaskDecisionLoader_1 = require("../api/TaskDecisionLoader");
var InfiniteScroll_1 = require("./InfiniteScroll");
var ListContainer_1 = require("../styled/ListContainer");
var ListWrapper_1 = require("../styled/ListWrapper");
var DateGroup_1 = require("../styled/DateGroup");
var DateGroupHeader_1 = require("../styled/DateGroupHeader");
var type_helpers_1 = require("../type-helpers");
var date_1 = require("../util/date");
var type_helpers_2 = require("../type-helpers");
var DecisionItem_1 = require("./DecisionItem");
var ResourcedTaskItem_1 = require("./ResourcedTaskItem");
// tslint:disable-next-line:variable-name
var LoadingWrapper = styled_components_1.default.div(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  margin: 20px 0;\n  height: ", ";\n"], ["\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  margin: 20px 0;\n  height: ", ";\n"])), function (props) { return props.height || 'auto'; });
var ResourcedItemList = /** @class */ (function (_super) {
    tslib_1.__extends(ResourcedItemList, _super);
    function ResourcedItemList(props) {
        var _this = _super.call(this, props) || this;
        _this.loadLatest = function (recentUpdateContext) {
            var _a = _this.props, initialQuery = _a.initialQuery, taskDecisionProvider = _a.taskDecisionProvider;
            var items = _this.state.items;
            taskDecisionProvider.then(function (provider) {
                TaskDecisionLoader_1.loadLatestItems(initialQuery, items || [], provider, recentUpdateContext).then(function (latestItems) {
                    if (_this.mounted) {
                        _this.setState({
                            items: latestItems,
                        });
                    }
                });
            });
        };
        _this.loadMore = function () {
            var nextQuery = _this.state.nextQuery;
            if (nextQuery) {
                _this.performQuery(nextQuery, false);
            }
        };
        _this.state = {
            loading: true,
            error: false,
        };
        return _this;
    }
    ResourcedItemList.prototype.componentDidMount = function () {
        this.mounted = true;
        this.performInitialQuery(this.props);
    };
    ResourcedItemList.prototype.componentWillUnmount = function () {
        this.mounted = false;
        this.unsubscribeRecentUpdates();
    };
    ResourcedItemList.prototype.componentWillReceiveProps = function (nextProps) {
        if (this.props.initialQuery !== nextProps.initialQuery ||
            this.props.taskDecisionProvider !== nextProps.taskDecisionProvider) {
            this.unsubscribeRecentUpdates();
            this.performInitialQuery(nextProps);
        }
    };
    ResourcedItemList.prototype.unsubscribeRecentUpdates = function () {
        var recentUpdatesId = this.recentUpdatesId;
        if (recentUpdatesId) {
            this.props.taskDecisionProvider.then(function (provider) {
                provider.unsubscribeRecentUpdates(recentUpdatesId);
            });
        }
        this.recentUpdatesId = undefined;
    };
    ResourcedItemList.prototype.performInitialQuery = function (props) {
        var _this = this;
        var initialQuery = props.initialQuery;
        this.performQuery(initialQuery, true, {
            id: function (id) {
                _this.recentUpdatesId = id;
            },
            recentUpdates: this.loadLatest,
        });
    };
    ResourcedItemList.prototype.performQuery = function (query, replaceAll, recentUpdatesListener) {
        var _this = this;
        var taskDecisionProvider = this.props.taskDecisionProvider;
        var items = replaceAll ? [] : this.state.items;
        this.setState({
            loading: true,
            error: false,
            items: items,
        });
        taskDecisionProvider.then(function (provider) {
            provider
                .getItems(query, recentUpdatesListener)
                .then(function (result) {
                if (!_this.mounted) {
                    return;
                }
                var items = result.items, nextQuery = result.nextQuery;
                var combinedItems;
                if (replaceAll) {
                    combinedItems = items;
                }
                else {
                    combinedItems = (_this.state.items || []).concat(items);
                }
                _this.setState({
                    items: combinedItems,
                    nextQuery: nextQuery,
                    loading: false,
                });
                var onUpdate = _this.props.onUpdate;
                if (onUpdate) {
                    onUpdate(combinedItems, items);
                }
            })
                .catch(function (err) {
                if (!_this.mounted) {
                    return;
                }
                _this.setState({
                    loading: false,
                    error: true,
                });
            });
        });
    };
    ResourcedItemList.prototype.renderItems = function () {
        var _a = this.props, appearance = _a.appearance, groupItems = _a.groupItems, initialQuery = _a.initialQuery;
        var items = this.state.items;
        if (!items) {
            return null;
        }
        var sortCriteria = initialQuery.sortCriteria;
        var renderedItems;
        if (groupItems && type_helpers_1.isDateSortCriteria(sortCriteria)) {
            renderedItems = this.renderItemsGroupedByDate(items);
        }
        else {
            renderedItems = this.renderItemsUngrouped(items);
        }
        return (React.createElement(ListContainer_1.default, { theme: { appearance: appearance } }, renderedItems));
    };
    ResourcedItemList.prototype.renderItemsUngrouped = function (items) {
        var _a = this.props, appearance = _a.appearance, renderDocument = _a.renderDocument, taskDecisionProvider = _a.taskDecisionProvider;
        return (React.createElement(ListWrapper_1.default, null, items.map(function (item) {
            var objectKey = type_helpers_2.toObjectKey(item);
            if (type_helpers_2.isDecision(item)) {
                return (React.createElement("li", { key: type_helpers_2.objectKeyToString(objectKey) },
                    React.createElement(DecisionItem_1.default, { appearance: appearance, participants: item.participants, creator: item.creator, lastUpdater: item.lastUpdater }, renderDocument(item.content, type_helpers_1.toRendererContext(objectKey)))));
            }
            if (type_helpers_2.isTask(item)) {
                return (React.createElement("li", { key: type_helpers_2.objectKeyToString(objectKey) },
                    React.createElement(ResourcedTaskItem_1.default, { key: type_helpers_2.objectKeyToString(objectKey), taskDecisionProvider: taskDecisionProvider, taskId: objectKey.localId, objectAri: objectKey.objectAri, containerAri: objectKey.containerAri, appearance: appearance, participants: item.participants, creator: item.creator, lastUpdater: item.lastUpdater }, renderDocument(item.content, type_helpers_1.toRendererContext(objectKey)))));
            }
            return null;
        })));
    };
    ResourcedItemList.prototype.renderItemsGroupedByDate = function (items) {
        var _this = this;
        var itemsByDate = this.groupItemsByDate(items);
        return (React.createElement(DateGroup_1.default, null, itemsByDate.map(function (_a) {
            var date = _a.date, items = _a.items;
            return (React.createElement("li", { key: date.toISOString() },
                React.createElement(DateGroupHeader_1.default, null, date_1.getFormattedDate(date)),
                _this.renderItemsUngrouped(items)));
        })));
    };
    ResourcedItemList.prototype.groupItemsByDate = function (items) {
        var groupByField = this.props.initialQuery.sortCriteria || constants_1.defaultSortCriteria;
        var lastDate;
        return items.reduce(function (groups, item) {
            var currentDate = date_1.getStartOfDate(item[groupByField]);
            if (date_1.isSameDate(lastDate, currentDate)) {
                var lastGroup = groups[groups.length - 1];
                lastGroup.items.push(item);
            }
            else {
                lastDate = currentDate;
                groups.push({
                    date: currentDate,
                    items: [item],
                });
            }
            return groups;
        }, []);
    };
    ResourcedItemList.prototype.render = function () {
        var _a = this.props, emptyComponent = _a.emptyComponent, errorComponent = _a.errorComponent, height = _a.height, useInfiniteScroll = _a.useInfiniteScroll;
        var _b = this.state, error = _b.error, items = _b.items, loading = _b.loading;
        var loadingSpinner;
        if (error && errorComponent) {
            return errorComponent || null;
        }
        if (loading) {
            var height_1;
            if (!items || items.length === 0) {
                height_1 = '100%';
            }
            loadingSpinner = (React.createElement(LoadingWrapper, { height: height_1 },
                React.createElement(spinner_1.default, { size: "medium" })));
        }
        else if (!items || !items.length) {
            return emptyComponent || null;
        }
        if (height && useInfiniteScroll) {
            return (React.createElement(InfiniteScroll_1.default, { height: height, onThresholdReached: this.loadMore },
                this.renderItems(),
                loadingSpinner));
        }
        return (React.createElement("div", null,
            this.renderItems(),
            loadingSpinner));
    };
    ResourcedItemList.defaultProps = {
        appearance: 'card',
    };
    return ResourcedItemList;
}(react_1.PureComponent));
exports.default = ResourcedItemList;
var templateObject_1;
