// @atlaskit/code/dist/cjs/themes/defaultTheme.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultColors = undefined;

var _theme = require('@atlaskit/theme');

var defaultColors = exports.defaultColors = {
  lineNumberColor: _theme.colors.N90,
  lineNumberBgColor: _theme.colors.N30,
  backgroundColor: _theme.colors.N20,
  textColor: _theme.colors.N800,
  substringColor: _theme.colors.N400,
  keywordColor: _theme.colors.B500,
  attributeColor: _theme.colors.T500,
  selectorTagColor: _theme.colors.B500,
  docTagColor: _theme.colors.Y300,
  nameColor: _theme.colors.B500,
  builtInColor: _theme.colors.B500,
  literalColor: _theme.colors.B500,
  bulletColor: _theme.colors.B500,
  codeColor: _theme.colors.B500,
  additionColor: _theme.colors.B500,
  regexpColor: _theme.colors.T300,
  symbolColor: _theme.colors.T300,
  variableColor: _theme.colors.T300,
  templateVariableColor: _theme.colors.T300,
  linkColor: _theme.colors.P300,
  selectorAttributeColor: _theme.colors.T300,
  selectorPseudoColor: _theme.colors.T300,
  typeColor: _theme.colors.T500,
  stringColor: _theme.colors.G400,
  selectorIdColor: _theme.colors.T500,
  selectorClassColor: _theme.colors.T500,
  quoteColor: _theme.colors.T500,
  templateTagColor: _theme.colors.T500,
  deletionColor: _theme.colors.T500,
  titleColor: _theme.colors.P300,
  sectionColor: _theme.colors.P300,
  commentColor: _theme.colors.N400,
  metaKeywordColor: _theme.colors.G400,
  metaColor: _theme.colors.N400,
  functionColor: _theme.colors.N800,
  numberColor: _theme.colors.B500
};
