// @atlaskit/editor-core/dist/es5/plugins/type-ahead/commands/select-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var prosemirror_state_1 = require("prosemirror-state");
var prosemirror_model_1 = require("prosemirror-model");
var prosemirror_utils_1 = require("prosemirror-utils");
var utils_1 = require("../../../utils");
var main_1 = require("../pm-plugins/main");
var find_query_mark_1 = require("../utils/find-query-mark");
var dismiss_1 = require("./dismiss");
exports.selectCurrentItem = function () { return function (state, dispatch) {
    var _a = main_1.pluginKey.getState(state), active = _a.active, currentIndex = _a.currentIndex, items = _a.items, typeAheadHandler = _a.typeAheadHandler;
    if (!active || !typeAheadHandler) {
        return false;
    }
    if (!typeAheadHandler.selectItem || !items[currentIndex]) {
        return exports.withTypeAheadQueryMarkPosition(state, function (start, end) {
            return exports.insertFallbackCommand(start, end)(state, dispatch);
        });
    }
    return exports.selectItem(typeAheadHandler, items[currentIndex])(state, dispatch);
}; };
exports.selectSingleItemOrDismiss = function () { return function (state, dispatch) {
    var _a = main_1.pluginKey.getState(state), active = _a.active, items = _a.items, typeAheadHandler = _a.typeAheadHandler;
    if (!active || !typeAheadHandler || !typeAheadHandler.selectItem) {
        return false;
    }
    if (items.length === 1) {
        return exports.selectItem(typeAheadHandler, items[0])(state, dispatch);
    }
    if (!items || items.length === 0) {
        return dismiss_1.dismissCommand()(state, dispatch);
    }
    return false;
}; };
exports.selectByIndex = function (index) { return function (state, dispatch) {
    var _a = main_1.pluginKey.getState(state), active = _a.active, items = _a.items, typeAheadHandler = _a.typeAheadHandler;
    if (!active ||
        !typeAheadHandler ||
        !typeAheadHandler.selectItem ||
        !items[index]) {
        return false;
    }
    return exports.selectItem(typeAheadHandler, items[index])(state, dispatch);
}; };
exports.selectItem = function (handler, item) { return function (state, dispatch) {
    return exports.withTypeAheadQueryMarkPosition(state, function (start, end) {
        var replaceWith = function (node) {
            var tr = state.tr;
            tr = tr
                .setMeta(main_1.pluginKey, { action: main_1.ACTIONS.SELECT_CURRENT })
                .replaceWith(start, end, prosemirror_model_1.Fragment.empty);
            /**
             *
             * Replacing a type ahead query mark with a block node.
             *
             */
            if (node.isBlock) {
                tr = prosemirror_utils_1.safeInsert(node)(tr);
                /**
                 *
                 * Replacing a type ahead query mark with an inline node.
                 *
                 */
            }
            else if (node.isInline) {
                var fragment = prosemirror_model_1.Fragment.fromArray([node, state.schema.text(' ')]);
                tr = tr.replaceWith(start, start, fragment);
                // This problem affects Chrome v58-62. See: https://github.com/ProseMirror/prosemirror/issues/710
                if (utils_1.isChromeWithSelectionBug) {
                    document.getSelection().empty();
                }
                // Placing cursor after node + space.
                tr = tr.setSelection(prosemirror_state_1.Selection.near(tr.doc.resolve(start + fragment.size)));
            }
            dispatch(tr);
            return true;
        };
        if (handler.selectItem(state, item, replaceWith) === false) {
            return exports.insertFallbackCommand(start, end)(state, dispatch);
        }
        return true;
    });
}; };
exports.insertFallbackCommand = function (start, end) { return function (state, dispatch) {
    var _a = main_1.pluginKey.getState(state), query = _a.query, trigger = _a.trigger;
    var node = state.schema.text(trigger + query);
    dispatch(state.tr.replaceWith(start, end, node));
    return true;
}; };
exports.withTypeAheadQueryMarkPosition = function (state, cb) {
    var doc = state.doc;
    var typeAheadQuery = state.schema.marks.typeAheadQuery;
    var queryMark = find_query_mark_1.findQueryMark(typeAheadQuery, doc, 0, doc.nodeSize - 2);
    if (!queryMark) {
        return false;
    }
    return cb(queryMark.start, queryMark.end);
};
