// @atlaskit/media-picker/dist/es5/service/mediaClientPool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var mediaClient_1 = require("./mediaClient");
var MediaClientPool = /** @class */ (function () {
    function MediaClientPool(apiUrl, authProvider) {
        var _this = this;
        this.apiUrl = apiUrl;
        this.authProvider = authProvider;
        this.clients = {};
        this.getMediaClient = function (collection) {
            return _this.getStoredClient(collection) || _this.createNewClient(collection);
        };
        this.createNewClient = function (collection) {
            var client = new mediaClient_1.MediaClient(_this.apiUrl, _this.authProvider, collection);
            _this.setClient(client, collection);
            return client;
        };
        this.getStoredClient = function (collection) {
            if (collection === undefined) {
                return _this.clientForAbsentCollection;
            }
            return _this.clients[collection];
        };
        this.setClient = function (client, collection) {
            if (collection === undefined) {
                _this.clientForAbsentCollection = client;
                return;
            }
            _this.clients[collection] = client;
        };
    }
    return MediaClientPool;
}());
exports.MediaClientPool = MediaClientPool;
