// @atlaskit/media-picker/dist/es5/outer/analytics/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var UserEvent = /** @class */ (function () {
    function UserEvent(name, properties) {
        this.prefix = 'files.mediapicker-web.';
        this.name = this.prefix + name;
        this.properties = properties;
    }
    UserEvent.prototype.toJSON = function () {
        return {
            name: this.name,
            properties: this.properties,
        };
    };
    return UserEvent;
}());
exports.UserEvent = UserEvent;
var MediaPickerLoaded = /** @class */ (function (_super) {
    tslib_1.__extends(MediaPickerLoaded, _super);
    function MediaPickerLoaded() {
        return _super.call(this, 'library.loaded', {}) || this;
    }
    return MediaPickerLoaded;
}(UserEvent));
exports.MediaPickerLoaded = MediaPickerLoaded;
var MPBrowserLoaded = /** @class */ (function (_super) {
    tslib_1.__extends(MPBrowserLoaded, _super);
    function MPBrowserLoaded() {
        return _super.call(this, 'browser.loaded', {}) || this;
    }
    return MPBrowserLoaded;
}(UserEvent));
exports.MPBrowserLoaded = MPBrowserLoaded;
var MPDropzoneLoaded = /** @class */ (function (_super) {
    tslib_1.__extends(MPDropzoneLoaded, _super);
    function MPDropzoneLoaded() {
        return _super.call(this, 'dropzone.loaded', {}) || this;
    }
    return MPDropzoneLoaded;
}(UserEvent));
exports.MPDropzoneLoaded = MPDropzoneLoaded;
var MPClipboardLoaded = /** @class */ (function (_super) {
    tslib_1.__extends(MPClipboardLoaded, _super);
    function MPClipboardLoaded() {
        return _super.call(this, 'cliboard.loaded', {}) || this;
    }
    return MPClipboardLoaded;
}(UserEvent));
exports.MPClipboardLoaded = MPClipboardLoaded;
/* Popup */
var MPPopupLoaded = /** @class */ (function (_super) {
    tslib_1.__extends(MPPopupLoaded, _super);
    function MPPopupLoaded() {
        return _super.call(this, 'popup.loaded', {}) || this;
    }
    return MPPopupLoaded;
}(UserEvent));
exports.MPPopupLoaded = MPPopupLoaded;
var MPPopupShown = /** @class */ (function (_super) {
    tslib_1.__extends(MPPopupShown, _super);
    function MPPopupShown() {
        return _super.call(this, 'popup.shown', {}) || this;
    }
    return MPPopupShown;
}(UserEvent));
exports.MPPopupShown = MPPopupShown;
var MPPopupHidden = /** @class */ (function (_super) {
    tslib_1.__extends(MPPopupHidden, _super);
    function MPPopupHidden() {
        return _super.call(this, 'popup.hidden', {}) || this;
    }
    return MPPopupHidden;
}(UserEvent));
exports.MPPopupHidden = MPPopupHidden;
/* File operations */
var MPFileUploadEnded = /** @class */ (function (_super) {
    tslib_1.__extends(MPFileUploadEnded, _super);
    function MPFileUploadEnded() {
        return _super.call(this, 'file.upload.ended', {}) || this;
    }
    return MPFileUploadEnded;
}(UserEvent));
exports.MPFileUploadEnded = MPFileUploadEnded;
var MPFileUploadStarted = /** @class */ (function (_super) {
    tslib_1.__extends(MPFileUploadStarted, _super);
    function MPFileUploadStarted() {
        return _super.call(this, 'file.upload.started', {}) || this;
    }
    return MPFileUploadStarted;
}(UserEvent));
exports.MPFileUploadStarted = MPFileUploadStarted;
var MPFileProcessingStarted = /** @class */ (function (_super) {
    tslib_1.__extends(MPFileProcessingStarted, _super);
    function MPFileProcessingStarted() {
        return _super.call(this, 'file.processing.started', {}) || this;
    }
    return MPFileProcessingStarted;
}(UserEvent));
exports.MPFileProcessingStarted = MPFileProcessingStarted;
var ErrorException = /** @class */ (function (_super) {
    tslib_1.__extends(ErrorException, _super);
    function ErrorException(errorAlias, errorStack, errorDescription) {
        var _this = this;
        var errorMessage = {
            alias: errorAlias,
            description: errorDescription,
            stacktrace: errorStack,
        };
        _this = _super.call(this, 'files.mediapicker-web.library.exception.fired', errorMessage) || this;
        return _this;
    }
    return ErrorException;
}(UserEvent));
exports.ErrorException = ErrorException;
