// @atlaskit/media-picker/dist/es5/outer/analytics/tracker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var herment_1 = require("../../lib/herment");
var UserTracker = /** @class */ (function () {
    function UserTracker() {
        this.defaults = {
            storage_key: 'media-picker.herment.storage',
            product: 'media-picker',
            subproduct: 'library',
            version: '1.0',
        };
        this.queue = [];
        this.queueProcessor = herment_1.default(tslib_1.__assign({}, this.defaults, { queue: this.queue }));
        if (this.queueProcessor) {
            this.queueProcessor.start();
        }
        this.defaultProperties = {};
    }
    UserTracker.prototype.track = function () {
        var _this = this;
        return function (userEvent) {
            var event = userEvent.toJSON();
            event.properties = tslib_1.__assign({}, _this.defaultProperties, event.properties);
            _this.queue.push(event);
        };
    };
    return UserTracker;
}());
exports.UserTracker = UserTracker;
