// @atlaskit/editor-core/dist/es5/plugins/layout/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var editor_common_1 = require("@atlaskit/editor-common");
var prosemirror_state_1 = require("prosemirror-state");
exports.pluginKey = new prosemirror_state_1.PluginKey('layout');
var layoutPlugin = {
    nodes: function () {
        return [
            { rank: 2400, name: 'layoutSection', node: editor_common_1.layoutSection },
            { rank: 2400, name: 'layoutColumn', node: editor_common_1.layoutColumn },
        ];
    },
    pmPlugins: function () {
        return [
            {
                rank: 2400,
                plugin: function (_a) {
                    var schema = _a.schema;
                    return new prosemirror_state_1.Plugin({
                        key: exports.pluginKey,
                        state: {
                            init: function () { return ({}); },
                            apply: function (tr, state) { return state; },
                        },
                        props: {},
                    });
                },
            },
        ];
    },
};
exports.default = layoutPlugin;
