// @atlaskit/comment/dist/esm/components/Author.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { Component } from 'react';
import Field from './Field';

var Author = function (_Component) {
  _inherits(Author, _Component);

  function Author() {
    _classCallCheck(this, Author);

    return _possibleConstructorReturn(this, (Author.__proto__ || _Object$getPrototypeOf(Author)).apply(this, arguments));
  }

  _createClass(Author, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          href = _props.href,
          onClick = _props.onClick,
          onFocus = _props.onFocus,
          onMouseOver = _props.onMouseOver;

      return React.createElement(
        Field,
        {
          hasAuthor: true,
          href: href,
          onClick: onClick,
          onFocus: onFocus,
          onMouseOver: onMouseOver
        },
        children
      );
    }
  }]);

  return Author;
}(Component);

export default Author;
