// @atlaskit/comment/dist/esm/components/Field.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { Component } from 'react';
import { Anchor, Span } from '../styled/FieldStyles';

var CommentField = function (_Component) {
  _inherits(CommentField, _Component);

  function CommentField() {
    _classCallCheck(this, CommentField);

    return _possibleConstructorReturn(this, (CommentField.__proto__ || _Object$getPrototypeOf(CommentField)).apply(this, arguments));
  }

  _createClass(CommentField, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          hasAuthor = _props.hasAuthor,
          href = _props.href,
          onClick = _props.onClick,
          onFocus = _props.onFocus,
          onMouseOver = _props.onMouseOver;
      /* eslint-disable jsx-a11y/no-static-element-interactions */

      return href ? React.createElement(
        Anchor,
        {
          href: href,
          hasAuthor: hasAuthor,
          onClick: onClick,
          onFocus: onFocus,
          onMouseOver: onMouseOver
        },
        children
      ) : React.createElement(
        Span,
        {
          hasAuthor: hasAuthor,
          onClick: onClick,
          onFocus: onFocus,
          onMouseOver: onMouseOver
        },
        children
      );
      /* eslint-enable jsx-a11y/no-static-element-interactions */
    }
  }]);

  return CommentField;
}(Component);

export default CommentField;
