// @atlaskit/comment/dist/esm/components/Time.js
import _Object$getPrototypeOf from 'babel-runtime/core-js/object/get-prototype-of';
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _createClass from 'babel-runtime/helpers/createClass';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { Component } from 'react';
import Field from './Field';

var Time = function (_Component) {
  _inherits(Time, _Component);

  function Time() {
    _classCallCheck(this, Time);

    return _possibleConstructorReturn(this, (Time.__proto__ || _Object$getPrototypeOf(Time)).apply(this, arguments));
  }

  _createClass(Time, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          href = _props.href,
          onClick = _props.onClick,
          onFocus = _props.onFocus,
          onMouseOver = _props.onMouseOver;

      return React.createElement(
        Field,
        {
          href: href,
          onClick: onClick,
          onFocus: onFocus,
          onMouseOver: onMouseOver
        },
        children
      );
    }
  }]);

  return Time;
}(Component);

export default Time;
