// @atlaskit/media-picker/dist/es2015/popup/components/app.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Component } from 'react';
import { connect, Provider } from 'react-redux';
import { ContextFactory } from '@atlaskit/media-core';
import ModalDialog from '@atlaskit/modal-dialog';
/* Components */
import Footer from './footer/footer';
import Sidebar from './sidebar/sidebar';
import UploadView from './views/upload/upload';
import GiphyView from './views/giphy/giphyView';
import Browser from './views/browser/browser';
import { Dropzone } from './dropzone/dropzone';
import MainEditorView from './views/editor/mainEditorView';
/* Configs */
import { RECENTS_COLLECTION } from '../config';
/* actions */
import { startApp } from '../actions/startApp';
import { hidePopup } from '../actions/hidePopup';
import { fileUploadsStart } from '../actions/fileUploadsStart';
import { fileUploadPreviewUpdate } from '../actions/fileUploadPreviewUpdate';
import { fileUploadProgress } from '../actions/fileUploadProgress';
import { fileUploadProcessingStart } from '../actions/fileUploadProcessingStart';
import { fileUploadEnd } from '../actions/fileUploadEnd';
import { fileUploadError } from '../actions/fileUploadError';
import { MediaPicker } from '../..';
import PassContext from './passContext';
import { MediaPickerPopupWrapper, SidebarWrapper, ViewWrapper } from './styled';
var App = /** @class */ (function (_super) {
    tslib_1.__extends(App, _super);
    function App(props) {
        var _this = _super.call(this, props) || this;
        _this.setDropzoneActive = function (isDropzoneActive) {
            _this.setState({
                isDropzoneActive: isDropzoneActive,
            });
        };
        var onStartApp = props.onStartApp, onUploadsStart = props.onUploadsStart, onUploadPreviewUpdate = props.onUploadPreviewUpdate, onUploadStatusUpdate = props.onUploadStatusUpdate, onUploadProcessing = props.onUploadProcessing, onUploadEnd = props.onUploadEnd, onUploadError = props.onUploadError, context = props.context;
        var userAuthProvider = context.config.userAuthProvider;
        if (!userAuthProvider) {
            throw new Error('userAuthProvider must be provided in the context');
        }
        _this.state = {
            isDropzoneActive: false,
        };
        var defaultConfig = {
            uploadParams: {
                collection: RECENTS_COLLECTION,
            },
        };
        // We can't just use the given context since the Cards in the recents view needs a different authProvider
        _this.mpContext = ContextFactory.create({
            serviceHost: context.config.serviceHost,
            authProvider: userAuthProvider,
        });
        _this.mpBrowser = MediaPicker('browser', _this.mpContext, tslib_1.__assign({}, defaultConfig, { multiple: true, useNewUploadService: _this.props.useNewUploadService }));
        _this.mpBrowser.on('uploads-start', onUploadsStart);
        _this.mpBrowser.on('upload-preview-update', onUploadPreviewUpdate);
        _this.mpBrowser.on('upload-status-update', onUploadStatusUpdate);
        _this.mpBrowser.on('upload-processing', onUploadProcessing);
        _this.mpBrowser.on('upload-end', onUploadEnd);
        _this.mpBrowser.on('upload-error', onUploadError);
        _this.mpDropzone = MediaPicker('dropzone', _this.mpContext, tslib_1.__assign({}, defaultConfig, { headless: true, useNewUploadService: _this.props.useNewUploadService }));
        _this.mpDropzone.on('drag-enter', function () { return _this.setDropzoneActive(true); });
        _this.mpDropzone.on('drag-leave', function () { return _this.setDropzoneActive(false); });
        _this.mpDropzone.on('uploads-start', onUploadsStart);
        _this.mpDropzone.on('upload-preview-update', onUploadPreviewUpdate);
        _this.mpDropzone.on('upload-status-update', onUploadStatusUpdate);
        _this.mpDropzone.on('upload-processing', onUploadProcessing);
        _this.mpDropzone.on('upload-end', onUploadEnd);
        _this.mpDropzone.on('upload-error', onUploadError);
        _this.mpBinary = MediaPicker('binary', _this.mpContext, tslib_1.__assign({}, defaultConfig, { useNewUploadService: _this.props.useNewUploadService }));
        _this.mpBinary.on('uploads-start', onUploadsStart);
        _this.mpBinary.on('upload-preview-update', onUploadPreviewUpdate);
        _this.mpBinary.on('upload-status-update', onUploadStatusUpdate);
        _this.mpBinary.on('upload-processing', onUploadProcessing);
        _this.mpBinary.on('upload-end', onUploadEnd);
        _this.mpBinary.on('upload-error', onUploadError);
        onStartApp({
            onCancelUpload: function (uploadId) {
                _this.mpBrowser.cancel(uploadId);
                _this.mpDropzone.cancel(uploadId);
                _this.mpBinary.cancel(uploadId);
            },
        });
        return _this;
    }
    App.prototype.componentWillReceiveProps = function (_a) {
        var isVisible = _a.isVisible;
        if (isVisible !== this.props.isVisible) {
            if (isVisible) {
                this.mpDropzone.activate();
            }
            else {
                this.mpDropzone.deactivate();
            }
        }
    };
    App.prototype.componentWillUnmount = function () {
        this.mpDropzone.deactivate();
    };
    App.prototype.render = function () {
        var _a = this.props, selectedServiceName = _a.selectedServiceName, isVisible = _a.isVisible, onClose = _a.onClose, store = _a.store;
        var isDropzoneActive = this.state.isDropzoneActive;
        if (!isVisible) {
            return null;
        }
        return (React.createElement(Provider, { store: store },
            React.createElement(ModalDialog, { onClose: onClose, width: "x-large", isChromeless: true },
                React.createElement(PassContext, { store: store },
                    React.createElement(MediaPickerPopupWrapper, null,
                        React.createElement(SidebarWrapper, null,
                            React.createElement(Sidebar, null)),
                        React.createElement(ViewWrapper, null,
                            this.renderCurrentView(selectedServiceName),
                            React.createElement(Footer, null)),
                        React.createElement(Dropzone, { isActive: isDropzoneActive }),
                        React.createElement(MainEditorView, { binaryUploader: this.mpBinary }))))));
    };
    App.prototype.renderCurrentView = function (selectedServiceName) {
        if (selectedServiceName === 'upload') {
            return (React.createElement(UploadView, { mpBrowser: this.mpBrowser, context: this.mpContext, recentsCollection: RECENTS_COLLECTION }));
        }
        else if (selectedServiceName === 'giphy') {
            return React.createElement(GiphyView, null);
        }
        else {
            return React.createElement(Browser, null);
        }
    };
    return App;
}(Component));
export { App };
var mapStateToProps = function (_a) {
    var view = _a.view, context = _a.context, useNewUploadService = _a.useNewUploadService;
    return ({
        selectedServiceName: view.service.name,
        isVisible: view.isVisible,
        useNewUploadService: useNewUploadService,
        context: context,
    });
};
var mapDispatchToProps = function (dispatch) { return ({
    onStartApp: function (payload) { return dispatch(startApp(payload)); },
    onUploadsStart: function (payload) {
        return dispatch(fileUploadsStart(payload));
    },
    onClose: function () { return dispatch(hidePopup()); },
    onUploadPreviewUpdate: function (payload) {
        return dispatch(fileUploadPreviewUpdate(payload));
    },
    onUploadStatusUpdate: function (payload) {
        return dispatch(fileUploadProgress(payload));
    },
    onUploadProcessing: function (payload) {
        return dispatch(fileUploadProcessingStart(payload));
    },
    onUploadEnd: function (payload) {
        return dispatch(fileUploadEnd(payload));
    },
    onUploadError: function (payload) {
        return dispatch(fileUploadError(payload));
    },
}); };
export default connect(mapStateToProps, mapDispatchToProps)(App);
