// @atlaskit/media-card/dist/es2015/links/embed/urlEmbedCard/styled.js
/* tslint:disable variable-name */
import * as tslib_1 from "tslib";
import styled from 'styled-components';
export var defaultWidth = 480;
export var defaultHeight = 180;
export var maxWidth = 480;
export var maxHeight = 360;
export var Wrapper = styled.div(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n  position: relative;\n  ", ";\n"], ["\n  position: relative;\n  ",
    ";\n"])), function (_a) {
    var aspectRatio = _a.aspectRatio, width = _a.width, height = _a.height;
    if (width || height) {
        return "\n        width: " + (width || defaultWidth) + "px;\n        height: " + (height || defaultHeight) + "px;\n        max-width: " + maxWidth + "px;\n        max-height: " + maxHeight + "px;\n      ";
    }
    else if (aspectRatio) {
        return "\n        width: " + defaultWidth + "px;\n        /*\n          - cannot use a % height when the parent container does not have a fixed height\n          - padding-bottom % is calculated on the width of the parent container\n         */\n        height: 0;\n        padding-bottom: " + 1 / aspectRatio * 100 + "%;\n      ";
    }
    else {
        return "\n        width: " + defaultWidth + "px;\n        height: " + defaultHeight + "px;\n      ";
    }
});
export var Iframe = styled.iframe(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n  position: absolute;\n  width: 100%;\n  height: 100%;\n  border: none;\n  border-radius: 3px;\n"], ["\n  position: absolute;\n  width: 100%;\n  height: 100%;\n  border: none;\n  border-radius: 3px;\n"])));
var templateObject_1, templateObject_2;
