// @atlaskit/media-picker/dist/es5/popup/components/views/browser/auth/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var react_1 = require("react");
var react_redux_1 = require("react-redux");
var button_1 = require("@atlaskit/button");
var dropbox_1 = require("@atlaskit/icon/glyph/dropbox");
var googledrive_1 = require("@atlaskit/icon/glyph/googledrive");
var actions_1 = require("../../../../actions");
var styled_1 = require("./styled");
var serviceDetails = {
    dropbox: {
        name: 'Dropbox',
        icon: React.createElement(dropbox_1.default, { label: "dropbox", size: "xlarge" }),
    },
    google: {
        name: 'Google Drive',
        icon: React.createElement(googledrive_1.default, { label: "drive", size: "xlarge" }),
    },
};
/**
 * Routing class that displays view depending on situation.
 */
var StatelessAuth = /** @class */ (function (_super) {
    tslib_1.__extends(StatelessAuth, _super);
    function StatelessAuth() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onClick = function () { return _this.props.onStartAuth(_this.props.service.name); };
        return _this;
    }
    StatelessAuth.prototype.render = function () {
        var service = this.props.service;
        var details = serviceDetails[service.name];
        if (!details) {
            return null;
        }
        var name = details.name, icon = details.icon;
        var btnLabel = "Connect to " + name;
        return (React.createElement(styled_1.ConnectWrapper, null,
            React.createElement(styled_1.Title, null,
                "Upload a file from ",
                name),
            React.createElement(styled_1.IconWrapper, null, icon),
            React.createElement(styled_1.ButtonWrapper, null,
                React.createElement(button_1.default, { appearance: "primary", className: "connectBtn", onClick: this.onClick }, btnLabel)),
            React.createElement(styled_1.TextDescription, null,
                "We'll open a new page to help you",
                React.createElement("br", null),
                " connect your ",
                name,
                " account")));
    };
    return StatelessAuth;
}(react_1.Component));
exports.StatelessAuth = StatelessAuth;
exports.default = react_redux_1.connect(function (state) { return ({
    service: state.view.service,
}); }, function (dispatch) { return ({
    onStartAuth: function (serviceName) { return dispatch(actions_1.startAuth(serviceName)); },
}); })(StatelessAuth);
