// @atlaskit/editor-core/dist/es2015/plugins/hyperlink/index.js
import * as React from 'react';
import { link, WithProviders } from '@atlaskit/editor-common';
import { createPlugin, hyperlinkPluginKey } from './pm-plugins/main';
import { createInputRulePlugin } from './pm-plugins/input-rule';
import { createKeymapPlugin } from './pm-plugins/keymap';
import HyperlinkEdit from './ui/HyperlinkEdit';
export { hyperlinkPluginKey };
var hyperlinkPlugin = {
    marks: function () {
        return [{ name: 'link', mark: link, rank: 100 }];
    },
    pmPlugins: function () {
        return [
            { rank: 900, plugin: function (_a) {
                    var schema = _a.schema, props = _a.props;
                    return createPlugin(schema, props);
                } },
            { rank: 910, plugin: function (_a) {
                    var schema = _a.schema;
                    return createInputRulePlugin(schema);
                } },
            {
                rank: 920,
                plugin: function (_a) {
                    var schema = _a.schema, props = _a.props;
                    return createKeymapPlugin(schema, props);
                },
            },
        ];
    },
    contentComponent: function (_a) {
        var editorView = _a.editorView, providerFactory = _a.providerFactory, appearance = _a.appearance, popupsMountPoint = _a.popupsMountPoint, popupsBoundariesElement = _a.popupsBoundariesElement, popupsScrollableElement = _a.popupsScrollableElement;
        if (appearance === 'message') {
            return null;
        }
        var renderNode = function (providers) {
            var pluginState = hyperlinkPluginKey.getState(editorView.state);
            return (React.createElement(HyperlinkEdit, { editorView: editorView, pluginState: pluginState, activityProvider: providers.activityProvider, popupsMountPoint: popupsMountPoint, popupsBoundariesElement: popupsBoundariesElement, popupsScrollableElement: popupsScrollableElement }));
        };
        return (React.createElement(WithProviders, { providerFactory: providerFactory, providers: ['activityProvider'], renderNode: renderNode }));
    },
};
export default hyperlinkPlugin;
