// @atlaskit/media-card/dist/es5/links/embed/urlEmbedCard/styled.js
"use strict";
/* tslint:disable variable-name */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var styled_components_1 = require("styled-components");
exports.defaultWidth = 480;
exports.defaultHeight = 180;
exports.maxWidth = 480;
exports.maxHeight = 360;
exports.Wrapper = styled_components_1.default.div(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n  position: relative;\n  ", ";\n"], ["\n  position: relative;\n  ",
    ";\n"])), function (_a) {
    var aspectRatio = _a.aspectRatio, width = _a.width, height = _a.height;
    if (width || height) {
        return "\n        width: " + (width || exports.defaultWidth) + "px;\n        height: " + (height || exports.defaultHeight) + "px;\n        max-width: " + exports.maxWidth + "px;\n        max-height: " + exports.maxHeight + "px;\n      ";
    }
    else if (aspectRatio) {
        return "\n        width: " + exports.defaultWidth + "px;\n        /*\n          - cannot use a % height when the parent container does not have a fixed height\n          - padding-bottom % is calculated on the width of the parent container\n         */\n        height: 0;\n        padding-bottom: " + 1 / aspectRatio * 100 + "%;\n      ";
    }
    else {
        return "\n        width: " + exports.defaultWidth + "px;\n        height: " + exports.defaultHeight + "px;\n      ";
    }
});
exports.Iframe = styled_components_1.default.iframe(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n  position: absolute;\n  width: 100%;\n  height: 100%;\n  border: none;\n  border-radius: 3px;\n"], ["\n  position: absolute;\n  width: 100%;\n  height: 100%;\n  border: none;\n  border-radius: 3px;\n"])));
var templateObject_1, templateObject_2;
