// @atlaskit/editor-core/dist/es2015/plugins/quick-insert/index.js
import * as React from 'react';
import InfoIcon from '@atlaskit/icon/glyph/editor/info';
import DateIcon from '@atlaskit/icon/glyph/editor/date';
import TableIcon from '@atlaskit/icon/glyph/editor/table';
import { createTableNode } from '../table/utils';
import { find } from './search';
var ITEMS = [
    {
        title: 'Panel',
        icon: function () { return React.createElement(InfoIcon, { label: "Insert Panel" }); },
        action: function (state, replaceWith) {
            return replaceWith(state.schema.nodes.panel.createChecked({}, state.schema.nodes.paragraph.createChecked()));
        },
    },
    {
        title: 'Table',
        icon: function () { return React.createElement(TableIcon, { label: "Insert table" }); },
        action: function (state, replaceWith) {
            return replaceWith(createTableNode(3, 3, state.schema));
        },
    },
    {
        title: 'Date',
        icon: function () { return React.createElement(DateIcon, { label: "Insert Date" }); },
        action: function (state, replaceWith) {
            return replaceWith(state.schema.nodes.date.createChecked({ timestamp: Date.now() }));
        },
    },
    {
        title: 'Browse all...',
        action: function () {
            document.location.href = 'https://atlassian.com';
            return true;
        },
    },
];
var quickInsertPlugin = {
    pluginsOptions: {
        typeAhead: {
            trigger: '/',
            getItems: function (query) {
                return new Promise(function (resolve) {
                    setTimeout(function () {
                        resolve(find(query, ITEMS, function (item) { return item.title; }));
                    }, 100);
                });
            },
            selectItem: function (state, item, replaceWith) {
                return item.action(state, replaceWith);
            },
        },
    },
};
export default quickInsertPlugin;
