// @atlaskit/media-core/dist/es5/services/collectionService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var createRequest_1 = require("./util/createRequest");
exports.DEFAULT_COLLECTION_PAGE_SIZE = 10;
var MediaCollectionService = /** @class */ (function () {
    function MediaCollectionService(config) {
        this.config = config;
    }
    MediaCollectionService.prototype.mapToMediaCollectionItem = function (item) {
        var id = item.id, type = item.type, occurrenceKey = item.occurrenceKey, details = item.details;
        if (type === 'file') {
            var fileDetails = details;
            return {
                type: 'file',
                details: tslib_1.__assign({ id: id,
                    occurrenceKey: occurrenceKey }, fileDetails),
            };
        }
        else {
            var linkDetails = details;
            return {
                type: 'link',
                details: tslib_1.__assign({ id: id, type: 'link', occurrenceKey: occurrenceKey, title: linkDetails.title || '' }, linkDetails),
            };
        }
    };
    MediaCollectionService.prototype.getCollectionItems = function (collectionName, limit, inclusiveStartKey, sortDirection, details) {
        var _this = this;
        if (limit === void 0) { limit = exports.DEFAULT_COLLECTION_PAGE_SIZE; }
        var request = createRequest_1.default({
            config: this.config,
            preventPreflight: true,
            collectionName: collectionName,
        });
        return request({
            url: "/collection/" + collectionName + "/items",
            params: {
                limit: limit,
                inclusiveStartKey: inclusiveStartKey,
                sortDirection: sortDirection,
                details: details,
            },
        }).response.then(function (response) {
            return {
                items: response.data.contents.map(_this.mapToMediaCollectionItem),
                nextInclusiveStartKey: response.data.nextInclusiveStartKey,
            };
        });
    };
    return MediaCollectionService;
}());
exports.MediaCollectionService = MediaCollectionService;
