// @atlaskit/media-core/dist/es5/services/dataUriService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var blobService_1 = require("./blobService");
var MediaDataUriService = /** @class */ (function () {
    function MediaDataUriService(authProvider, serviceHost, collectionName) {
        this.authProvider = authProvider;
        this.serviceHost = serviceHost;
        this.collectionName = collectionName;
        this.blobService = new blobService_1.MediaBlobService(authProvider, serviceHost, collectionName);
    }
    MediaDataUriService.prototype.fetchOriginalDataUri = function (mediaItem) {
        return this.blobService.fetchOriginalBlob(mediaItem).then(this.readBlob);
    };
    MediaDataUriService.prototype.fetchImageDataUri = function (mediaItem, options) {
        return this.blobService
            .fetchImageBlob(mediaItem, options)
            .then(this.readBlob);
    };
    MediaDataUriService.prototype.readBlob = function (blob) {
        return new Promise(function (resolve, reject) {
            var reader = new FileReader();
            reader.addEventListener('load', function () { return resolve(reader.result); });
            reader.addEventListener('error', function () { return reject(reader.error); });
            reader.readAsDataURL(blob);
        });
    };
    return MediaDataUriService;
}());
exports.MediaDataUriService = MediaDataUriService;
