// @atlaskit/media-core/dist/es5/services/fileService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var createRequest_1 = require("./util/createRequest");
var MediaFileService = /** @class */ (function () {
    function MediaFileService(config, fileItemCache) {
        this.config = config;
        this.fileItemCache = fileItemCache;
    }
    MediaFileService.prototype.getFileItem = function (fileId, collectionName) {
        var _this = this;
        var cacheKey = [fileId, 'file'].join('-');
        var cachedValue = this.fileItemCache.get(cacheKey);
        if (cachedValue) {
            return Promise.resolve(cachedValue);
        }
        else {
            var request = createRequest_1.default({
                config: this.config,
                collectionName: collectionName,
                preventPreflight: true,
            });
            return request({ url: "/file/" + fileId })
                .response.then(function (json) { return json.data; })
                .then(function (fileDetails) {
                var fileItem = {
                    type: 'file',
                    details: {
                        id: fileDetails.id,
                        mediaType: fileDetails.mediaType,
                        mimeType: fileDetails.mimeType,
                        name: fileDetails.name,
                        processingStatus: fileDetails.processingStatus,
                        size: fileDetails.size,
                        artifacts: fileDetails.artifacts,
                    },
                };
                if (fileDetails.processingStatus === 'succeeded') {
                    _this.fileItemCache.set(cacheKey, fileItem);
                }
                return fileItem;
            });
        }
    };
    return MediaFileService;
}());
exports.MediaFileService = MediaFileService;
