// @atlaskit/comment/dist/cjs/components/Author.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Field = require('./Field');

var _Field2 = _interopRequireDefault(_Field);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Author = function (_Component) {
  (0, _inherits3.default)(Author, _Component);

  function Author() {
    (0, _classCallCheck3.default)(this, Author);
    return (0, _possibleConstructorReturn3.default)(this, (Author.__proto__ || (0, _getPrototypeOf2.default)(Author)).apply(this, arguments));
  }

  (0, _createClass3.default)(Author, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          href = _props.href,
          onClick = _props.onClick,
          onFocus = _props.onFocus,
          onMouseOver = _props.onMouseOver;

      return _react2.default.createElement(
        _Field2.default,
        {
          hasAuthor: true,
          href: href,
          onClick: onClick,
          onFocus: onFocus,
          onMouseOver: onMouseOver
        },
        children
      );
    }
  }]);
  return Author;
}(_react.Component);

exports.default = Author;
