// @atlaskit/comment/dist/cjs/components/Comment.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Layout = require('./Layout');

var _Layout2 = _interopRequireDefault(_Layout);

var _Header = require('./Header');

var _Header2 = _interopRequireDefault(_Header);

var _Footer = require('./Footer');

var _Footer2 = _interopRequireDefault(_Footer);

var _CommentStyles = require('../styled/CommentStyles');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Comment = function (_Component) {
  (0, _inherits3.default)(Comment, _Component);

  function Comment() {
    (0, _classCallCheck3.default)(this, Comment);
    return (0, _possibleConstructorReturn3.default)(this, (Comment.__proto__ || (0, _getPrototypeOf2.default)(Comment)).apply(this, arguments));
  }

  (0, _createClass3.default)(Comment, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          actions = _props.actions,
          author = _props.author,
          avatar = _props.avatar,
          children = _props.children,
          content = _props.content,
          edited = _props.edited,
          errorActions = _props.errorActions,
          errorIconLabel = _props.errorIconLabel,
          highlighted = _props.highlighted,
          isError = _props.isError,
          isSaving = _props.isSaving,
          restrictedTo = _props.restrictedTo,
          savingText = _props.savingText,
          time = _props.time,
          type = _props.type,
          id = _props.id;


      var headerProps = {
        author: author,
        edited: edited,
        isError: isError,
        isSaving: isSaving,
        restrictedTo: restrictedTo,
        savingText: savingText,
        time: time,
        type: type
      };
      var footerProps = {
        actions: actions,
        errorActions: errorActions,
        errorIconLabel: errorIconLabel,
        isError: isError,
        isSaving: isSaving
      };
      var layoutContent = _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(_Header2.default, headerProps),
        _react2.default.createElement(
          _CommentStyles.Content,
          { isDisabled: isSaving || isError },
          content
        ),
        _react2.default.createElement(_Footer2.default, footerProps)
      );

      return _react2.default.createElement(
        _Layout2.default,
        {
          id: id,
          avatar: avatar,
          content: layoutContent,
          highlighted: highlighted
        },
        children
      );
    }
  }]);
  return Comment;
}(_react.Component);

Comment.defaultProps = {
  actions: [],
  restrictedTo: '',
  highlighted: false,
  isSaving: false,
  savingText: 'Sending...',
  isError: false,
  errorActions: [],
  errorIconLabel: ''
};
exports.default = Comment;
