// @atlaskit/comment/dist/cjs/components/Field.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _FieldStyles = require('../styled/FieldStyles');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CommentField = function (_Component) {
  (0, _inherits3.default)(CommentField, _Component);

  function CommentField() {
    (0, _classCallCheck3.default)(this, CommentField);
    return (0, _possibleConstructorReturn3.default)(this, (CommentField.__proto__ || (0, _getPrototypeOf2.default)(CommentField)).apply(this, arguments));
  }

  (0, _createClass3.default)(CommentField, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          hasAuthor = _props.hasAuthor,
          href = _props.href,
          onClick = _props.onClick,
          onFocus = _props.onFocus,
          onMouseOver = _props.onMouseOver;
      /* eslint-disable jsx-a11y/no-static-element-interactions */

      return href ? _react2.default.createElement(
        _FieldStyles.Anchor,
        {
          href: href,
          hasAuthor: hasAuthor,
          onClick: onClick,
          onFocus: onFocus,
          onMouseOver: onMouseOver
        },
        children
      ) : _react2.default.createElement(
        _FieldStyles.Span,
        {
          hasAuthor: hasAuthor,
          onClick: onClick,
          onFocus: onFocus,
          onMouseOver: onMouseOver
        },
        children
      );
      /* eslint-enable jsx-a11y/no-static-element-interactions */
    }
  }]);
  return CommentField;
}(_react.Component);

exports.default = CommentField;
