// @atlaskit/badge/dist/cjs/Badge/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _theme = require('@atlaskit/theme');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styled = require('./styled');

var _theme2 = require('../theme');

var theme = _interopRequireWildcard(_theme2);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getValue(value, max) {
  if (value < 0) {
    return '0';
  }
  if (max > 0 && value > max) {
    return max + '+';
  }
  if (value === Infinity) {
    return '∞';
  }
  return String(value);
}

var Badge = function (_PureComponent) {
  (0, _inherits3.default)(Badge, _PureComponent);

  function Badge() {
    (0, _classCallCheck3.default)(this, Badge);
    return (0, _possibleConstructorReturn3.default)(this, (Badge.__proto__ || (0, _getPrototypeOf2.default)(Badge)).apply(this, arguments));
  }

  (0, _createClass3.default)(Badge, [{
    key: 'componentWillUpdate',
    value: function componentWillUpdate(nextProps) {
      var _props = this.props,
          onValueUpdated = _props.onValueUpdated,
          oldValue = _props.value;
      var newValue = nextProps.value;


      if (onValueUpdated && newValue !== oldValue) {
        onValueUpdated({ oldValue: oldValue, newValue: newValue });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          appearance = _props2.appearance,
          max = _props2.max,
          value = _props2.value;

      var computedValue = getValue(value, max);
      return _react2.default.createElement(
        _theme.Appearance,
        { props: appearance, theme: theme },
        function (styleProps) {
          return _react2.default.createElement(
            _styled.BadgeElement,
            styleProps,
            computedValue
          );
        }
      );
    }
  }]);
  return Badge;
}(_react.PureComponent);

Badge.defaultProps = {
  appearance: 'default',
  max: 99,
  value: 0
};
exports.default = Badge;
