// @atlaskit/editor-core/dist/es2015/plugins/block-type/ui/ToolbarBlockType/index.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import ExpandIcon from '@atlaskit/icon/glyph/chevron-down';
import TextStyleIcon from '@atlaskit/icon/glyph/editor/text-style';
import { analyticsService as analytics } from '../../../../analytics';
import ToolbarButton from '../../../../ui/ToolbarButton';
import DropdownMenu from '../../../../ui/DropdownMenu';
import { ButtonContent, Separator, Wrapper, MenuWrapper, ExpandIconWrapper, } from '../../../../ui/styles';
var ToolbarBlockType = /** @class */ (function (_super) {
    tslib_1.__extends(ToolbarBlockType, _super);
    function ToolbarBlockType(props) {
        var _this = _super.call(this, props) || this;
        _this.onOpenChange = function (attrs) {
            _this.setState({
                active: attrs.isOpen,
            });
        };
        _this.handleTriggerClick = function () {
            _this.onOpenChange({ isOpen: !_this.state.active });
        };
        _this.createItems = function () {
            var _a = _this.state, currentBlockType = _a.currentBlockType, availableBlockTypes = _a.availableBlockTypes;
            var items = [];
            availableBlockTypes.forEach(function (blockType, blockTypeNo) {
                items.push({
                    content: blockType.title,
                    value: blockType,
                    // ED-2853, hiding tooltips as shortcuts are not working atm.
                    // tooltipDescription: tooltip(findKeymapByDescription(blockType.title)),
                    // tooltipPosition: 'right',
                    isActive: currentBlockType === blockType,
                });
            });
            return [
                {
                    items: items,
                },
            ];
        };
        _this.handlePluginStateChange = function (pluginState) {
            _this.setState({
                active: _this.state.active,
                availableBlockTypes: pluginState.availableBlockTypes,
                currentBlockType: pluginState.currentBlockType,
                blockTypesDisabled: pluginState.blockTypesDisabled,
            });
        };
        _this.handleSelectBlockType = function (_a) {
            var item = _a.item;
            var blockType = item.value;
            var availableBlockTypes = _this.state.availableBlockTypes;
            _this.props.pluginState.setBlockType(blockType.name, _this.props.editorView);
            _this.setState({
                active: false,
                availableBlockTypes: availableBlockTypes,
                currentBlockType: blockType,
            });
            analytics.trackEvent("atlassian.editor.format." + blockType.name + ".button");
        };
        var pluginState = props.pluginState;
        _this.state = {
            active: false,
            availableBlockTypes: pluginState.availableBlockTypes,
            currentBlockType: pluginState.currentBlockType,
            blockTypesDisabled: pluginState.blockTypesDisabled,
        };
        return _this;
    }
    ToolbarBlockType.prototype.componentDidMount = function () {
        this.props.pluginState.subscribe(this.handlePluginStateChange);
    };
    ToolbarBlockType.prototype.componentWillUnmount = function () {
        this.props.pluginState.unsubscribe(this.handlePluginStateChange);
    };
    ToolbarBlockType.prototype.render = function () {
        var _this = this;
        var _a = this.state, active = _a.active, currentBlockType = _a.currentBlockType, blockTypesDisabled = _a.blockTypesDisabled, availableBlockTypes = _a.availableBlockTypes;
        var _b = this.props, popupsMountPoint = _b.popupsMountPoint, popupsBoundariesElement = _b.popupsBoundariesElement, popupsScrollableElement = _b.popupsScrollableElement, isSmall = _b.isSmall;
        var blockTypeTitles = availableBlockTypes
            .filter(function (blockType) { return blockType.name === currentBlockType.name; })
            .map(function (blockType) { return blockType.title; });
        var toolbarButtonFactory = function (disabled) { return (React.createElement(ToolbarButton, { spacing: "none", selected: active, disabled: disabled, onClick: _this.handleTriggerClick, iconAfter: React.createElement(Wrapper, { isSmall: isSmall },
                isSmall && React.createElement(TextStyleIcon, { label: "Change formatting" }),
                React.createElement(ExpandIconWrapper, null,
                    React.createElement(ExpandIcon, { label: "Change formatting" }))) }, !isSmall && (React.createElement(ButtonContent, null, blockTypeTitles[0] || 'Normal text')))); };
        if (!this.props.isDisabled && !blockTypesDisabled) {
            var items = this.createItems();
            return (React.createElement(MenuWrapper, null,
                React.createElement(DropdownMenu, { items: items, onOpenChange: this.onOpenChange, onItemActivated: this.handleSelectBlockType, isOpen: active, mountTo: popupsMountPoint, boundariesElement: popupsBoundariesElement, scrollableElement: popupsScrollableElement, fitHeight: 360, fitWidth: 106 }, toolbarButtonFactory(false)),
                React.createElement(Separator, null)));
        }
        return (React.createElement(Wrapper, null,
            toolbarButtonFactory(true),
            React.createElement(Separator, null)));
    };
    return ToolbarBlockType;
}(React.PureComponent));
export default ToolbarBlockType;
