// @atlaskit/media-picker/dist/es5/popup/middleware/getPreview.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var getPreview_1 = require("../actions/getPreview");
var sendUploadEvent_1 = require("../actions/sendUploadEvent");
function default_1(fetcher) {
    return function (store) { return function (next) { return function (action) {
        if (getPreview_1.isGetPreviewAction(action)) {
            getPreview(fetcher, store, action);
        }
        return next(action);
    }; }; };
}
exports.default = default_1;
function getPreview(fetcher, store, _a) {
    var uploadId = _a.uploadId, file = _a.file, collection = _a.collection;
    var _b = store.getState(), apiUrl = _b.apiUrl, userAuthProvider = _b.userAuthProvider;
    return userAuthProvider()
        .then(function (auth) { return fetcher.getPreview(apiUrl, auth, file.id, collection); })
        .then(function (preview) {
        return store.dispatch(sendUploadEvent_1.sendUploadEvent({
            event: {
                name: 'upload-preview-update',
                data: {
                    file: file,
                    preview: preview,
                },
            },
            uploadId: uploadId,
        }));
    });
}
exports.getPreview = getPreview;
