// @atlaskit/media-picker/dist/es5/popup/middleware/importFiles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var uuid = require("uuid");
var startImport_1 = require("../actions/startImport");
var finalizeUpload_1 = require("../actions/finalizeUpload");
var remoteUploadStart_1 = require("../actions/remoteUploadStart");
var getPreview_1 = require("../actions/getPreview");
var handleCloudFetchingEvent_1 = require("../actions/handleCloudFetchingEvent");
var setEventProxy_1 = require("../actions/setEventProxy");
var hidePopup_1 = require("../actions/hidePopup");
var config_1 = require("../config");
var remoteUploadActivity_1 = require("../tools/websocket/upload/remoteUploadActivity");
var file_1 = require("../../domain/file");
var sendUploadEvent_1 = require("../actions/sendUploadEvent");
exports.isRemoteFileItem = function (item) {
    return ['dropbox', 'google', 'giphy'].indexOf(item.serviceName) !== -1;
};
exports.isRemoteService = function (serviceName) {
    return ['dropbox', 'google', 'giphy'].indexOf(serviceName) !== -1;
};
var mapSelectedItemToSelectedUploadFile = function (_a) {
    var id = _a.id, name = _a.name, mimeType = _a.mimeType, size = _a.size, date = _a.date, serviceName = _a.serviceName, accountId = _a.accountId;
    return ({
        file: {
            id: id,
            name: name,
            size: size,
            creationDate: date || Date.now(),
            type: mimeType,
        },
        serviceName: serviceName,
        accountId: accountId,
        uploadId: uuid.v4(),
    });
};
function importFilesMiddleware(eventEmitter, wsProvider) {
    return function (store) { return function (next) { return function (action) {
        if (startImport_1.isStartImportAction(action)) {
            importFiles(eventEmitter, store, wsProvider);
        }
        return next(action);
    }; }; };
}
exports.importFilesMiddleware = importFilesMiddleware;
function importFiles(eventEmitter, store, wsProvider) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var _a, apiUrl, uploads, tenant, selectedItems, userAuthProvider, auth, selectedUploadFiles;
        return tslib_1.__generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = store.getState(), apiUrl = _a.apiUrl, uploads = _a.uploads, tenant = _a.tenant, selectedItems = _a.selectedItems, userAuthProvider = _a.userAuthProvider;
                    store.dispatch(hidePopup_1.hidePopup());
                    return [4 /*yield*/, userAuthProvider()];
                case 1:
                    auth = _b.sent();
                    selectedUploadFiles = selectedItems.map(mapSelectedItemToSelectedUploadFile);
                    eventEmitter.emitUploadsStart(selectedUploadFiles.map(function (_a) {
                        var file = _a.file, uploadId = _a.uploadId;
                        return file_1.copyMediaFileForUpload(file, uploadId);
                    }));
                    selectedUploadFiles.forEach(function (selectedUploadFile) {
                        var file = selectedUploadFile.file, serviceName = selectedUploadFile.serviceName, uploadId = selectedUploadFile.uploadId;
                        var selectedItemId = file.id;
                        if (serviceName === 'upload') {
                            var localUpload = uploads[selectedItemId];
                            exports.importFilesFromLocalUpload(selectedItemId, tenant, uploadId, store, localUpload);
                        }
                        else if (serviceName === 'recent_files') {
                            exports.importFilesFromRecentFiles(selectedUploadFile, tenant, store);
                        }
                        else if (exports.isRemoteService(serviceName)) {
                            var wsConnectionHolder = wsProvider.getWsConnectionHolder(apiUrl, auth);
                            exports.importFilesFromRemoteService(selectedUploadFile, tenant, store, wsConnectionHolder);
                        }
                    });
                    return [2 /*return*/];
            }
        });
    });
}
exports.importFiles = importFiles;
exports.importFilesFromLocalUpload = function (selectedItemId, tenant, uploadId, store, localUpload) {
    localUpload.events.forEach(function (originalEvent) {
        var event = tslib_1.__assign({}, originalEvent);
        if (event.name === 'upload-processing') {
            var file = event.data.file;
            var source = {
                id: file.publicId,
                collection: config_1.RECENTS_COLLECTION,
            };
            store.dispatch(finalizeUpload_1.finalizeUpload(file, uploadId, source, tenant));
        }
        else if (event.name !== 'upload-end') {
            store.dispatch(sendUploadEvent_1.sendUploadEvent({ event: event, uploadId: uploadId }));
        }
    });
    store.dispatch(setEventProxy_1.setEventProxy(selectedItemId, uploadId));
};
exports.importFilesFromRecentFiles = function (selectedUploadFile, tenant, store) {
    var file = selectedUploadFile.file, uploadId = selectedUploadFile.uploadId;
    var source = {
        id: file.id,
        collection: config_1.RECENTS_COLLECTION,
    };
    store.dispatch(finalizeUpload_1.finalizeUpload(file, uploadId, source, tenant));
    store.dispatch(getPreview_1.getPreview(uploadId, file, config_1.RECENTS_COLLECTION));
};
exports.importFilesFromRemoteService = function (selectedUploadFile, tenant, store, wsConnectionHolder) {
    var uploadId = selectedUploadFile.uploadId, serviceName = selectedUploadFile.serviceName, accountId = selectedUploadFile.accountId, file = selectedUploadFile.file;
    var uploadActivity = new remoteUploadActivity_1.RemoteUploadActivity(uploadId, function (event, payload) {
        // TODO figure out the difference between this uploadId and the last MSW-405
        var uploadId = payload.uploadId;
        var newFile = tslib_1.__assign({}, file, { id: uploadId, creationDate: Date.now() });
        store.dispatch(handleCloudFetchingEvent_1.handleCloudFetchingEvent(newFile, event, payload));
    });
    uploadActivity.on('Started', function () {
        store.dispatch(remoteUploadStart_1.remoteUploadStart(uploadId, tenant));
    });
    wsConnectionHolder.openConnection(uploadActivity);
    wsConnectionHolder.send({
        type: 'fetchFile',
        params: {
            serviceName: serviceName,
            accountId: accountId,
            fileId: file.id,
            fileName: file.name,
            collection: config_1.RECENTS_COLLECTION,
            jobId: uploadId,
        },
    });
};
