// @atlaskit/editor-common/dist/es5/schema/nodes/tableNodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util_shared_styles_1 = require("@atlaskit/util-shared-styles");
var utils_1 = require("../../utils");
var styles_1 = require("../../styles");
var getCellAttrs = function (dom) {
    var widthAttr = dom.getAttribute('data-colwidth');
    var width = widthAttr && /^\d+(,\d+)*$/.test(widthAttr)
        ? widthAttr.split(',').map(function (str) { return Number(str); })
        : null;
    var colspan = Number(dom.getAttribute('colspan') || 1);
    return {
        colspan: colspan,
        rowspan: Number(dom.getAttribute('rowspan') || 1),
        colwidth: width && width.length === colspan ? width : null,
        background: dom.style.backgroundColor || null,
    };
};
var setCellAttrs = function (node) {
    var attrs = {};
    if (node.attrs.colspan !== 1) {
        attrs.colspan = node.attrs.colspan;
    }
    if (node.attrs.rowspan !== 1) {
        attrs.rowspan = node.attrs.rowspan;
    }
    if (node.attrs.colwidth) {
        attrs['data-colwidth'] = node.attrs.colwidth.join(',');
    }
    if (node.attrs.background) {
        var background = node.attrs.background;
        var nodeType = node.type.name;
        // to ensure that we don't overwrite product's style:
        // - it clears background color for <th> if its set to grey
        // - it clears background color for <td> if its set to white
        var ignored = (nodeType === 'tableHeader' &&
            background === exports.tableBackgroundColorNames.get('grey')) ||
            (nodeType === 'tableCell' &&
                background === exports.tableBackgroundColorNames.get('white'));
        if (!ignored) {
            var color = nodeType === 'tableCell'
                ? utils_1.hexToRgba(background, styles_1.akEditorTableCellBackgroundOpacity)
                : background;
            attrs.style = (attrs.style || '') + "background-color: " + color + ";";
        }
    }
    return attrs;
};
exports.tableBackgroundColorPalette = new Map();
/** New borders for colors in the color picker */
exports.tableBackgroundBorderColors = {
    blue: util_shared_styles_1.akColorB75,
    teal: util_shared_styles_1.akColorT75,
    red: util_shared_styles_1.akColorR75,
    grey: util_shared_styles_1.akColorN40,
    purple: util_shared_styles_1.akColorP75,
    green: util_shared_styles_1.akColorG75,
    yellow: util_shared_styles_1.akColorY75,
    white: util_shared_styles_1.akColorN40,
};
exports.tableBackgroundColorNames = new Map();
[
    [util_shared_styles_1.akColorB50, 'Blue'],
    [util_shared_styles_1.akColorT50, 'Teal'],
    [util_shared_styles_1.akColorR50, 'Red'],
    [util_shared_styles_1.akColorN20, 'Grey'],
    [util_shared_styles_1.akColorP50, 'Purple'],
    [util_shared_styles_1.akColorG50, 'Green'],
    [util_shared_styles_1.akColorY50, 'Yellow'],
    [util_shared_styles_1.akColorN0, 'White'],
].forEach(function (_a) {
    var colorValue = _a[0], colorName = _a[1];
    exports.tableBackgroundColorPalette.set(colorValue.toLowerCase(), colorName);
    exports.tableBackgroundColorNames.set(colorName.toLowerCase(), colorValue.toLowerCase());
});
function calcTableColumnWidths(node) {
    var tableColumnWidths = [];
    var isNumberColumnEnabled = node.attrs.isNumberColumnEnabled;
    node.forEach(function (rowNode, _, i) {
        rowNode.forEach(function (colNode, _, j) {
            var colwidth = colNode.attrs.colwidth || [0];
            if (isNumberColumnEnabled && j === 0) {
                if (!colwidth) {
                    colwidth = [styles_1.akEditorTableNumberColumnWidth];
                }
            }
            // if we have a colwidth attr for this cell, and it contains new
            // colwidths we haven't seen for the whole table yet, add those
            // (colwidths over the table are defined as-we-go)
            if (colwidth && colwidth.length + j > tableColumnWidths.length) {
                tableColumnWidths = tableColumnWidths.slice(0, j).concat(colwidth);
            }
        });
    });
    return tableColumnWidths;
}
exports.calcTableColumnWidths = calcTableColumnWidths;
// "any", because NodeSpec doesn't support "tableRole" yet
exports.table = {
    content: 'tableRow+',
    attrs: {
        isNumberColumnEnabled: { default: false },
        layout: { default: 'default' },
        __autoSize: { default: false },
    },
    tableRole: 'table',
    isolating: true,
    group: 'block',
    parseDOM: [
        {
            tag: 'table',
            getAttrs: function (dom) { return ({
                isNumberColumnEnabled: dom.getAttribute('data-number-column') === 'true' ? true : false,
                layout: dom.getAttribute('data-layout') || 'default',
                __autoSize: dom.getAttribute('data-autosize') === 'true' ? true : false,
            }); },
        },
    ],
    toDOM: function (node) {
        var attrs = {
            'data-number-column': node.attrs.isNumberColumnEnabled,
            'data-layout': node.attrs.layout,
            'data-autosize': node.attrs.__autoSize,
        };
        return ['table', attrs, ['tbody', 0]];
    },
};
exports.tableToJSON = function (node) { return ({
    attrs: Object.keys(node.attrs)
        .filter(function (key) { return !key.startsWith('__'); })
        .reduce(function (obj, key) {
        obj[key] = node.attrs[key];
        return obj;
    }, {}),
}); };
exports.tableRow = {
    content: '(tableCell | tableHeader)+',
    tableRole: 'row',
    parseDOM: [{ tag: 'tr' }],
    toDOM: function () {
        return ['tr', 0];
    },
};
var cellAttrs = {
    colspan: { default: 1 },
    rowspan: { default: 1 },
    colwidth: { default: null },
    background: { default: null },
};
exports.tableCell = {
    content: '(paragraph | panel | blockquote | orderedList | bulletList | rule | heading | codeBlock |  mediaGroup | mediaSingle | applicationCard | decisionList | taskList | extension)+',
    attrs: cellAttrs,
    tableRole: 'cell',
    isolating: true,
    parseDOM: [
        {
            tag: 'td',
            getAttrs: function (dom) { return getCellAttrs(dom); },
        },
    ],
    toDOM: function (node) {
        return ['td', setCellAttrs(node), 0];
    },
};
exports.toJSONTableCell = function (node) { return ({
    attrs: Object.keys(node.attrs).reduce(function (obj, key) {
        if (cellAttrs[key].default !== node.attrs[key]) {
            obj[key] = node.attrs[key];
        }
        return obj;
    }, {}),
}); };
exports.tableHeader = {
    content: '(paragraph | panel | blockquote | orderedList | bulletList | rule | heading | codeBlock | mediaGroup | mediaSingle  | applicationCard | decisionList | taskList | extension)+',
    attrs: cellAttrs,
    tableRole: 'header_cell',
    isolating: true,
    parseDOM: [
        {
            tag: 'th',
            getAttrs: function (dom) { return getCellAttrs(dom); },
        },
    ],
    toDOM: function (node) {
        return ['th', setCellAttrs(node), 0];
    },
};
exports.toJSONTableHeader = exports.toJSONTableCell;
