// @atlaskit/editor-common/dist/es2015/schema/marks/action.js
import { LINK, COLOR } from '../groups';
export var action = {
    excludes: COLOR,
    group: LINK,
    attrs: {
        key: { default: null },
        title: { default: null },
        target: { default: null },
        parameters: { default: null },
    },
    inclusive: false,
    parseDOM: [
        {
            tag: 'span[data-mark-type="action"]',
            getAttrs: function (dom) {
                var key = dom.getAttribute('data-action-mark-key');
                var targetReceiver = dom.getAttribute('data-action-mark-target-receiver');
                var targetKey = dom.getAttribute('data-action-mark-target-key');
                var title = dom.getAttribute('data-action-mark-title');
                return {
                    key: key,
                    title: title,
                    target: {
                        receiver: targetReceiver,
                        key: targetKey,
                    },
                };
            },
        },
    ],
    toDOM: function (node) {
        var _a = node.attrs, title = _a.title, key = _a.key, target = _a.target;
        return [
            'span',
            {
                'data-mark-type': 'action',
                'data-action-mark-key': key,
                'data-action-mark-title': title,
                'data-action-mark-target-receiver': target && target.receiver,
                'data-action-mark-target-key': target && target.key,
            },
        ];
    },
};
