// @atlaskit/editor-core/dist/es5/plugins/mentions/ui/ToolbarMention/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var react_1 = require("react");
var mention_1 = require("@atlaskit/icon/glyph/editor/mention");
var analytics_1 = require("../../../../analytics");
var ToolbarButton_1 = require("../../../../ui/ToolbarButton");
var ToolbarMention = /** @class */ (function (_super) {
    tslib_1.__extends(ToolbarMention, _super);
    function ToolbarMention() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = { disabled: false };
        _this.handlePluginStateChange = function (pluginState) {
            _this.setState({
                disabled: !pluginState.enabled,
            });
        };
        _this.handleInsertMention = function () {
            _this.pluginState.insertMentionQuery();
            return true;
        };
        return _this;
    }
    ToolbarMention.prototype.componentWillMount = function () {
        this.setPluginState(this.props);
    };
    ToolbarMention.prototype.componentWillUpdate = function (nextProps) {
        if (!this.pluginState) {
            this.setPluginState(nextProps);
        }
    };
    ToolbarMention.prototype.componentWillUnmount = function () {
        var pluginState = this.pluginState;
        if (pluginState) {
            pluginState.unsubscribe(this.handlePluginStateChange);
        }
    };
    ToolbarMention.prototype.render = function () {
        var disabled = this.state.disabled;
        var _a = this.props, isDisabled = _a.isDisabled, isReducedSpacing = _a.isReducedSpacing;
        if (!this.pluginState) {
            return null;
        }
        return (React.createElement(ToolbarButton_1.default, { spacing: isReducedSpacing ? 'none' : 'default', onClick: this.handleInsertMention, disabled: disabled || isDisabled, title: "Mention a person (@)", iconBefore: React.createElement(mention_1.default, { label: "Add mention" }) }));
    };
    ToolbarMention.prototype.setPluginState = function (props) {
        var editorView = props.editorView, pluginKey = props.pluginKey;
        if (!editorView) {
            return;
        }
        var pluginState = pluginKey.getState(editorView.state);
        if (pluginState) {
            this.pluginState = pluginState;
            pluginState.subscribe(this.handlePluginStateChange);
        }
    };
    tslib_1.__decorate([
        analytics_1.analyticsDecorator('atlassian.fabric.mention.picker.trigger.button')
    ], ToolbarMention.prototype, "handleInsertMention", void 0);
    return ToolbarMention;
}(react_1.PureComponent));
exports.default = ToolbarMention;
