// @atlaskit/flag/dist/cjs/components/Flag/styledFlag.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Description = exports.Expander = exports.Content = exports.DismissButton = exports.Title = exports.Icon = exports.Header = undefined;

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  background-color: ', ';\n  border-radius: ', 'px;\n  box-sizing: border-box;\n  box-shadow: ', ';\n  color: ', ';\n  padding: ', 'px;\n  transition: background-color 200ms;\n  width: 100%;\n  z-index: 600;\n\n  &:focus {\n    outline: none;\n    box-shadow: 0 0 0 2px ', ';\n  }\n'], ['\n  background-color: ', ';\n  border-radius: ', 'px;\n  box-sizing: border-box;\n  box-shadow: ', ';\n  color: ', ';\n  padding: ', 'px;\n  transition: background-color 200ms;\n  width: 100%;\n  z-index: 600;\n\n  &:focus {\n    outline: none;\n    box-shadow: 0 0 0 2px ', ';\n  }\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n  align-items: center;\n  height: ', 'px;\n'], ['\n  display: flex;\n  align-items: center;\n  height: ', 'px;\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  flex: 0 0 auto;\n  width: ', 'px;\n'], ['\n  flex: 0 0 auto;\n  width: ', 'px;\n']),
    _templateObject4 = (0, _taggedTemplateLiteral3.default)(['\n  color: ', ';\n  font-weight: 600;\n  flex: 1;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  white-space: nowrap;\n'], ['\n  color: ', ';\n  font-weight: 600;\n  flex: 1;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  white-space: nowrap;\n']),
    _templateObject5 = (0, _taggedTemplateLiteral3.default)(['\n  appearance: none;\n  background: none;\n  border: none;\n  border-radius: ', 'px;\n  color: ', ';\n  cursor: pointer;\n  flex: 0 0 auto;\n  margin-left: ', 'px;\n  padding: 0;\n  white-space: nowrap;\n\n  &:focus {\n    outline: none;\n    box-shadow: 0 0 0 2px ', ';\n  }\n'], ['\n  appearance: none;\n  background: none;\n  border: none;\n  border-radius: ', 'px;\n  color: ', ';\n  cursor: pointer;\n  flex: 0 0 auto;\n  margin-left: ', 'px;\n  padding: 0;\n  white-space: nowrap;\n\n  &:focus {\n    outline: none;\n    box-shadow: 0 0 0 2px ', ';\n  }\n']),
    _templateObject6 = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n  flex: 1 1 100%;\n  flex-direction: column;\n  justify-content: center;\n  min-width: 0;\n  padding: 0 0 0 ', 'px;\n'], ['\n  display: flex;\n  flex: 1 1 100%;\n  flex-direction: column;\n  justify-content: center;\n  min-width: 0;\n  padding: 0 0 0 ', 'px;\n']),
    _templateObject7 = (0, _taggedTemplateLiteral3.default)(['\n  max-height: ', 'px;\n  opacity: ', ';\n  overflow: ', ';\n  transition: max-height 0.3s, opacity 0.3s;\n'], ['\n  max-height: ', 'px;\n  opacity: ', ';\n  overflow: ', ';\n  transition: max-height 0.3s, opacity 0.3s;\n']),
    _templateObject8 = (0, _taggedTemplateLiteral3.default)(['\n  color: ', ';\n  word-wrap: break-word;\n'], ['\n  color: ', ';\n  word-wrap: break-word;\n']);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _theme = require('@atlaskit/theme');

var _theme2 = require('../../theme');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getBoxShadow = function getBoxShadow(props) {
  var borderColor = (0, _theme2.flagBorderColor)(props);
  var shadowColor = (0, _theme2.flagShadowColor)(props);

  var border = borderColor && '0 0 1px ' + borderColor;
  var shadow = '0 20px 32px -8px ' + shadowColor;

  return [border, shadow].filter(function (p) {
    return p;
  }).join(',');
};

exports.default = _styledComponents2.default.div(_templateObject, _theme2.flagBackgroundColor, _theme.borderRadius, getBoxShadow, _theme2.flagTextColor, _theme.math.multiply(_theme.gridSize, 2), _theme2.flagFocusRingColor);

// Header

var Header = exports.Header = _styledComponents2.default.div(_templateObject2, _theme.math.multiply(_theme.gridSize, 4));

var Icon = exports.Icon = _styledComponents2.default.div(_templateObject3, _theme.math.multiply(_theme.gridSize, 5));

var Title = exports.Title = _styledComponents2.default.span(_templateObject4, _theme2.flagTextColor);

var DismissButton = exports.DismissButton = _styledComponents2.default.button(_templateObject5, _theme.borderRadius, _theme2.flagTextColor, _theme.gridSize, _theme2.flagFocusRingColor);

// Content
var Content = exports.Content = _styledComponents2.default.div(_templateObject6, _theme.math.multiply(_theme.gridSize, 5));

var Expander = exports.Expander = _styledComponents2.default.div(_templateObject7, function (_ref) {
  var isExpanded = _ref.isExpanded;
  return isExpanded ? 150 : 0;
}, function (_ref2) {
  var isExpanded = _ref2.isExpanded;
  return isExpanded ? 1 : 0;
}, function (_ref3) {
  var isExpanded = _ref3.isExpanded;
  return isExpanded ? 'visible' : 'hidden';
});

var Description = exports.Description = _styledComponents2.default.div(_templateObject8, _theme2.flagTextColor);
