// @atlaskit/media-card/dist/es5/app/styled/ActionsView.js
"use strict";
/* tslint:disable:variable-name */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var styled_components_1 = require("styled-components");
var util_shared_styles_1 = require("@atlaskit/util-shared-styles");
var media_ui_1 = require("@atlaskit/media-ui");
exports.Actions = styled_components_1.default.div(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n  ", " font-size: 14px;\n"], ["\n  ", " font-size: 14px;\n"])), media_ui_1.center);
exports.ActionsMenu = styled_components_1.default.div(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n  display: flex;\n  margin-left: ", "px;\n"], ["\n  display: flex;\n  margin-left: ", "px;\n"])), util_shared_styles_1.akGridSizeUnitless / 2);
var Message = styled_components_1.default.span(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n  margin-right: ", "px;\n"], ["\n  margin-right: ", "px;\n"])), function (_a) {
    var tryAgain = _a.tryAgain;
    return (tryAgain ? 0 : 5);
});
exports.FailureMessageBlock = styled_components_1.default.div(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject([""], [""])));
exports.FailureMessage = styled_components_1.default(Message)(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n  color: ", ";\n  vertical-align: middle;\n  margin-left: 5px;\n"], ["\n  color: ", ";\n  vertical-align: middle;\n  margin-left: 5px;\n"])), util_shared_styles_1.akColorR400);
exports.SuccessMessage = styled_components_1.default(Message)(templateObject_6 || (templateObject_6 = tslib_1.__makeTemplateObject([""], [""])));
exports.ActionButtonWrapper = styled_components_1.default.div(templateObject_7 || (templateObject_7 = tslib_1.__makeTemplateObject(["\n  margin-left: 5px;\n"], ["\n  margin-left: 5px;\n"])));
exports.ButtonWrapper = styled_components_1.default.div(templateObject_8 || (templateObject_8 = tslib_1.__makeTemplateObject(["\n  ", ";\n"], ["\n  ",
    ";\n"])), function (_a) {
    var isDark = _a.isDark;
    if (isDark) {
        return "\n\n        /* former styles to look good on darker background */\n        & > button {\n\n          color: " + util_shared_styles_1.akColorN0 + " !important;\n          background-color: rgba(255, 255, 255, .08);\n\n          &:hover {\n            background: hsla(0,0%,100%,.12);\n          }\n\n          &:active {\n            background: rgba(76,154,255,.32);\n          }\n\n          &:focus {\n            box-shadow: 0 0 0 2px rgba(38,132,255,.6);\n          }\n        }\n\n      ";
    }
    else {
        return '';
    }
});
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8;
