// @atlaskit/task-decision/dist/es5/type-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDecision = function (item) {
    return !!(item && item.type === 'DECISION');
};
exports.isTask = function (item) {
    return !!(item && item.type === 'TASK');
};
exports.isServiceDecision = function (item) {
    return !!(item && item.type === 'DECISION');
};
exports.isServiceTask = function (item) {
    return !!(item && item.type === 'TASK');
};
exports.isDateSortCriteria = function (sortCriteria) {
    return !sortCriteria ||
        sortCriteria === 'creationDate' ||
        sortCriteria === 'lastUpdateDate';
};
exports.toObjectKey = function (item) {
    var containerAri = item.containerAri, localId = item.localId, objectAri = item.objectAri;
    return {
        containerAri: containerAri,
        localId: localId,
        objectAri: objectAri,
    };
};
exports.toRendererContext = function (item) {
    var containerAri = item.containerAri, objectAri = item.objectAri;
    return {
        containerAri: containerAri,
        objectAri: objectAri,
    };
};
exports.objectKeyToString = function (objectKey) {
    var containerAri = objectKey.containerAri, objectAri = objectKey.objectAri, localId = objectKey.localId;
    return containerAri + ":" + objectAri + ":" + localId;
};
exports.toggleTaskState = function (state) {
    return state === 'DONE' ? 'TODO' : 'DONE';
};
