// @atlaskit/media-card/dist/es5/utils/cardGenericViewSmall/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var react_1 = require("react");
var cx = require("classnames");
var utils_1 = require("../../utils");
var mediaTypeIcon_1 = require("../../utils/mediaTypeIcon");
var cardActions_1 = require("../../utils/cardActions");
var infoView_1 = require("./infoView");
var thumbnailView_1 = require("./thumbnailView");
var styled_1 = require("./styled");
var CardGenericViewSmall = /** @class */ (function (_super) {
    tslib_1.__extends(CardGenericViewSmall, _super);
    function CardGenericViewSmall() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            isMenuExpanded: false,
        };
        return _this;
    }
    CardGenericViewSmall.prototype.render = function () {
        var error = this.props.error;
        if (error) {
            return this.renderError();
        }
        else {
            return this.renderCard();
        }
    };
    CardGenericViewSmall.prototype.renderFileCard = function () {
        var _a = this.props, loading = _a.loading, mediaType = _a.mediaType, thumbnailUrl = _a.thumbnailUrl, title = _a.title, subtitle = _a.subtitle;
        return this.formatCard(React.createElement(thumbnailView_1.ThumbnailView, { type: "file", url: thumbnailUrl, isLoading: loading }), React.createElement(infoView_1.InfoView, { icon: React.createElement(mediaTypeIcon_1.MediaTypeIcon, { type: mediaType }), title: title, subtitle: subtitle, isLink: false, isLoading: loading }));
    };
    CardGenericViewSmall.prototype.renderLinkCard = function () {
        var _a = this.props, loading = _a.loading, iconUrl = _a.iconUrl, thumbnailUrl = _a.thumbnailUrl, title = _a.title, subtitle = _a.subtitle;
        return this.formatCard(React.createElement(thumbnailView_1.ThumbnailView, { type: "link", url: thumbnailUrl, isLoading: loading }), React.createElement(infoView_1.InfoView, { icon: iconUrl ? React.createElement("img", { src: iconUrl }) : null, title: title, subtitle: subtitle, isLink: true, isLoading: loading }));
    };
    CardGenericViewSmall.prototype.renderCard = function () {
        var type = this.props.type;
        if (type === 'link') {
            return this.renderLinkCard();
        }
        else {
            return this.renderFileCard();
        }
    };
    CardGenericViewSmall.prototype.renderError = function () {
        var _a = this.props, error = _a.error, onRetry = _a.onRetry;
        var retryComponent = onRetry ? (React.createElement(styled_1.Retry, { onClick: onRetry }, "Retry")) : null;
        return this.formatCard(React.createElement(utils_1.ErrorIcon, null), React.createElement(styled_1.ErrorWrapper, null,
            React.createElement(styled_1.Error, { className: "error" }, error),
            retryComponent));
    };
    CardGenericViewSmall.prototype.formatCard = function (left, right) {
        var _a = this.props, actions = _a.actions, loading = _a.loading, error = _a.error;
        var cardClass = cx('media-card', { loading: loading });
        return (React.createElement(styled_1.SmallCard, { hasError: !!error, className: cardClass },
            React.createElement(styled_1.ImgWrapper, { shadow: !loading && !error }, left),
            React.createElement(styled_1.InfoWrapper, null, right),
            React.createElement(styled_1.ActionsWrapper, null, actions ? React.createElement(cardActions_1.default, { actions: actions }) : null)));
    };
    return CardGenericViewSmall;
}(react_1.Component));
exports.CardGenericViewSmall = CardGenericViewSmall;
