// @atlaskit/editor-core/dist/es2015/plugins/unsupported-content/index.js
import { confluenceUnsupportedBlock, confluenceUnsupportedInline, } from '@atlaskit/editor-common';
import { Plugin, PluginKey } from 'prosemirror-state';
import { nodeViewFactory } from '../../nodeviews';
import ReactUnsupportedBlockNode from './nodeviews/unsupported-block';
import ReactUnsupportedInlineNode from './nodeviews/unsupported-inline';
import { traverseNode } from './utils';
export var pluginKey = new PluginKey('unsupportedContentPlugin');
var createPlugin = function (schema, providerFactory) {
    return new Plugin({
        state: {
            init: function (config, state) {
                traverseNode(state.doc, schema);
            },
            apply: function (tr, pluginState, oldState, newState) {
                return pluginState;
            },
        },
        key: pluginKey,
        props: {
            nodeViews: {
                confluenceUnsupportedBlock: nodeViewFactory(providerFactory, {
                    confluenceUnsupportedBlock: ReactUnsupportedBlockNode,
                }),
                confluenceUnsupportedInline: nodeViewFactory(providerFactory, {
                    confluenceUnsupportedInline: ReactUnsupportedInlineNode,
                }),
            },
        },
    });
};
var unsupportedContentPlugin = {
    nodes: function () {
        return [
            {
                rank: 1300,
                name: 'confluenceUnsupportedBlock',
                node: confluenceUnsupportedBlock,
            },
            {
                rank: 1310,
                name: 'confluenceUnsupportedInline',
                node: confluenceUnsupportedInline,
            },
        ];
    },
    pmPlugins: function () {
        return [
            {
                rank: 1320,
                plugin: function (_a) {
                    var schema = _a.schema, providerFactory = _a.providerFactory;
                    return createPlugin(schema, providerFactory);
                },
            },
        ];
    },
};
export default unsupportedContentPlugin;
