// @atlaskit/smart-card/dist/es2015/block/CardView/AlertView/index.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import * as debounce from 'lodash.debounce';
import { Ellipsify } from '@atlaskit/media-ui';
import { Wrapper } from './styled';
var contentWidthWhenCardIs400px = 384;
var AlertView = /** @class */ (function (_super) {
    tslib_1.__extends(AlertView, _super);
    function AlertView() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {};
        _this.handleTryAgain = function (event) {
            event.preventDefault();
            event.stopPropagation();
            var onTryAgain = _this.props.onTryAgain;
            if (onTryAgain) {
                onTryAgain();
            }
        };
        _this.handleCancel = function (event) {
            event.preventDefault();
            event.stopPropagation();
            var onCancel = _this.props.onCancel;
            if (onCancel) {
                onCancel();
            }
        };
        _this.handleMount = function (el) {
            _this.el = el;
        };
        _this.handleResize = debounce(function () {
            if (_this.el) {
                _this.setState({ width: _this.el.clientWidth });
            }
        }, 250);
        return _this;
    }
    AlertView.prototype.componentDidMount = function () {
        this.handleResize();
        window.addEventListener('resize', this.handleResize);
    };
    AlertView.prototype.componentWillUnMount = function () {
        window.removeEventListener('resize', this.handleResize);
    };
    AlertView.prototype.renderContent = function () {
        var _a = this.props, type = _a.type, message = _a.message;
        var width = this.state.width;
        var text = type === 'success' ? message : 'Something went wrong.';
        if (width && width < contentWidthWhenCardIs400px) {
            return React.createElement(Ellipsify, { text: text, lines: 2, inline: true });
        }
        else {
            return React.createElement(Ellipsify, { text: text, lines: 1, inline: true });
        }
    };
    AlertView.prototype.renderRetryAndCancel = function () {
        var type = this.props.type;
        if (type === 'success') {
            return null;
        }
        return (React.createElement("span", null,
            React.createElement("a", { onClick: this.handleTryAgain }, "Try again"),
            " or",
            ' ',
            React.createElement("a", { onClick: this.handleCancel }, "cancel"),
            "."));
    };
    AlertView.prototype.render = function () {
        var _a = this.props, type = _a.type, style = _a.style;
        return (React.createElement(Wrapper, { innerRef: this.handleMount, type: type, style: style },
            this.renderContent(),
            " ",
            this.renderRetryAndCancel()));
    };
    return AlertView;
}(React.Component));
export default AlertView;
