// @atlaskit/editor-core/dist/es2015/plugins/collab-edit/utils.js
import { Decoration } from 'prosemirror-view';
import { akColorR400, akColorY400, akColorG400, akColorB400, akColorT400, akColorP400, akColorN800, } from '@atlaskit/util-shared-styles';
import { hexToRgba } from '@atlaskit/editor-common';
export var colors = [
    akColorR400,
    akColorY400,
    akColorG400,
    akColorT400,
    akColorB400,
    akColorP400,
    akColorN800,
].map(function (solid) { return ({
    solid: solid,
    selection: hexToRgba(solid, 0.2),
}); });
// tslint:disable:no-bitwise
export var getAvatarColor = function (str) {
    var hash = 0;
    for (var i = 0; i < str.length; i++) {
        hash = (hash << 5) - hash + str.charCodeAt(i);
        hash = hash & hash;
    }
    var index = Math.abs(hash) % colors.length;
    return { index: index, color: colors[index] };
};
// tslint:enable:no-bitwise
export var findPointers = function (id, decorations) {
    return decorations
        .find()
        .reduce(function (arr, deco) {
        return deco.spec.pointer.sessionId === id ? arr.concat(deco) : arr;
    }, []);
};
function style(options) {
    var color = (options && options.color) || 'black';
    return "border-left: 1px solid " + color + "; border-right: 1px solid " + color + "; margin-right: -2px;";
}
export var createTelepointers = function (from, to, sessionId, isSelection, initial) {
    var decorations = [];
    var avatarColor = getAvatarColor(sessionId);
    var color = avatarColor.index.toString();
    if (isSelection) {
        var className = "telepointer color-" + color + " telepointer-selection";
        decorations.push(Decoration.inline(from, to, { class: className, 'data-initial': initial }, { pointer: { sessionId: sessionId } }));
    }
    var cursor = document.createElement('span');
    cursor.textContent = '\u200b';
    cursor.className = "telepointer color-" + color + " telepointer-selection-badge";
    cursor.style.cssText = style({ color: avatarColor.color.solid }) + ";";
    cursor.setAttribute('data-initial', initial);
    return decorations.concat(Decoration.widget(to, cursor, { pointer: { sessionId: sessionId } }));
};
