// @atlaskit/renderer/dist/es5/email/serializers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var blockquote_1 = require("./nodes/blockquote");
var bullet_list_1 = require("./nodes/bullet-list");
var code_block_1 = require("./nodes/code-block");
var decision_item_1 = require("./nodes/decision-item");
var decision_list_1 = require("./nodes/decision-list");
var emoji_1 = require("./nodes/emoji");
var hard_break_1 = require("./nodes/hard-break");
var heading_1 = require("./nodes/heading");
var list_item_1 = require("./nodes/list-item");
var mention_1 = require("./nodes/mention");
var ordered_list_1 = require("./nodes/ordered-list");
var panel_1 = require("./nodes/panel");
var paragraph_1 = require("./nodes/paragraph");
var rule_1 = require("./nodes/rule");
var table_1 = require("./nodes/table");
var table_cell_1 = require("./nodes/table-cell");
var table_header_1 = require("./nodes/table-header");
var table_row_1 = require("./nodes/table-row");
var text_1 = require("./nodes/text");
var unknown_block_1 = require("./nodes/unknown-block");
var code_1 = require("./marks/code");
var strong_1 = require("./marks/strong");
var link_1 = require("./marks/link");
var strike_1 = require("./marks/strike");
var strong_2 = require("./marks/strong");
var subsup_1 = require("./marks/subsup");
var text_color_1 = require("./marks/text-color");
var underline_1 = require("./marks/underline");
var renderNothing = function () { return ''; };
exports.nodeSerializers = {
    applicationCard: renderNothing,
    bodiedExtension: renderNothing,
    blockquote: blockquote_1.default,
    bulletList: bullet_list_1.default,
    codeBlock: code_block_1.default,
    decisionList: decision_list_1.default,
    decisionItem: decision_item_1.default,
    emoji: emoji_1.default,
    image: renderNothing,
    hardBreak: hard_break_1.default,
    heading: heading_1.default,
    listItem: list_item_1.default,
    mediaGroup: renderNothing,
    mention: mention_1.default,
    orderedList: ordered_list_1.default,
    panel: panel_1.default,
    paragraph: paragraph_1.default,
    rule: rule_1.default,
    table: table_1.default,
    tableCell: table_cell_1.default,
    tableHeader: table_header_1.default,
    tableRow: table_row_1.default,
    taskList: renderNothing,
    text: text_1.default,
    unknownBlock: unknown_block_1.default,
};
exports.markSerializers = {
    code: code_1.default,
    em: strong_1.default,
    link: link_1.default,
    strike: strike_1.default,
    strong: strong_2.default,
    subsup: subsup_1.default,
    textColor: text_color_1.default,
    underline: underline_1.default,
};
