// @atlaskit/editor-core/dist/es2015/plugins/block-type/types.js
// The names of the blocks don't map precisely to schema nodes, because
// of concepts like "paragraph" <-> "Normal text" and "Unknown".
//
// Rather than half-match half-not, this plugin introduces its own
// nomenclature for what 'block type' is active.
export var NORMAL_TEXT = {
    name: 'normal',
    title: 'Normal text',
    nodeName: 'paragraph',
};
export var HEADING_1 = {
    name: 'heading1',
    title: 'Heading 1',
    nodeName: 'heading',
};
export var HEADING_2 = {
    name: 'heading2',
    title: 'Heading 2',
    nodeName: 'heading',
};
export var HEADING_3 = {
    name: 'heading3',
    title: 'Heading 3',
    nodeName: 'heading',
};
export var HEADING_4 = {
    name: 'heading4',
    title: 'Heading 4',
    nodeName: 'heading',
};
export var HEADING_5 = {
    name: 'heading5',
    title: 'Heading 5',
    nodeName: 'heading',
};
export var HEADING_6 = {
    name: 'heading6',
    title: 'Heading 6',
    nodeName: 'heading',
};
export var BLOCK_QUOTE = {
    name: 'blockquote',
    title: 'Block quote',
    nodeName: 'blockquote',
};
export var CODE_BLOCK = {
    name: 'codeblock',
    title: 'Code block',
    nodeName: 'codeBlock',
};
export var PANEL = {
    name: 'panel',
    title: 'Panel',
    nodeName: 'panel',
};
export var OTHER = {
    name: 'other',
    title: 'Other…',
    nodeName: '',
};
export var ALL_BLOCK_TYPES = [
    NORMAL_TEXT,
    HEADING_1,
    HEADING_2,
    HEADING_3,
    HEADING_4,
    HEADING_5,
    HEADING_6,
    BLOCK_QUOTE,
    CODE_BLOCK,
    PANEL,
];
