// @atlaskit/editor-core/dist/es5/plugins/table/pm-plugins/table-column-resizing-plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var prosemirror_state_1 = require("prosemirror-state");
var prosemirror_tables_1 = require("prosemirror-tables");
var main_1 = require("../pm-plugins/main");
exports.pluginKey = new prosemirror_state_1.PluginKey('tableColumnResizingCustomPlugin');
var updateControls = function (state) {
    var tableElement = main_1.stateKey.getState(state).tableElement;
    if (!tableElement) {
        return;
    }
    var cols = tableElement.querySelector('tr').children;
    var columnControls = tableElement.parentElement.querySelectorAll('.table-column');
    var rows = tableElement.querySelectorAll('tr');
    var rowControls = tableElement.parentElement.parentElement.querySelectorAll('.table-row');
    // update column controls width on resize
    for (var i = 0, count = columnControls.length; i < count; i++) {
        columnControls[i].style.width = cols[i].offsetWidth + 1 + "px";
    }
    // update rows controls height on resize
    for (var i = 0, count = rowControls.length; i < count; i++) {
        rowControls[i].style.height = rows[i].offsetHeight + 1 + "px";
    }
    var rightShadow = tableElement.parentElement.parentElement.querySelector('.table-shadow.-right');
    if (rightShadow) {
        var _a = tableElement.parentElement, offsetWidth = _a.offsetWidth, scrollLeft = _a.scrollLeft;
        var diff = tableElement.offsetWidth - offsetWidth;
        var scrollDiff = scrollLeft - diff > 0 ? scrollLeft - diff : 0;
        var width = diff > 0 ? Math.min(diff, 10) : 0;
        var container = tableElement.parentElement.parentElement;
        var paddingLeft = getComputedStyle(container).paddingLeft;
        var paddingLeftPx = paddingLeft
            ? Number(paddingLeft.substr(0, paddingLeft.length - 2))
            : 0;
        rightShadow.style.width = width + "px";
        rightShadow.style.left = offsetWidth -
            width -
            scrollDiff +
            paddingLeftPx + "px";
    }
};
var plugin = new prosemirror_state_1.Plugin({
    key: exports.pluginKey,
    props: {
        handleDOMEvents: {
            mousemove: function (view) {
                var dragging = prosemirror_tables_1.columnResizingPluginKey.getState(view.state).dragging;
                if (dragging) {
                    updateControls(view.state);
                }
                return false;
            },
            mouseleave: function (view) {
                updateControls(view.state);
                return false;
            },
        },
        attributes: function () { return ({ class: 'table-resizing' }); },
    },
});
exports.default = plugin;
