// @atlaskit/reactions/dist/es5/internal/flash-animation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var typestyle_1 = require("typestyle");
var cx = require("classnames");
var react_transition_group_1 = require("react-transition-group");
var theme_1 = require("@atlaskit/theme");
var containerStyle = typestyle_1.style({
    width: '100%',
    height: '100%',
});
var flashTime = 700;
var flashAnimation = typestyle_1.keyframes({
    '0%': {
        backgroundColor: 'transparent',
    },
    '20%': {
        backgroundColor: theme_1.colors.B75,
    },
    '75%': {
        backgroundColor: theme_1.colors.B75,
    },
    '100%': {
        backgroundColor: 'transparent',
    },
});
var flashStyle = typestyle_1.style({
    animation: flashAnimation + " " + flashTime + "ms ease-in-out",
});
/**
 * Flash animation background component. See Reaction component for usage.
 */
var FlashAnimation = /** @class */ (function (_super) {
    tslib_1.__extends(FlashAnimation, _super);
    function FlashAnimation(props) {
        var _this = _super.call(this, props) || this;
        _this.flash = function () {
            _this.setState(function (_a) {
                var flashing = _a.flashing;
                if (flashing) {
                    return null;
                }
                return { flashing: true };
            });
        };
        _this.resetState = function () {
            _this.setState({ flashing: false });
        };
        _this.state = {
            flashing: false,
        };
        return _this;
    }
    FlashAnimation.prototype.render = function () {
        var classNameProp = this.props.className;
        var className = cx(containerStyle, classNameProp);
        return (React.createElement(react_transition_group_1.CSSTransition, { classNames: {
                enter: flashStyle,
            }, in: this.state.flashing, timeout: flashTime, onEntered: this.resetState },
            React.createElement("div", { className: className }, this.props.children)));
    };
    FlashAnimation.defaultProps = {
        className: undefined,
        flash: false,
    };
    return FlashAnimation;
}(React.PureComponent));
exports.default = FlashAnimation;
