// @atlaskit/editor-core/dist/es5/plugins/type-ahead/ui/TypeAheadItemsList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_components_1 = require("styled-components");
var item_1 = require("@atlaskit/item");
var theme_1 = require("@atlaskit/theme");
var itemTheme = (_a = {},
    _a[item_1.itemThemeNamespace] = {
        padding: {
            default: {
                bottom: 8,
                left: 12,
                right: 12,
                top: 8,
            },
        },
        borderRadius: function () { return 0; },
        selected: {
            background: theme_1.themed({ light: theme_1.colors.N20, dark: theme_1.colors.DN70 }),
            text: theme_1.themed({ light: theme_1.colors.N800, dark: theme_1.colors.DN600 }),
            secondaryText: theme_1.themed({ light: theme_1.colors.N200, dark: theme_1.colors.DN300 }),
        },
    },
    _a);
function TypeAheadItemsList(_a) {
    var items = _a.items, currentIndex = _a.currentIndex, insertByIndex = _a.insertByIndex;
    if (!Array.isArray(items)) {
        return null;
    }
    return (React.createElement(styled_components_1.ThemeProvider, { theme: itemTheme },
        React.createElement(item_1.ItemGroup, null, items.map(function (item, index) { return (React.createElement(item_1.default, { key: item.title, onClick: function () { return insertByIndex(index); }, elemBefore: item.icon ? item.icon() : null, isSelected: index === currentIndex }, item.title)); }))));
}
exports.TypeAheadItemsList = TypeAheadItemsList;
var _a;
