// @atlaskit/editor-core/dist/es5/plugins/code-block/pm-plugins/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var prosemirror_commands_1 = require("prosemirror-commands");
var prosemirror_state_1 = require("prosemirror-state");
var prosemirror_utils_1 = require("prosemirror-utils");
var keymaps_1 = require("./keymaps");
var CodeBlockState = /** @class */ (function () {
    function CodeBlockState(state) {
        this.toolbarVisible = false;
        this.domEvent = false;
        this.uniqueId = undefined;
        this.editorFocused = false;
        this.changeHandlers = [];
        this.focusHandlers = [];
        this.changeHandlers = [];
        this.focusHandlers = [];
        this.state = state;
    }
    CodeBlockState.prototype.subscribe = function (cb) {
        this.changeHandlers.push(cb);
        cb(this);
    };
    CodeBlockState.prototype.unsubscribe = function (cb) {
        this.changeHandlers = this.changeHandlers.filter(function (ch) { return ch !== cb; });
    };
    CodeBlockState.prototype.subscribeFocusHandlers = function (cb) {
        this.focusHandlers.push(cb);
    };
    CodeBlockState.prototype.unsubscribeFocusHandlers = function (cb) {
        this.focusHandlers = this.focusHandlers.filter(function (ch) { return ch !== cb; });
    };
    CodeBlockState.prototype.updateLanguage = function (language, view) {
        if (this.activeCodeBlock) {
            prosemirror_commands_1.setBlockType(view.state.schema.nodes.codeBlock, {
                language: language,
                uniqueId: this.uniqueId,
            })(view.state, view.dispatch);
            if (this.focusHandlers.length > 0) {
                this.triggerFocus();
            }
            else {
                view.focus();
            }
        }
    };
    CodeBlockState.prototype.removeCodeBlock = function (view) {
        var _a = view.state, tr = _a.tr, schema = _a.schema, dispatch = view.dispatch;
        dispatch(prosemirror_utils_1.removeParentNodeOfType(schema.nodes.codeBlock)(tr));
        view.focus();
    };
    CodeBlockState.prototype.updateEditorFocused = function (editorFocused) {
        this.editorFocused = editorFocused;
    };
    CodeBlockState.prototype.setLanguages = function (supportedLanguages) {
        this.supportedLanguages = supportedLanguages;
    };
    CodeBlockState.prototype.update = function (state, domAtPos, domEvent) {
        if (domEvent === void 0) { domEvent = false; }
        this.state = state;
        var codeBlockNode = this.activeCodeBlockNode();
        if ((domEvent && codeBlockNode) || codeBlockNode !== this.activeCodeBlock) {
            this.domEvent = domEvent;
            var newElement = codeBlockNode && this.activeCodeBlockElement(domAtPos);
            this.toolbarVisible =
                this.editorFocused &&
                    !!codeBlockNode &&
                    (domEvent || this.element !== newElement);
            this.activeCodeBlock = codeBlockNode;
            this.language =
                (codeBlockNode && codeBlockNode.attrs['language']) || undefined;
            this.element = newElement;
            this.uniqueId = codeBlockNode && codeBlockNode.attrs['uniqueId'];
            this.triggerOnChange();
        }
    };
    CodeBlockState.prototype.triggerOnChange = function () {
        var _this = this;
        this.changeHandlers.forEach(function (cb) { return cb(_this); });
    };
    CodeBlockState.prototype.triggerFocus = function () {
        var _this = this;
        this.focusHandlers.forEach(function (cb) { return cb(_this.uniqueId); });
    };
    CodeBlockState.prototype.activeCodeBlockElement = function (domAtPos) {
        var _a = this.state, selection = _a.selection, codeBlock = _a.schema.nodes.codeBlock;
        return prosemirror_utils_1.findParentDomRefOfType(codeBlock, domAtPos)(selection);
    };
    CodeBlockState.prototype.activeCodeBlockNode = function () {
        var _a = this.state, selection = _a.selection, codeBlock = _a.schema.nodes.codeBlock;
        var parent = prosemirror_utils_1.findParentNodeOfType(codeBlock)(selection);
        if (parent) {
            return parent.node;
        }
    };
    return CodeBlockState;
}());
exports.CodeBlockState = CodeBlockState;
exports.stateKey = new prosemirror_state_1.PluginKey('codeBlockPlugin');
exports.plugin = new prosemirror_state_1.Plugin({
    state: {
        init: function (config, state) {
            return new CodeBlockState(state);
        },
        apply: function (tr, pluginState, oldState, newState) {
            var stored = tr.getMeta(exports.stateKey);
            if (stored) {
                pluginState.update(newState, stored.domAtPos, stored.domEvent);
            }
            return pluginState;
        },
    },
    key: exports.stateKey,
    view: function (editorView) {
        var domAtPos = editorView.domAtPos.bind(editorView);
        exports.stateKey.getState(editorView.state).update(editorView.state, domAtPos);
        return {
            update: function (view, prevState) {
                exports.stateKey.getState(view.state).update(view.state, domAtPos);
            },
        };
    },
    props: {
        handleClick: function (view, event) {
            exports.stateKey
                .getState(view.state)
                .update(view.state, view.domAtPos.bind(view), true);
            return false;
        },
        handleDOMEvents: {
            focus: function (view, event) {
                exports.stateKey.getState(view.state).updateEditorFocused(true);
                return false;
            },
            blur: function (view, event) {
                var pluginState = exports.stateKey.getState(view.state);
                pluginState.updateEditorFocused(false);
                pluginState.update(view.state, view.domAtPos.bind(view), true);
                return false;
            },
        },
    },
});
var plugins = function (schema) {
    return [exports.plugin, keymaps_1.default(schema)].filter(function (plugin) { return !!plugin; });
};
exports.default = plugins;
