// @atlaskit/editor-core/dist/es2015/styles/index.js
import { fontSize } from '@atlaskit/theme';
import { akCodeFontFamily, akColorB100, akColorB400, akColorB50, akColorN0, akColorN20, akColorN30, akColorN40, akColorN100, akColorN500, akColorN700, akColorN900, akZIndexLayer, } from '@atlaskit/util-shared-styles';
export var akEditorCodeFontFamily = akCodeFontFamily;
export var akEditorInactiveForeground = akColorN500;
export var akEditorFocus = akColorB100;
export var akEditorSubtleAccent = akColorN40;
export var akEditorActiveBackground = akColorN500;
export var akEditorActiveForeground = akColorN0;
export var akEditorBlockquoteBorderColor = akColorN40;
export var akEditorDropdownActiveBackground = akColorN900;
export var akEditorPopupBackground = akColorN700;
export var akEditorPopupText = akColorB50;
export var akEditorPrimaryButton = akColorB400;
export var akEditorCodeBackground = akColorN20;
export var akEditorCodeBlockPadding = '12px';
export var akEditorCodeInlinePadding = '2px 4px';
export var akEditorFloatingPanelZIndex = akZIndexLayer;
export var akEditorMentionSelected = akColorN100;
export var akEditorRuleBackground = akColorN30;
export var akEditorRuleBorderRadius = '1px';
export var defaultEditorFontStyles = "\n  font-size: " + fontSize() + "px;\n  font-weight: normal;\n";
export var akEditorToolbarKeylineHeight = 2;
