// @atlaskit/renderer/dist/es5/react/nodes/mediaSingle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var editor_common_1 = require("@atlaskit/editor-common");
var React = require("react");
var react_1 = require("react");
var DEFAULT_WIDTH = 250;
var DEFAULT_HEIGHT = 200;
var MediaSingle = /** @class */ (function (_super) {
    tslib_1.__extends(MediaSingle, _super);
    function MediaSingle(props) {
        var _this = _super.call(this, props) || this;
        _this.onExternalImageLoaded = function (_a) {
            var width = _a.width, height = _a.height;
            _this.setState({
                width: width,
                height: height,
            });
        };
        _this.state = {};
        return _this;
    }
    MediaSingle.prototype.render = function () {
        var props = this.props;
        var child = React.Children.only(React.Children.toArray(props.children)[0]);
        var media = React.cloneElement(child, {
            cardDimensions: {
                width: '100%',
                height: '100%',
            },
            onExternalImageLoaded: this.onExternalImageLoaded,
            disableOverlay: true,
        });
        var _a = child.props, width = _a.width, height = _a.height, type = _a.type;
        if (type === 'external') {
            var _b = this.state, stateWidth = _b.width, stateHeight = _b.height;
            if (width === null) {
                width = stateWidth || DEFAULT_WIDTH;
            }
            if (height === null) {
                height = stateHeight || DEFAULT_HEIGHT;
            }
        }
        return (React.createElement(editor_common_1.MediaSingle, { layout: props.layout, width: width, height: height }, media));
    };
    return MediaSingle;
}(react_1.Component));
exports.default = MediaSingle;
