// @atlaskit/media-picker/dist/es5/popup/components/views/editor/errorView/errorView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var react_1 = require("react");
// import Button from '@atlaskit/button';
var escHelper_1 = require("../escHelper");
var phrases_1 = require("../phrases");
var styles_1 = require("../styles");
var styles_2 = require("./styles");
var icons_1 = require("../../../../../icons");
var ErrorView = /** @class */ (function (_super) {
    tslib_1.__extends(ErrorView, _super);
    function ErrorView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ErrorView.prototype.componentDidMount = function () {
        this.escHelper = new escHelper_1.EscHelper(this.props.onCancel);
    };
    ErrorView.prototype.componentWillUnmount = function () {
        this.escHelper.teardown();
    };
    ErrorView.prototype.render = function () {
        return (React.createElement(styles_1.CenterView, null,
            React.createElement(styles_2.ErrorPopup, null,
                React.createElement(styles_2.ErrorIconWrapper, null, icons_1.errorIcon),
                React.createElement(styles_2.ErrorMessage, null, this.props.message),
                React.createElement(styles_2.ErrorHint, null, this.renderHint()),
                this.renderTryAgainButton(),
                this.renderCancelButton())));
    };
    ErrorView.prototype.renderHint = function () {
        var onRetry = this.props.onRetry;
        if (onRetry) {
            return React.createElement("span", null, phrases_1.errorHintRetry);
        }
        return React.createElement("span", null, phrases_1.errorHintCritical);
    };
    ErrorView.prototype.renderTryAgainButton = function () {
        var onRetry = this.props.onRetry;
        if (onRetry) {
            return (React.createElement(styles_2.ErrorButton, { appearance: "primary", onClick: onRetry }, phrases_1.errorButtonRetry));
        }
        return null;
    };
    ErrorView.prototype.renderCancelButton = function () {
        var _a = this.props, onCancel = _a.onCancel, onRetry = _a.onRetry;
        return (React.createElement(styles_2.ErrorButton, { appearance: "subtle", onClick: onCancel }, onRetry ? phrases_1.errorButtonCancel : phrases_1.errorButtonClose));
    };
    return ErrorView;
}(react_1.Component));
exports.ErrorView = ErrorView;
