// @atlaskit/reactions/dist/es2015/internal/flash-animation.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { style, keyframes } from 'typestyle';
import * as cx from 'classnames';
import { CSSTransition } from 'react-transition-group';
import { colors } from '@atlaskit/theme';
var containerStyle = style({
    width: '100%',
    height: '100%',
});
var flashTime = 700;
var flashAnimation = keyframes({
    '0%': {
        backgroundColor: 'transparent',
    },
    '20%': {
        backgroundColor: colors.B75,
    },
    '75%': {
        backgroundColor: colors.B75,
    },
    '100%': {
        backgroundColor: 'transparent',
    },
});
var flashStyle = style({
    animation: flashAnimation + " " + flashTime + "ms ease-in-out",
});
/**
 * Flash animation background component. See Reaction component for usage.
 */
var FlashAnimation = /** @class */ (function (_super) {
    tslib_1.__extends(FlashAnimation, _super);
    function FlashAnimation(props) {
        var _this = _super.call(this, props) || this;
        _this.flash = function () {
            _this.setState(function (_a) {
                var flashing = _a.flashing;
                if (flashing) {
                    return null;
                }
                return { flashing: true };
            });
        };
        _this.resetState = function () {
            _this.setState({ flashing: false });
        };
        _this.state = {
            flashing: false,
        };
        return _this;
    }
    FlashAnimation.prototype.render = function () {
        var classNameProp = this.props.className;
        var className = cx(containerStyle, classNameProp);
        return (React.createElement(CSSTransition, { classNames: {
                enter: flashStyle,
            }, in: this.state.flashing, timeout: flashTime, onEntered: this.resetState },
            React.createElement("div", { className: className }, this.props.children)));
    };
    FlashAnimation.defaultProps = {
        className: undefined,
        flash: false,
    };
    return FlashAnimation;
}(React.PureComponent));
export default FlashAnimation;
