// @atlaskit/renderer/dist/es5/text/nodes/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var applicationCard_1 = require("./applicationCard");
var blockquote_1 = require("./blockquote");
var bulletList_1 = require("./bulletList");
var hardBreak_1 = require("./hardBreak");
var heading_1 = require("./heading");
var listItem_1 = require("./listItem");
var mediaGroup_1 = require("./mediaGroup");
var orderedList_1 = require("./orderedList");
var panel_1 = require("./panel");
var paragraph_1 = require("./paragraph");
var rule_1 = require("./rule");
var table_1 = require("./table");
var unknown_1 = require("./unknown");
exports.nodeToReducerMapping = {
    applicationCard: applicationCard_1.default,
    blockquote: blockquote_1.default,
    bulletList: bulletList_1.default,
    hardBreak: hardBreak_1.default,
    heading: heading_1.default,
    listItem: listItem_1.default,
    mediaGroup: mediaGroup_1.default,
    orderedList: orderedList_1.default,
    panel: panel_1.default,
    paragraph: paragraph_1.default,
    rule: rule_1.default,
    table: table_1.default,
    unknown: unknown_1.default,
};
