// @atlaskit/media-picker/dist/es5/popup/components/views/editor/editorView/toolbar/toolbar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var react_1 = require("react");
var lineWidthButton_1 = require("./buttons/lineWidthButton");
var colorButton_1 = require("./buttons/colorButton");
var toolButton_1 = require("./buttons/toolButton");
var lineWidthPopup_1 = require("./popups/lineWidthPopup");
var colorPopup_1 = require("./popups/colorPopup");
var styles_1 = require("./styles");
var phrases_1 = require("../../phrases");
exports.tools = [
    'arrow',
    'rectangle',
    'oval',
    'line',
    'text',
    'blur',
    'brush',
];
var Toolbar = /** @class */ (function (_super) {
    tslib_1.__extends(Toolbar, _super);
    function Toolbar(props) {
        var _this = _super.call(this, props) || this;
        _this.state = { popup: 'none' };
        return _this;
    }
    Toolbar.prototype.render = function () {
        var _this = this;
        var color = this.props.color;
        var onColorButtonClick = function () { return _this.showOrHidePopup('color'); };
        var onLineWidthButtonClick = function () { return _this.showOrHidePopup('lineWidth'); };
        var isColorButtonActive = this.state.popup === 'color';
        var isLineWidthButtonActive = this.state.popup === 'lineWidth';
        return (React.createElement(styles_1.ToolbarContainer, null,
            React.createElement(styles_1.CenterButtons, null,
                this.renderToolButtons(),
                React.createElement(lineWidthButton_1.LineWidthButton, { isActive: isLineWidthButtonActive, onClick: onLineWidthButtonClick }),
                React.createElement(colorButton_1.ColorButton, { color: color, isActive: isColorButtonActive, onClick: onColorButtonClick })),
            React.createElement(styles_1.RightButtons, null,
                React.createElement(styles_1.RightButton, { appearance: "primary", theme: "dark", onClick: this.props.onSave }, phrases_1.buttonSave),
                React.createElement(styles_1.RightButton, { appearance: "subtle", onClick: this.props.onCancel, theme: "dark" }, phrases_1.buttonCancel)),
            this.renderPopup()));
    };
    Toolbar.prototype.renderToolButtons = function () {
        var _a = this.props, activeTool = _a.tool, onToolChanged = _a.onToolChanged;
        var onToolClick = function (tool) { return onToolChanged(tool); };
        return exports.tools.map(function (tool) { return (React.createElement(toolButton_1.ToolButton, { key: tool, tool: tool, activeTool: activeTool, onToolClick: onToolClick })); });
    };
    Toolbar.prototype.showOrHidePopup = function (target) {
        if (this.state.popup === target) {
            this.setState({ popup: 'none' });
        }
        else {
            this.setState({ popup: target });
        }
    };
    Toolbar.prototype.renderPopup = function () {
        var _this = this;
        var _a = this.props, color = _a.color, lineWidth = _a.lineWidth, onColorChanged = _a.onColorChanged, onLineWidthChanged = _a.onLineWidthChanged;
        var popup = this.state.popup;
        if (popup === 'color') {
            var onPickColor = function (color) {
                onColorChanged(color);
                _this.setState({ popup: 'none' });
            };
            return React.createElement(colorPopup_1.ColorPopup, { onPickColor: onPickColor, color: color });
        }
        if (popup === 'lineWidth') {
            var onLineWidthClick = function (lineWidth) {
                onLineWidthChanged(lineWidth);
                _this.setState({ popup: 'none' });
            };
            return (React.createElement(lineWidthPopup_1.LineWidthPopup, { onLineWidthClick: onLineWidthClick, lineWidth: lineWidth }));
        }
        return null;
    };
    return Toolbar;
}(react_1.Component));
exports.Toolbar = Toolbar;
