// @atlaskit/media-picker/dist/es5/domain/file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function copyMediaFileForUpload(_a, uploadId) {
    var name = _a.name, size = _a.size, creationDate = _a.creationDate, type = _a.type;
    // We dont' use spread here because user upload events are not sanitized
    return {
        id: uploadId,
        name: name,
        size: size,
        creationDate: creationDate,
        type: type,
    };
}
exports.copyMediaFileForUpload = copyMediaFileForUpload;
function copyPublicMediaFileForUpload(_a, uploadId) {
    var name = _a.name, size = _a.size, creationDate = _a.creationDate, type = _a.type, publicId = _a.publicId;
    // We dont' use spread here because user upload events are not sanitized
    return {
        id: uploadId,
        name: name,
        size: size,
        creationDate: creationDate,
        type: type,
        publicId: publicId,
    };
}
exports.copyPublicMediaFileForUpload = copyPublicMediaFileForUpload;
function isPublicMediaFile(file) {
    return !!file.publicId;
}
exports.isPublicMediaFile = isPublicMediaFile;
