// @atlaskit/editor-core/dist/es5/plugins/width/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var prosemirror_state_1 = require("prosemirror-state");
exports.pluginKey = new prosemirror_state_1.PluginKey('widthPlugin');
function createPlugin(dispatch) {
    return new prosemirror_state_1.Plugin({
        key: exports.pluginKey,
        state: {
            init: function () { return undefined; },
            apply: function (tr, width) {
                var newWidth = tr.getMeta(exports.pluginKey);
                if (newWidth && width !== newWidth) {
                    dispatch(exports.pluginKey, newWidth);
                    return newWidth;
                }
                return width;
            },
        },
    });
}
exports.createPlugin = createPlugin;
var widthPlugin = {
    pmPlugins: function () { return [
        {
            rank: 10000,
            plugin: function (_a) {
                var dispatch = _a.dispatch;
                return createPlugin(dispatch);
            },
        },
    ]; },
};
exports.default = widthPlugin;
