// @atlaskit/comment/dist/cjs/styled/HeaderStyles.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TopItemsContainer = exports.TopItem = exports.RestrictedIconWrapper = exports.Restricted = exports.BulletSpacer = undefined;

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  padding-right: ', 'px;\n'], ['\n  padding-right: ', 'px;\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  color: ', ';\n  display: flex;\n'], ['\n  color: ', ';\n  display: flex;\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  margin-right: ', 'px;\n'], ['\n  margin-right: ', 'px;\n']),
    _templateObject4 = (0, _taggedTemplateLiteral3.default)(['\n  display: inline-block;\n  margin-left: ', 'px;\n\n  [dir=\'rtl\'] & {\n    margin-left: 0;\n    margin-right: ', 'px;\n  }\n\n  &:first-child {\n    margin-left: 0;\n\n    [dir=\'rtl\'] & {\n      margin-right: 0;\n    }\n  }\n'], ['\n  display: inline-block;\n  margin-left: ', 'px;\n\n  [dir=\'rtl\'] & {\n    margin-left: 0;\n    margin-right: ', 'px;\n  }\n\n  &:first-child {\n    margin-left: 0;\n\n    [dir=\'rtl\'] & {\n      margin-right: 0;\n    }\n  }\n']),
    _templateObject5 = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n'], ['\n  display: flex;\n']);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _theme = require('@atlaskit/theme');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ThemeColor = {
  Restricted: {
    text: _theme.colors.N100A
  }
};

var BulletSpacer = exports.BulletSpacer = _styledComponents2.default.span(_templateObject, (0, _theme.gridSize)() / 2);

var Restricted = exports.Restricted = _styledComponents2.default.div(_templateObject2, ThemeColor.Restricted.text);

var RestrictedIconWrapper = exports.RestrictedIconWrapper = _styledComponents2.default.span(_templateObject3, (0, _theme.gridSize)() / 2);
RestrictedIconWrapper.displayName = 'RestrictedIconWrapper';

var TopItem = exports.TopItem = _styledComponents2.default.div(_templateObject4, (0, _theme.gridSize)(), (0, _theme.gridSize)());

var TopItemsContainer = exports.TopItemsContainer = _styledComponents2.default.div(_templateObject5);
