// @atlaskit/editor-core/dist/es2015/plugins/extension/nodeviews/extension.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { ContentNodeView } from '../../../nodeviews';
import Extension from '../ui/Extension';
var ExtensionNode = /** @class */ (function (_super) {
    tslib_1.__extends(ExtensionNode, _super);
    function ExtensionNode(node, view, providerFactory, extensionHandlers) {
        var _this = _super.call(this, node, view) || this;
        var elementType = node.type.name === 'inlineExtension' ? 'span' : 'div';
        _this.node = node;
        _this.view = view;
        _this.providerFactory = providerFactory;
        _this.domRef = document.createElement(elementType);
        // @see ED-3790
        _this.domRef.className = node.type.name + "View-container";
        _this.extensionHandlers = extensionHandlers;
        _this.renderReactComponent(node);
        return _this;
    }
    Object.defineProperty(ExtensionNode.prototype, "dom", {
        get: function () {
            return this.domRef;
        },
        enumerable: true,
        configurable: true
    });
    ExtensionNode.prototype.update = function (node) {
        // @see https://github.com/ProseMirror/prosemirror/issues/648
        var isValidUpdate = this.node.type.name === node.type.name;
        if (isValidUpdate) {
            this.renderReactComponent(node);
        }
        return isValidUpdate;
    };
    ExtensionNode.prototype.destroy = function () {
        ReactDOM.unmountComponentAtNode(this.domRef);
        this.domRef = undefined;
        _super.prototype.destroy.call(this);
    };
    ExtensionNode.prototype.ignoreMutation = function (mutation) {
        // Extensions can perform async operations that will change the DOM.
        // To avoid having their tree rebuilt, we need to ignore the mutation.
        return true;
    };
    ExtensionNode.prototype.renderReactComponent = function (node) {
        ReactDOM.render(React.createElement(Extension, { editorView: this.view, node: node, providerFactory: this.providerFactory, handleContentDOMRef: this.handleRef, extensionHandlers: this.extensionHandlers }), this.domRef);
    };
    return ExtensionNode;
}(ContentNodeView));
export default function ExtensionNodeView(providerFactory, extensionHandlers) {
    return function (node, view, getPos) {
        return new ExtensionNode(node, view, providerFactory, extensionHandlers);
    };
}
