// @atlaskit/smart-card/dist/es5/block/DetailView/Users/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var avatar_1 = require("@atlaskit/avatar");
var styled_1 = require("./styled");
exports.maxAvatarCount = 5;
var Users = /** @class */ (function (_super) {
    tslib_1.__extends(Users, _super);
    function Users() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /* prevent the parent link handler from opening a URL when clicked */
        _this.handleAvatarClick = function (_a) {
            var event = _a.event;
            event.preventDefault();
            event.stopPropagation();
        };
        /* prevent the parent link handler from opening a URL when clicked */
        _this.handleMoreClick = function (event) {
            event.preventDefault();
            event.stopPropagation();
        };
        return _this;
    }
    Users.prototype.render = function () {
        var _a = this.props.users, users = _a === void 0 ? [] : _a;
        if (users.length === 0) {
            return null;
        }
        return (React.createElement(styled_1.Wrapper, null,
            React.createElement(avatar_1.AvatarGroup, { maxCount: exports.maxAvatarCount, appearance: "stack", size: "small", data: users.map(function (user) { return ({
                    name: user.name,
                    src: user.icon,
                    size: 'small',
                }); }), onAvatarClick: this.handleAvatarClick, onMoreClick: this.handleMoreClick })));
    };
    return Users;
}(React.Component));
exports.default = Users;
