// @atlaskit/smart-card/dist/es5/block/DetailView/Widgets/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var badge_1 = require("@atlaskit/badge");
var lozenge_1 = require("@atlaskit/lozenge");
var tooltip_1 = require("@atlaskit/tooltip");
var media_ui_1 = require("@atlaskit/media-ui");
var styled_1 = require("./styled");
var Widgets = /** @class */ (function (_super) {
    tslib_1.__extends(Widgets, _super);
    function Widgets() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Widgets.prototype.renderTitle = function (title) {
        return React.createElement(styled_1.Title, { key: "title" },
            title,
            ":");
    };
    Widgets.prototype.renderIcon = function (icon) {
        return React.createElement(media_ui_1.IconImage, { key: "icon", src: icon });
    };
    Widgets.prototype.renderBadge = function (badge) {
        return (React.createElement(badge_1.default, { key: "badge", appearance: badge.appearance || 'default', value: badge.value, max: badge.max }));
    };
    Widgets.prototype.renderLozenge = function (lozenge) {
        return (React.createElement(lozenge_1.default, { key: "lozenge", appearance: lozenge.appearance || 'default', isBold: lozenge.isBold }, lozenge.text));
    };
    Widgets.prototype.renderText = function (text) {
        return React.createElement(styled_1.Text, { key: "text" }, text);
    };
    Widgets.prototype.renderWidget = function (key, detail) {
        var title = detail.title, text = detail.text, icon = detail.icon, badge = detail.badge, lozenge = detail.lozenge, tooltip = detail.tooltip;
        var attrs = [];
        if (title) {
            attrs.push(this.renderTitle(title));
        }
        if (icon) {
            attrs.push(this.renderIcon(icon));
        }
        if (badge) {
            attrs.push(this.renderBadge(badge));
        }
        if (lozenge) {
            attrs.push(this.renderLozenge(lozenge));
        }
        if (text) {
            attrs.push(this.renderText(text));
        }
        if (attrs.length === 0) {
            // tslint:disable-next-line
            console.warn("Widgets: A widget doesn't contain any supported attributes: " + JSON.stringify(attrs, null, 2));
            return null;
        }
        return (React.createElement(styled_1.WidgetWrapper, { key: key },
            React.createElement(tooltip_1.default, { content: tooltip },
                React.createElement(styled_1.WidgetDetails, null, attrs))));
    };
    Widgets.prototype.render = function () {
        var _this = this;
        var _a = this.props.details, details = _a === void 0 ? [] : _a;
        if (details.length === 0) {
            return null;
        }
        return (React.createElement(styled_1.Wrapper, null, details.map(function (detail, index) { return _this.renderWidget(index, detail); })));
    };
    return Widgets;
}(React.Component));
exports.default = Widgets;
