// @atlaskit/media-picker/dist/es2015/index.js
import { BinaryUploader, } from './components/binary';
import { Browser, } from './components/browser';
import { Clipboard, } from './components/clipboard';
import { Dropzone, } from './components/dropzone';
import { Popup } from './components/popup';
import { UserTracker } from './outer/analytics/tracker';
import { handleError } from './util/handleError';
var trackEvent = new UserTracker().track();
export { BinaryUploader, Browser, Clipboard, Dropzone, Popup };
// TODO: Remove factory ??
export function MediaPicker(componentName, context, pickerConfig) {
    if (context) {
        var analyticsContext = { trackEvent: trackEvent };
        switch (componentName) {
            case 'binary':
                return new BinaryUploader(analyticsContext, context, pickerConfig);
            case 'browser':
                return new Browser(analyticsContext, context, pickerConfig);
            case 'clipboard':
                return new Clipboard(analyticsContext, context, pickerConfig);
            case 'dropzone':
                return new Dropzone(analyticsContext, context, pickerConfig);
            case 'popup':
                return new Popup(analyticsContext, context, pickerConfig);
            default:
                var message = "The component " + componentName + " does not exist";
                handleError('wrong_component', message);
                throw new Error(message);
        }
    }
    else {
        var constructors = {
            binary: BinaryUploader,
            browser: Browser,
            clipboard: Clipboard,
            dropzone: Dropzone,
            popup: Popup,
        };
        return constructors[componentName];
    }
}
