// @atlaskit/smart-card/dist/es5/block/extractPropsFromJSONLD/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var genericExtractPropsFromJSONLD_1 = require("../../genericExtractPropsFromJSONLD");
var extractPropsFromObject_1 = require("./extractPropsFromObject");
var extractPropsFromDocument_1 = require("./extractPropsFromDocument");
var extractPropsFromSpreadsheet_1 = require("./extractPropsFromSpreadsheet");
var blockExtractorPrioritiesByType = {
    Object: 0,
    Document: 5,
    Spreadsheet: 10,
};
var blockExtractorFunctionsByType = {
    Object: extractPropsFromObject_1.extractPropsFromObject,
    Document: extractPropsFromDocument_1.extractPropsFromDocument,
    Spreadsheet: extractPropsFromSpreadsheet_1.extractPropsFromSpreadsheet,
};
function extractPropsFromJSONLD(json) {
    return genericExtractPropsFromJSONLD_1.genericExtractPropsFromJSONLD({
        extractorPrioritiesByType: blockExtractorPrioritiesByType,
        extractorFunctionsByType: blockExtractorFunctionsByType,
        defaultExtractorFunction: extractPropsFromObject_1.extractPropsFromObject,
        json: json,
    });
}
exports.extractPropsFromJSONLD = extractPropsFromJSONLD;
