// @atlaskit/media-picker/dist/es5/popup/services/cloud-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
// We still need postis here to communicate with the "link-account-handler" iframe
var postis = require("postis");
var url = require("url");
var uuid = require("uuid");
var auth_1 = require("../domain/auth");
var objectToQueryString_1 = require("../tools/objectToQueryString");
var CloudService = /** @class */ (function () {
    function CloudService(userAuthProvider) {
        this.userAuthProvider = userAuthProvider;
    }
    CloudService.prototype.startAuth = function (apiUrl, redirectUrl, serviceName) {
        var _this = this;
        var win = window.open('', '_blank');
        return this.userAuthProvider()
            .then(function (auth) {
            return new Promise(function (resolve, reject) {
                var channelId = uuid.v4();
                var authParams = auth_1.mapAuthToQueryParameters(auth);
                var queryString = objectToQueryString_1.objectToQueryString(tslib_1.__assign({}, authParams, { redirectUrl: redirectUrl + "?channelId=" + channelId }));
                var url = _this.pickerUrl(apiUrl) + "/service/" + serviceName + "?" + queryString;
                // Electron does not support location.assign so we must use the
                // string setter to assign a new location to the window
                win.location = url;
                var channel = postis({
                    window: win,
                    scope: channelId,
                });
                channel.ready(function () {
                    channel.listen('auth-callback-received', function () {
                        // notify auth window to close itself
                        channel.send({ method: 'auth-callback-done', params: {} });
                        // unregister the channel listener
                        channel.destroy();
                        resolve();
                        // TODO: MSW-69 what happens if this times out?
                    });
                });
            });
        })
            .catch(function (e) {
            if (win) {
                win.close();
            }
            throw e;
        });
    };
    CloudService.prototype.fileStoreUrl = function (apiUrl) {
        var _a = url.parse(apiUrl), protocol = _a.protocol, host = _a.host;
        return protocol + "//" + host;
    };
    CloudService.prototype.pickerUrl = function (apiUrl) {
        return this.fileStoreUrl(apiUrl) + "/picker";
    };
    return CloudService;
}());
exports.CloudService = CloudService;
