// @atlaskit/media-ui/dist/es5/CardPreview/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var react_render_image_1 = require("react-render-image");
var image_1 = require("@atlaskit/icon/glyph/image");
var NoImageIcon_1 = require("./NoImageIcon");
var styled_1 = require("./styled");
exports.LoadingView = function () { return (React.createElement(styled_1.IconWrapper, null,
    React.createElement(image_1.default, { size: "xlarge", label: "loading" }))); };
exports.NoImageView = function () { return (React.createElement(styled_1.IconWrapper, null,
    React.createElement(NoImageIcon_1.default, null))); };
exports.LoadedView = function (_a) {
    var url = _a.url;
    return (React.createElement(styled_1.ImageWrapper, { url: url }));
};
var CardPreview = /** @class */ (function (_super) {
    tslib_1.__extends(CardPreview, _super);
    function CardPreview() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CardPreview.prototype.renderContent = function () {
        var _a = this.props, isPlaceholder = _a.isPlaceholder, url = _a.url;
        if (isPlaceholder) {
            return React.createElement(exports.LoadingView, null);
        }
        if (!url) {
            return React.createElement(exports.NoImageView, null);
        }
        return (React.createElement(react_render_image_1.default, { src: url, loading: React.createElement(exports.LoadingView, null), loaded: React.createElement(exports.LoadedView, { url: url }), errored: React.createElement(exports.NoImageView, null) }));
    };
    CardPreview.prototype.render = function () {
        return React.createElement(styled_1.Wrapper, null, this.renderContent());
    };
    return CardPreview;
}(React.Component));
exports.CardPreview = CardPreview;
