// @atlaskit/media-card/dist/es2015/app/styled/ActionsView.js
/* tslint:disable:variable-name */
import * as tslib_1 from "tslib";
import styled from 'styled-components';
import { akGridSizeUnitless, akColorR400, akColorN0, } from '@atlaskit/util-shared-styles';
import { center } from '@atlaskit/media-ui';
export var Actions = styled.div(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n  ", " font-size: 14px;\n"], ["\n  ", " font-size: 14px;\n"])), center);
export var ActionsMenu = styled.div(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n  display: flex;\n  margin-left: ", "px;\n"], ["\n  display: flex;\n  margin-left: ", "px;\n"])), akGridSizeUnitless / 2);
var Message = styled.span(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n  margin-right: ", "px;\n"], ["\n  margin-right: ", "px;\n"])), function (_a) {
    var tryAgain = _a.tryAgain;
    return (tryAgain ? 0 : 5);
});
export var FailureMessageBlock = styled.div(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject([""], [""])));
export var FailureMessage = styled(Message)(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n  color: ", ";\n  vertical-align: middle;\n  margin-left: 5px;\n"], ["\n  color: ", ";\n  vertical-align: middle;\n  margin-left: 5px;\n"])), akColorR400);
export var SuccessMessage = styled(Message)(templateObject_6 || (templateObject_6 = tslib_1.__makeTemplateObject([""], [""])));
export var ActionButtonWrapper = styled.div(templateObject_7 || (templateObject_7 = tslib_1.__makeTemplateObject(["\n  margin-left: 5px;\n"], ["\n  margin-left: 5px;\n"])));
export var ButtonWrapper = styled.div(templateObject_8 || (templateObject_8 = tslib_1.__makeTemplateObject(["\n  ", ";\n"], ["\n  ",
    ";\n"])), function (_a) {
    var isDark = _a.isDark;
    if (isDark) {
        return "\n\n        /* former styles to look good on darker background */\n        & > button {\n\n          color: " + akColorN0 + " !important;\n          background-color: rgba(255, 255, 255, .08);\n\n          &:hover {\n            background: hsla(0,0%,100%,.12);\n          }\n\n          &:active {\n            background: rgba(76,154,255,.32);\n          }\n\n          &:focus {\n            box-shadow: 0 0 0 2px rgba(38,132,255,.6);\n          }\n        }\n\n      ";
    }
    else {
        return '';
    }
});
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8;
