// @atlaskit/media-card/dist/es2015/app/styled/AvatarStack.js
/* tslint:disable:variable-name */
import * as tslib_1 from "tslib";
import styled, { css } from 'styled-components';
import { akGridSizeUnitless, akFontFamily, akColorN20, akColorN500, } from '@atlaskit/util-shared-styles';
// copied from @atlaskit/avatar
export var AVATAR_SIZES = {
    xsmall: akGridSizeUnitless * 2,
    small: akGridSizeUnitless * 3,
    medium: akGridSizeUnitless * 4,
    large: akGridSizeUnitless * 6,
    xlarge: akGridSizeUnitless * 12,
};
var borderWidth = 3;
export var StackWrapper = styled.div(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n  display: inline-flex;\n  flex-direction: row-reverse;\n"], ["\n  display: inline-flex;\n  flex-direction: row-reverse;\n"])));
export var ItemWrapper = styled.div(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n  background-color: ", ";\n  border-radius: 100%;\n  display: flex;\n  padding: ", "px;\n  position: relative;\n\n  /* overlap the avatars */\n  & + & {\n    margin-right: -", "px;\n  }\n"], ["\n  background-color: ", ";\n  border-radius: 100%;\n  display: flex;\n  padding: ", "px;\n  position: relative;\n\n  /* overlap the avatars */\n  & + & {\n    margin-right: -", "px;\n  }\n"])), function (_a) {
    var borderColor = _a.borderColor;
    return borderColor;
}, borderWidth, akGridSizeUnitless / 2 + borderWidth);
export var MoreAvatar = styled.div(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n  align-items: center;\n  background-color: ", ";\n  border-radius: 100%;\n  color: ", ";\n  display: flex;\n  font-family: ", ";\n  font-size: 11px;\n  justify-content: center;\n  margin-right: -7px;\n\n  ", ";\n"], ["\n  align-items: center;\n  background-color: ", ";\n  border-radius: 100%;\n  color: ", ";\n  display: flex;\n  font-family: ", ";\n  font-size: 11px;\n  justify-content: center;\n  margin-right: -7px;\n\n  ",
    ";\n"])), akColorN20, akColorN500, akFontFamily, function (_a) {
    var avatarSize = _a.avatarSize;
    return css(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n    height: ", "px;\n    width: ", "px;\n  "], ["\n    height: ", "px;\n    width: ", "px;\n  "])), AVATAR_SIZES[avatarSize], AVATAR_SIZES[avatarSize]);
});
var templateObject_1, templateObject_2, templateObject_4, templateObject_3;
