// @atlaskit/media-picker/dist/es5/popup/components/views/giphy/giphyView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var react_1 = require("react");
var react_redux_1 = require("react-redux");
var debounce = require("lodash.debounce");
var field_text_1 = require("@atlaskit/field-text");
var button_1 = require("@atlaskit/button");
var spinner_1 = require("@atlaskit/spinner");
var media_card_1 = require("@atlaskit/media-card");
var bricksGrid_1 = require("./bricksGrid");
var fileClick_1 = require("../../../actions/fileClick");
var gridCellScaler_1 = require("../../../tools/gridCellScaler");
var searchGiphy_1 = require("../../../actions/searchGiphy");
var styles_1 = require("./styles");
var icons_1 = require("../../../../icons");
var NUMBER_OF_COLUMNS = 4;
var GAP_SIZE = 5;
var CONTAINER_WIDTH = 677;
var GiphyView = /** @class */ (function (_super) {
    tslib_1.__extends(GiphyView, _super);
    function GiphyView(props) {
        var _this = _super.call(this, props) || this;
        _this.getContent = function () {
            var _a = _this.props, hasError = _a.hasError, isLoading = _a.isLoading, cardModels = _a.cardModels;
            if (hasError) {
                return _this.renderError();
            }
            if (!isLoading && cardModels.length === 0) {
                return _this.renderEmptyState();
            }
            return _this.renderSearchResults();
        };
        _this.renderError = function () {
            return (React.createElement(styles_1.WarningContainer, null,
                React.createElement(styles_1.WarningIconWrapper, null, icons_1.errorIcon),
                React.createElement(styles_1.WarningHeading, null, "Ouch! We could not retrieve any GIFs"),
                React.createElement(styles_1.WarningSuggestion, null, "Check your network connection"),
                React.createElement(button_1.default, { onClick: _this.handleRetryButtonClick }, "Try again")));
        };
        _this.renderEmptyState = function () {
            var query = _this.state.query;
            // The GIF used in this error state is too large to store as a data URI (> 3.2 MB)
            return (React.createElement(styles_1.WarningContainer, null,
                React.createElement(styles_1.WarningImage, { src: "https://media1.giphy.com/media/10YK5Hh53nC3dK/200w.gif" }),
                React.createElement(styles_1.WarningHeading, null, "Hello? Was it me you're looking for?"),
                React.createElement(styles_1.WarningSuggestion, null,
                    "We couldn't find anything for \"",
                    query,
                    "\"")));
        };
        _this.renderSearchResults = function () {
            var _a = _this.props, isLoading = _a.isLoading, cardModels = _a.cardModels, totalResultCount = _a.totalResultCount;
            var isThereAreMoreResults = totalResultCount === undefined ||
                cardModels.length < totalResultCount - 1;
            var shouldShowLoadMoreButton = isLoading || isThereAreMoreResults;
            var loadMoreButton = shouldShowLoadMoreButton && _this.renderLoadMoreButton();
            return (React.createElement("div", null,
                _this.renderMasonaryLayout(_this.props.cardModels),
                loadMoreButton));
        };
        _this.renderMasonaryLayout = function (cardModels) {
            if (cardModels.length === 0) {
                return null;
            }
            var cards = cardModels.map(function (cardModel, i) {
                var dataURI = cardModel.dataURI, metadata = cardModel.metadata, actualDimensions = cardModel.dimensions;
                var selectedItems = _this.props.selectedItems;
                var selected = selectedItems.some(function (item) { return item.id === metadata.id && item.serviceName === 'giphy'; });
                var dimensions = gridCellScaler_1.default(tslib_1.__assign({}, actualDimensions, { gapSize: GAP_SIZE, containerWidth: CONTAINER_WIDTH, numberOfColumns: NUMBER_OF_COLUMNS }));
                return (React.createElement(styles_1.GridCell, { key: i + "-metadata.id", width: dimensions.width },
                    React.createElement(media_card_1.CardView, { status: "complete", dataURI: dataURI, metadata: metadata, dimensions: dimensions, selectable: true, selected: selected, onClick: _this.createClickHandler(cardModel) })));
            });
            return (React.createElement(bricksGrid_1.BricksLayout, { id: "mediapicker-gif-layout", sizes: [{ columns: NUMBER_OF_COLUMNS, gutter: GAP_SIZE }] }, cards));
        };
        _this.renderLoadMoreButton = function () {
            var isLoading = _this.props.isLoading;
            var iconAfter = isLoading ? React.createElement(spinner_1.default, null) : undefined;
            return (React.createElement(styles_1.ButtonContainer, null,
                React.createElement(button_1.default, { onClick: _this.handleLoadMoreButtonClick, isDisabled: isLoading, iconAfter: iconAfter }, "Load more GIFs")));
        };
        // private scaleThumbnailGif = ({
        //   width,
        //   height,
        // }: {
        //   width: number;
        //   height: number;
        // }) => {
        //   const desiredWith = Math.floor(
        //     (CONTAINER_WIDTH - GAP_SIZE * (NUMBER_OF_COLUMNS - 1)) /
        //       NUMBER_OF_COLUMNS,
        //   );
        //
        //   return {
        //     width: desiredWith,
        //     height: Math.round(desiredWith / width * height),
        //   };
        // };
        _this.createSearchChangeHandler = function () {
            var onSearchQueryChange = _this.props.onSearchQueryChange;
            var debouncedOnSearchQueryChange = debounce(onSearchQueryChange, 1000);
            return function (e) {
                var query = e.currentTarget.value;
                _this.setState({
                    query: query,
                });
                debouncedOnSearchQueryChange(query);
            };
        };
        _this.createClickHandler = function (cardModel) { return function () {
            _this.props.onCardClick(cardModel);
        }; };
        _this.handleLoadMoreButtonClick = function () {
            var onLoadMoreButtonClick = _this.props.onLoadMoreButtonClick;
            onLoadMoreButtonClick(_this.state.query, true);
        };
        _this.handleRetryButtonClick = function () {
            var onSearchQueryChange = _this.props.onSearchQueryChange;
            onSearchQueryChange(_this.state.query);
        };
        _this.state = {
            query: '',
        };
        _this.searchChangeHandler = _this.createSearchChangeHandler();
        return _this;
    }
    GiphyView.prototype.componentDidUpdate = function (_a) {
        var oldOnSearchQueryChange = _a.onSearchQueryChange;
        var newOnSearchQueryChange = this.props.onSearchQueryChange;
        if (oldOnSearchQueryChange !== newOnSearchQueryChange) {
            this.createSearchChangeHandler();
        }
    };
    GiphyView.prototype.render = function () {
        var query = this.state.query;
        return (React.createElement(styles_1.Container, { id: "mediapicker-giphy-container" },
            React.createElement(styles_1.Title, null, "GIPHY"),
            React.createElement(field_text_1.default, { label: "", placeholder: "Search all the GIFs!", onChange: this.searchChangeHandler, shouldFitContainer: true, value: query }),
            this.getContent()));
    };
    return GiphyView;
}(react_1.Component));
exports.GiphyView = GiphyView;
exports.default = react_redux_1.connect(function (state) { return ({
    hasError: state.view.hasError,
    isLoading: state.view.isLoading,
    cardModels: state.giphy.imageCardModels,
    totalResultCount: state.giphy.totalResultCount,
    selectedItems: state.selectedItems,
}); }, function (dispatch) { return ({
    onSearchQueryChange: function (query) { return dispatch(searchGiphy_1.searchGiphy(query, false)); },
    onLoadMoreButtonClick: function (query, shouldAppendResults) {
        return dispatch(searchGiphy_1.searchGiphy(query, shouldAppendResults));
    },
    onCardClick: function (cardModel) {
        var _a = cardModel.metadata, id = _a.id, name = _a.name, size = _a.size;
        dispatch(fileClick_1.fileClick({
            mimeType: 'image/gif',
            id: id || '',
            name: name || '',
            size: size || 0,
            date: Date.now(),
        }, 'giphy'));
    },
}); })(GiphyView);
