// @atlaskit/task-decision/dist/es2015/api/TaskDecisionLoader.js
/**
 * Grabs the latest Items from the service, deduplicating against existing items.
 */
export var loadLatestItems = function (query, existing, provider, recentUpdateContext) {
    if (recentUpdateContext.localId) {
        // Retry until localId is found
        return retryIteration(function () { return provider.getItems(query).then(function (r) { return r.items; }); }, recentUpdateContext).then(function (items) { return mergeItems(existing, items); });
    }
    // Just load
    return provider
        .getItems(query)
        .then(function (response) { return mergeItems(existing, response.items); });
};
/**
 * Grabs the latest Decisions from the service, deduplicating against existing Decisions.
 */
export var loadLatestDecisions = function (query, existing, provider) {
    return provider
        .getDecisions(query)
        .then(function (response) { return mergeItems(existing, response.decisions); });
};
/**
 * Grabs the latest Tasks from the service, deduplicating against existing Tasks.
 */
export var loadLatestTasks = function (query, existing, provider) {
    return provider
        .getTasks(query)
        .then(function (response) { return mergeItems(existing, response.tasks); });
};
var retryDelaysInMilliseconds = [
    500,
    1000,
    1500,
    2500,
    4000,
    6000,
    8000,
    10000,
];
export var retryIteration = function (loader, recentUpdateContext, retry) {
    if (retry === void 0) { retry = 0; }
    return loadWithDelay(loader, retryDelaysInMilliseconds[retry]).then(function (items) {
        if (items.filter(function (item) { return item.localId === recentUpdateContext.localId; })
            .length > 0) {
            return items;
        }
        var delay = retryDelaysInMilliseconds[retry || 0];
        if (!delay) {
            // Give up - just retry what we've got.
            return items;
        }
        return retryIteration(loader, recentUpdateContext, retry + 1);
    });
};
export var loadWithDelay = function (loader, delay) {
    return new Promise(function (resolve) {
        setTimeout(function () {
            loader().then(function (items) {
                resolve(items);
            });
        }, delay);
    });
};
export var mergeItems = function (existingItems, newItems) {
    var newIds = new Set(newItems.map(function (item) { return item.localId; }));
    var unchangedItems = existingItems.filter(function (item) { return !newIds.has(item.localId); });
    return newItems.concat(unchangedItems);
};
