// @atlaskit/editor-core/dist/es5/plugins/text-formatting/pm-plugins/smart-input-rule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var prosemirror_inputrules_1 = require("prosemirror-inputrules");
var analytics_1 = require("../../../analytics");
var input_rules_1 = require("../../../utils/input-rules");
var prosemirror_state_1 = require("prosemirror-state");
/**
 * Creates an InputRuleHandler that will match on a regular expression of the
 * form `(prefix, content, suffix?)`, and replace it with some given text,
 * maintaining prefix and suffix around the replacement.
 *
 * @param text text to replace with
 */
function replaceTextUsingCaptureGroup(text, trackingEventName) {
    return function (state, match, start, end) {
        var prefix = match[1], suffix = match[3];
        var replacement = (prefix || '') + text + (suffix || '');
        if (trackingEventName) {
            analytics_1.analyticsService.trackEvent("atlassian.editor.format." + trackingEventName + ".autoformatting");
        }
        var tr = state.tr;
        tr.replaceWith(start, end, state.schema.text(replacement));
        tr.setSelection(prosemirror_state_1.Selection.near(tr.doc.resolve(tr.selection.to)));
        return tr;
    };
}
function createReplacementRule(to, from, trackingEventName) {
    return input_rules_1.createInputRule(from, replaceTextUsingCaptureGroup(to, trackingEventName));
}
function createReplacementRules(replMap, trackingEventName) {
    var rules = [];
    for (var _i = 0, _a = Object.keys(replMap); _i < _a.length; _i++) {
        var replacement = _a[_i];
        var regex = replMap[replacement];
        rules.push(createReplacementRule(replacement, regex, trackingEventName));
    }
    return rules;
}
// We don't agressively upgrade single quotes to smart quotes because
// they may clash with an emoji. Only do that when we have a matching
// single quote, or a contraction.
function createSingleQuotesRules(trackingEventName) {
    return [
        // wrapped text
        input_rules_1.createInputRule(/(\s+|^)'(.*)'$/, function (state, match, start, end) {
            var spacing = match[1], innerContent = match[2];
            var replacement = spacing + '‘' + innerContent + '’';
            analytics_1.analyticsService.trackEvent("atlassian.editor.format." + trackingEventName + ".autoformatting");
            return state.tr.insertText(replacement, start, end);
        }),
        // apostrophe
        createReplacementRule('’', /(\w)(')(\w)$/, trackingEventName),
    ];
}
exports.default = prosemirror_inputrules_1.inputRules({
    rules: createReplacementRules({
        Atlassian: /(\s+|^)(atlassian)(\s)$/,
        Jira: /(\s+|^)(jira|JIRA)(\s)$/,
        Bitbucket: /(\s+|^)(bitbucket|BitBucket)(\s)$/,
        Hipchat: /(\s+|^)(hipchat|HipChat)(\s)$/,
        Trello: /(\s+|^)(trello)(\s)$/,
    }, 'product').concat(createReplacementRules({
        '→': /(\s+|^)(--?>)(\s)$/,
        '←': /(\s+|^)(<--?)(\s)$/,
    }, 'arrow'), createReplacementRules({
        '–': /(\s+|^)(--)(\s)$/,
        '…': /()(\.\.\.)$/,
    }, 'typography'), createReplacementRules({
        '“': /((?:^|[\s\{\[\(\<'"\u2018\u201C]))(")$/,
        '”': /"$/,
    }, 'quote'), createSingleQuotesRules('quote')),
});
