// @atlaskit/media-picker/dist/es5/popup/components/navigation/navigation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var react_1 = require("react");
var react_redux_1 = require("react-redux");
var dropdown_menu_1 = require("@atlaskit/dropdown-menu");
var refresh_1 = require("@atlaskit/icon/glyph/refresh");
var settings_1 = require("@atlaskit/icon/glyph/settings");
var actions_1 = require("../../actions");
var changeCloudAccountFolder_1 = require("../../actions/changeCloudAccountFolder");
var changeAccount_1 = require("../../actions/changeAccount");
var styled_1 = require("./styled");
var SERVICENAME = {
    dropbox: 'Dropbox',
    google: 'Google Drive',
};
var Navigation = /** @class */ (function (_super) {
    tslib_1.__extends(Navigation, _super);
    function Navigation(props) {
        var _this = _super.call(this, props) || this;
        _this.onRefreshButtonClick = function () {
            var _a = _this.props, service = _a.service, path = _a.path, onChangePath = _a.onChangePath;
            onChangePath(service.name, service.accountId, path);
        };
        _this.onChangeAccountHandler = function (type, id) { return function () {
            var onChangeAccount = _this.props.onChangeAccount;
            onChangeAccount(type, id);
        }; };
        _this.onUnlinkAccountHandler = function (name, accountId) { return function () {
            var onUnlinkAccount = _this.props.onUnlinkAccount;
            onUnlinkAccount(name, accountId);
        }; };
        _this.onStartAuthHandler = function (name) { return function () {
            var onStartAuth = _this.props.onStartAuth;
            onStartAuth(name);
        }; };
        _this.handleOpenChange = function (attrs) {
            _this.setState({ dropdownOpen: attrs.isOpen });
        };
        _this.state = {
            dropdownOpen: false,
        };
        return _this;
    }
    Navigation.prototype.render = function () {
        var _a = this.props, service = _a.service, path = _a.path;
        var breadcrumbs = this.generateBreadcrumbs(service, path);
        var accountsDropdown = this.getAccountsDropdown();
        return (React.createElement(styled_1.FolderViewerNavigation, null,
            breadcrumbs,
            React.createElement(styled_1.ControlsWrapper, null,
                React.createElement(styled_1.Controls, null,
                    React.createElement(styled_1.ControlButton, { onClick: this.onRefreshButtonClick, iconBefore: React.createElement(refresh_1.default, { label: "refresh" }) }),
                    accountsDropdown))));
    };
    Navigation.prototype.getAccountButton = function () {
        var dropdownOpen = this.state.dropdownOpen;
        return (React.createElement(styled_1.AccountItemButton, { isSelected: dropdownOpen, iconBefore: React.createElement(settings_1.default, { label: "account settings" }) }));
    };
    Navigation.prototype.getAccountsDropdownItems = function () {
        var _this = this;
        var _a = this.props, service = _a.service, accounts = _a.accounts;
        var availableAccounts = accounts.filter(function (account) { return account.type === service.name; });
        var dropdownAccountItems = availableAccounts.map(function (_a) {
            var id = _a.id, displayName = _a.displayName, type = _a.type;
            return (React.createElement(dropdown_menu_1.DropdownItem, { key: id, onClick: _this.onChangeAccountHandler(type, id) }, id === service.accountId ? React.createElement("b", null, displayName) : displayName));
        });
        var dropdownActionItems = [
            React.createElement(dropdown_menu_1.DropdownItem, { key: "add", onClick: this.onStartAuthHandler(service.name) }, "Add account"),
            React.createElement(dropdown_menu_1.DropdownItem, { key: "unlink", onClick: this.onUnlinkAccountHandler(service.name, service.accountId) }, "Unlink Account"),
        ];
        return [
            React.createElement(dropdown_menu_1.DropdownItemGroup, { key: "accounts", title: "Accounts" }, dropdownAccountItems),
            React.createElement(dropdown_menu_1.DropdownItemGroup, { key: "actions", title: "Actions" }, dropdownActionItems),
        ];
    };
    Navigation.prototype.getAccountsDropdown = function () {
        var items = this.getAccountsDropdownItems();
        return (React.createElement(styled_1.AccountDropdownWrapper, null,
            React.createElement(dropdown_menu_1.default, { onOpenChange: this.handleOpenChange, trigger: this.getAccountButton(), position: "bottom right" }, items)));
    };
    Navigation.prototype.generateBreadcrumbs = function (service, path) {
        var _this = this;
        var serviceName = SERVICENAME[service.name] || service.name;
        var fullPath = [{ id: '', name: serviceName }].concat(path);
        var breadcrumbs = fullPath
            .slice(-2)
            .map(function (folderReference) {
            var index = fullPath.indexOf(folderReference);
            return fullPath.slice(0, index + 1);
        })
            .map(function (path, index, allPaths) {
            var isLast = index === allPaths.length - 1;
            return _this.renderBreadcrumb(service, path, isLast);
        });
        return React.createElement(styled_1.BreadCrumbs, null, breadcrumbs);
    };
    Navigation.prototype.renderBreadcrumb = function (service, path, isLast) {
        var onChangePath = this.props.onChangePath;
        if (path.length === 0) {
            return null;
        }
        var folder = path[path.length - 1];
        var onClick = function () {
            return onChangePath(service.name, service.accountId, path.slice(1));
        };
        return (React.createElement(styled_1.BreadCrumbLink, { key: folder.id, onClick: onClick, isLast: isLast },
            React.createElement(styled_1.BreadCrumbLinkLabel, { title: folder.name, isLast: isLast }, folder.name),
            React.createElement(styled_1.BreadCrumbLinkSeparator, { isLast: isLast }, "/")));
    };
    return Navigation;
}(react_1.Component));
exports.Navigation = Navigation;
exports.default = react_redux_1.connect(function (_a) {
    var accounts = _a.accounts, view = _a.view;
    return ({
        accounts: accounts,
        path: view.path,
        service: view.service,
    });
}, function (dispatch) { return ({
    onChangeAccount: function (serviceName, accountId) {
        return dispatch(changeAccount_1.changeAccount(serviceName, accountId));
    },
    onChangePath: function (serviceName, accountId, path) {
        return dispatch(changeCloudAccountFolder_1.changeCloudAccountFolder(serviceName, accountId, path.slice()));
    },
    onStartAuth: function (serviceName) { return dispatch(actions_1.startAuth(serviceName)); },
    onUnlinkAccount: function (serviceName, accountId) {
        return dispatch(actions_1.requestUnlinkCloudAccount({
            id: accountId,
            name: serviceName,
        }));
    },
}); })(Navigation);
