// @atlaskit/media-picker/dist/es5/popup/components/footer/footer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var react_1 = require("react");
var react_redux_1 = require("react-redux");
var styled_1 = require("./styled");
var actions_1 = require("../../actions");
var Footer = /** @class */ (function (_super) {
    tslib_1.__extends(Footer, _super);
    function Footer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Footer.prototype.renderCancelButton = function () {
        var _a = this.props, canCancel = _a.canCancel, onCancel = _a.onCancel;
        return (React.createElement(styled_1.CancelButton, { appearance: "subtle", onClick: onCancel, isDisabled: !canCancel }, "Cancel"));
    };
    Footer.prototype.renderInsertButton = function () {
        var _a = this.props, selectedItems = _a.selectedItems, canInsert = _a.canInsert, onInsert = _a.onInsert;
        var itemCount = selectedItems.length;
        if (itemCount === 0) {
            return null;
        }
        var onClick = function () { return onInsert(selectedItems); };
        return (React.createElement(styled_1.InsertButton, { appearance: "primary", onClick: onClick, isDisabled: !canInsert }, "Insert  " + itemCount + " " + (itemCount === 1 ? 'file' : 'files')));
    };
    Footer.prototype.render = function () {
        return (React.createElement(styled_1.Wrapper, null,
            this.renderInsertButton(),
            this.renderCancelButton()));
    };
    return Footer;
}(react_1.Component));
exports.Footer = Footer;
var mapStateToProps = function (_a) {
    var isUploading = _a.isUploading, isCancelling = _a.isCancelling, selectedItems = _a.selectedItems;
    var isUploadingOrCancelling = isUploading || isCancelling;
    return {
        selectedItems: selectedItems,
        canInsert: !isUploadingOrCancelling,
        canCancel: !isUploadingOrCancelling,
    };
};
var mapDispatchToProps = function (dispatch) { return ({
    onInsert: function (selectedItems) { return dispatch(actions_1.startImport()); },
    onCancel: function () { return dispatch(actions_1.hidePopup()); },
}); };
exports.default = react_redux_1.connect(mapStateToProps, mapDispatchToProps)(Footer);
