// @atlaskit/media-core/dist/es2015/services/linkService.js
import createRequest from './util/createRequest';
var MediaLinkService = /** @class */ (function () {
    function MediaLinkService(config) {
        this.config = config;
    }
    MediaLinkService.prototype.getLinkItem = function (linkId, collectionName) {
        var request = createRequest({
            config: this.config,
            collectionName: collectionName,
            preventPreflight: true,
        });
        return request({ url: "/link/" + linkId })
            .response.then(function (json) { return json.data; })
            .then(function (linkDetails) {
            return {
                type: 'link',
                details: {
                    id: linkDetails.id,
                    url: linkDetails.url,
                    type: linkDetails.metadata.type,
                    title: linkDetails.metadata.title,
                    description: linkDetails.metadata.description,
                    site: linkDetails.metadata.site,
                    author: linkDetails.metadata.author,
                    date: linkDetails.metadata.date,
                    resources: linkDetails.metadata.resources,
                },
            };
        });
    };
    MediaLinkService.prototype.addLinkItem = function (url, collectionName, metadata) {
        var request = createRequest({
            config: this.config,
            collectionName: collectionName,
        });
        return request({
            method: 'post',
            url: '/link',
            data: { url: url, metadata: metadata },
        }).response.then(function (json) { return json.data.id; });
    };
    return MediaLinkService;
}());
export { MediaLinkService };
