// @atlaskit/media-picker/dist/es2015/popup/components/views/browser/auth/auth.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Component } from 'react';
import { connect } from 'react-redux';
import AkButton from '@atlaskit/button';
import DropboxIcon from '@atlaskit/icon/glyph/dropbox';
import GoogledriveIcon from '@atlaskit/icon/glyph/googledrive';
import { startAuth } from '../../../../actions';
import { ButtonWrapper, ConnectWrapper, IconWrapper, TextDescription, Title, } from './styled';
var serviceDetails = {
    dropbox: {
        name: 'Dropbox',
        icon: React.createElement(DropboxIcon, { label: "dropbox", size: "xlarge" }),
    },
    google: {
        name: 'Google Drive',
        icon: React.createElement(GoogledriveIcon, { label: "drive", size: "xlarge" }),
    },
};
/**
 * Routing class that displays view depending on situation.
 */
var StatelessAuth = /** @class */ (function (_super) {
    tslib_1.__extends(StatelessAuth, _super);
    function StatelessAuth() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onClick = function () { return _this.props.onStartAuth(_this.props.service.name); };
        return _this;
    }
    StatelessAuth.prototype.render = function () {
        var service = this.props.service;
        var details = serviceDetails[service.name];
        if (!details) {
            return null;
        }
        var name = details.name, icon = details.icon;
        var btnLabel = "Connect to " + name;
        return (React.createElement(ConnectWrapper, null,
            React.createElement(Title, null,
                "Upload a file from ",
                name),
            React.createElement(IconWrapper, null, icon),
            React.createElement(ButtonWrapper, null,
                React.createElement(AkButton, { appearance: "primary", className: "connectBtn", onClick: this.onClick }, btnLabel)),
            React.createElement(TextDescription, null,
                "We'll open a new page to help you",
                React.createElement("br", null),
                " connect your ",
                name,
                " account")));
    };
    return StatelessAuth;
}(Component));
export { StatelessAuth };
export default connect(function (state) { return ({
    service: state.view.service,
}); }, function (dispatch) { return ({
    onStartAuth: function (serviceName) { return dispatch(startAuth(serviceName)); },
}); })(StatelessAuth);
