// @atlaskit/editor-core/dist/es5/styles/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var theme_1 = require("@atlaskit/theme");
var util_shared_styles_1 = require("@atlaskit/util-shared-styles");
exports.akEditorCodeFontFamily = util_shared_styles_1.akCodeFontFamily;
exports.akEditorInactiveForeground = util_shared_styles_1.akColorN500;
exports.akEditorFocus = util_shared_styles_1.akColorB100;
exports.akEditorSubtleAccent = util_shared_styles_1.akColorN40;
exports.akEditorActiveBackground = util_shared_styles_1.akColorN500;
exports.akEditorActiveForeground = util_shared_styles_1.akColorN0;
exports.akEditorBlockquoteBorderColor = util_shared_styles_1.akColorN40;
exports.akEditorDropdownActiveBackground = util_shared_styles_1.akColorN900;
exports.akEditorPopupBackground = util_shared_styles_1.akColorN700;
exports.akEditorPopupText = util_shared_styles_1.akColorB50;
exports.akEditorPrimaryButton = util_shared_styles_1.akColorB400;
exports.akEditorCodeBackground = util_shared_styles_1.akColorN20;
exports.akEditorCodeBlockPadding = '12px';
exports.akEditorCodeInlinePadding = '2px 4px';
exports.akEditorFloatingPanelZIndex = util_shared_styles_1.akZIndexLayer;
exports.akEditorMentionSelected = util_shared_styles_1.akColorN100;
exports.akEditorRuleBackground = util_shared_styles_1.akColorN30;
exports.akEditorRuleBorderRadius = '1px';
exports.defaultEditorFontStyles = "\n  font-size: " + theme_1.fontSize() + "px;\n  font-weight: normal;\n";
exports.akEditorToolbarKeylineHeight = 2;
