// @atlaskit/editor-core/dist/es5/plugins/unsupported-content/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var editor_common_1 = require("@atlaskit/editor-common");
var prosemirror_state_1 = require("prosemirror-state");
var nodeviews_1 = require("../../nodeviews");
var unsupported_block_1 = require("./nodeviews/unsupported-block");
var unsupported_inline_1 = require("./nodeviews/unsupported-inline");
var utils_1 = require("./utils");
exports.pluginKey = new prosemirror_state_1.PluginKey('unsupportedContentPlugin');
var createPlugin = function (schema, providerFactory) {
    return new prosemirror_state_1.Plugin({
        state: {
            init: function (config, state) {
                utils_1.traverseNode(state.doc, schema);
            },
            apply: function (tr, pluginState, oldState, newState) {
                return pluginState;
            },
        },
        key: exports.pluginKey,
        props: {
            nodeViews: {
                confluenceUnsupportedBlock: nodeviews_1.nodeViewFactory(providerFactory, {
                    confluenceUnsupportedBlock: unsupported_block_1.default,
                }),
                confluenceUnsupportedInline: nodeviews_1.nodeViewFactory(providerFactory, {
                    confluenceUnsupportedInline: unsupported_inline_1.default,
                }),
            },
        },
    });
};
var unsupportedContentPlugin = {
    nodes: function () {
        return [
            {
                rank: 1300,
                name: 'confluenceUnsupportedBlock',
                node: editor_common_1.confluenceUnsupportedBlock,
            },
            {
                rank: 1310,
                name: 'confluenceUnsupportedInline',
                node: editor_common_1.confluenceUnsupportedInline,
            },
        ];
    },
    pmPlugins: function () {
        return [
            {
                rank: 1320,
                plugin: function (_a) {
                    var schema = _a.schema, providerFactory = _a.providerFactory;
                    return createPlugin(schema, providerFactory);
                },
            },
        ];
    },
};
exports.default = unsupportedContentPlugin;
