// @atlaskit/editor-core/dist/es2015/nodeviews/ui/wrapper-click-area.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
import styled from 'styled-components';
import { setNodeSelection } from '../../utils';
var Wrapper = styled.div(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n  width: 100%;\n"], ["\n  width: 100%;\n"])));
Wrapper.displayName = 'WrapperClickArea';
// tslint:disable-next-line:variable-name
export default function wrapComponentWithClickArea(ReactComponent) {
    return /** @class */ (function (_super) {
        tslib_1.__extends(WrapperClickArea, _super);
        function WrapperClickArea() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.state = { selected: false };
            _this.handleDocumentSelectionChange = function (anchorPos, headPos) {
                var getPos = _this.props.getPos;
                var nodePos = getPos();
                _this.setState({
                    selected: nodePos >= anchorPos && nodePos < headPos,
                });
            };
            _this.onClick = function () {
                var _a = _this.props, getPos = _a.getPos, view = _a.view;
                setNodeSelection(view, getPos());
            };
            return _this;
        }
        WrapperClickArea.prototype.componentDidMount = function () {
            var pluginState = this.props.pluginState;
            pluginState.subscribe(this.handleDocumentSelectionChange);
        };
        WrapperClickArea.prototype.componentWillUnmount = function () {
            var pluginState = this.props.pluginState;
            pluginState.unsubscribe(this.handleDocumentSelectionChange);
        };
        WrapperClickArea.prototype.render = function () {
            return (React.createElement(Wrapper, { onClick: this.onClick },
                React.createElement(ReactComponent, tslib_1.__assign({}, this.props, { selected: this.state.selected }))));
        };
        return WrapperClickArea;
    }(PureComponent));
}
var templateObject_1;
