// @atlaskit/media-picker/dist/es2015/popup/components/navigation/navigation.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Component } from 'react';
import { connect } from 'react-redux';
import DropdownMenu, { DropdownItemGroup, DropdownItem, } from '@atlaskit/dropdown-menu';
import RefreshIcon from '@atlaskit/icon/glyph/refresh';
import SettingsIcon from '@atlaskit/icon/glyph/settings';
import { requestUnlinkCloudAccount, startAuth } from '../../actions';
import { changeCloudAccountFolder } from '../../actions/changeCloudAccountFolder';
import { changeAccount } from '../../actions/changeAccount';
import { FolderViewerNavigation, ControlsWrapper, Controls, ControlButton, BreadCrumbs, BreadCrumbLink, BreadCrumbLinkLabel, BreadCrumbLinkSeparator, AccountItemButton, AccountDropdownWrapper, } from './styled';
var SERVICENAME = {
    dropbox: 'Dropbox',
    google: 'Google Drive',
};
var Navigation = /** @class */ (function (_super) {
    tslib_1.__extends(Navigation, _super);
    function Navigation(props) {
        var _this = _super.call(this, props) || this;
        _this.onRefreshButtonClick = function () {
            var _a = _this.props, service = _a.service, path = _a.path, onChangePath = _a.onChangePath;
            onChangePath(service.name, service.accountId, path);
        };
        _this.onChangeAccountHandler = function (type, id) { return function () {
            var onChangeAccount = _this.props.onChangeAccount;
            onChangeAccount(type, id);
        }; };
        _this.onUnlinkAccountHandler = function (name, accountId) { return function () {
            var onUnlinkAccount = _this.props.onUnlinkAccount;
            onUnlinkAccount(name, accountId);
        }; };
        _this.onStartAuthHandler = function (name) { return function () {
            var onStartAuth = _this.props.onStartAuth;
            onStartAuth(name);
        }; };
        _this.handleOpenChange = function (attrs) {
            _this.setState({ dropdownOpen: attrs.isOpen });
        };
        _this.state = {
            dropdownOpen: false,
        };
        return _this;
    }
    Navigation.prototype.render = function () {
        var _a = this.props, service = _a.service, path = _a.path;
        var breadcrumbs = this.generateBreadcrumbs(service, path);
        var accountsDropdown = this.getAccountsDropdown();
        return (React.createElement(FolderViewerNavigation, null,
            breadcrumbs,
            React.createElement(ControlsWrapper, null,
                React.createElement(Controls, null,
                    React.createElement(ControlButton, { onClick: this.onRefreshButtonClick, iconBefore: React.createElement(RefreshIcon, { label: "refresh" }) }),
                    accountsDropdown))));
    };
    Navigation.prototype.getAccountButton = function () {
        var dropdownOpen = this.state.dropdownOpen;
        return (React.createElement(AccountItemButton, { isSelected: dropdownOpen, iconBefore: React.createElement(SettingsIcon, { label: "account settings" }) }));
    };
    Navigation.prototype.getAccountsDropdownItems = function () {
        var _this = this;
        var _a = this.props, service = _a.service, accounts = _a.accounts;
        var availableAccounts = accounts.filter(function (account) { return account.type === service.name; });
        var dropdownAccountItems = availableAccounts.map(function (_a) {
            var id = _a.id, displayName = _a.displayName, type = _a.type;
            return (React.createElement(DropdownItem, { key: id, onClick: _this.onChangeAccountHandler(type, id) }, id === service.accountId ? React.createElement("b", null, displayName) : displayName));
        });
        var dropdownActionItems = [
            React.createElement(DropdownItem, { key: "add", onClick: this.onStartAuthHandler(service.name) }, "Add account"),
            React.createElement(DropdownItem, { key: "unlink", onClick: this.onUnlinkAccountHandler(service.name, service.accountId) }, "Unlink Account"),
        ];
        return [
            React.createElement(DropdownItemGroup, { key: "accounts", title: "Accounts" }, dropdownAccountItems),
            React.createElement(DropdownItemGroup, { key: "actions", title: "Actions" }, dropdownActionItems),
        ];
    };
    Navigation.prototype.getAccountsDropdown = function () {
        var items = this.getAccountsDropdownItems();
        return (React.createElement(AccountDropdownWrapper, null,
            React.createElement(DropdownMenu, { onOpenChange: this.handleOpenChange, trigger: this.getAccountButton(), position: "bottom right" }, items)));
    };
    Navigation.prototype.generateBreadcrumbs = function (service, path) {
        var _this = this;
        var serviceName = SERVICENAME[service.name] || service.name;
        var fullPath = [{ id: '', name: serviceName }].concat(path);
        var breadcrumbs = fullPath
            .slice(-2)
            .map(function (folderReference) {
            var index = fullPath.indexOf(folderReference);
            return fullPath.slice(0, index + 1);
        })
            .map(function (path, index, allPaths) {
            var isLast = index === allPaths.length - 1;
            return _this.renderBreadcrumb(service, path, isLast);
        });
        return React.createElement(BreadCrumbs, null, breadcrumbs);
    };
    Navigation.prototype.renderBreadcrumb = function (service, path, isLast) {
        var onChangePath = this.props.onChangePath;
        if (path.length === 0) {
            return null;
        }
        var folder = path[path.length - 1];
        var onClick = function () {
            return onChangePath(service.name, service.accountId, path.slice(1));
        };
        return (React.createElement(BreadCrumbLink, { key: folder.id, onClick: onClick, isLast: isLast },
            React.createElement(BreadCrumbLinkLabel, { title: folder.name, isLast: isLast }, folder.name),
            React.createElement(BreadCrumbLinkSeparator, { isLast: isLast }, "/")));
    };
    return Navigation;
}(Component));
export { Navigation };
export default connect(function (_a) {
    var accounts = _a.accounts, view = _a.view;
    return ({
        accounts: accounts,
        path: view.path,
        service: view.service,
    });
}, function (dispatch) { return ({
    onChangeAccount: function (serviceName, accountId) {
        return dispatch(changeAccount(serviceName, accountId));
    },
    onChangePath: function (serviceName, accountId, path) {
        return dispatch(changeCloudAccountFolder(serviceName, accountId, path.slice()));
    },
    onStartAuth: function (serviceName) { return dispatch(startAuth(serviceName)); },
    onUnlinkAccount: function (serviceName, accountId) {
        return dispatch(requestUnlinkCloudAccount({
            id: accountId,
            name: serviceName,
        }));
    },
}); })(Navigation);
