// @atlaskit/comment/dist/esm/styled/HeaderStyles.js
import _taggedTemplateLiteral from 'babel-runtime/helpers/taggedTemplateLiteral';

var _templateObject = _taggedTemplateLiteral(['\n  padding-right: ', 'px;\n'], ['\n  padding-right: ', 'px;\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n  color: ', ';\n  display: flex;\n'], ['\n  color: ', ';\n  display: flex;\n']),
    _templateObject3 = _taggedTemplateLiteral(['\n  margin-right: ', 'px;\n'], ['\n  margin-right: ', 'px;\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n  display: inline-block;\n  margin-left: ', 'px;\n\n  [dir=\'rtl\'] & {\n    margin-left: 0;\n    margin-right: ', 'px;\n  }\n\n  &:first-child {\n    margin-left: 0;\n\n    [dir=\'rtl\'] & {\n      margin-right: 0;\n    }\n  }\n'], ['\n  display: inline-block;\n  margin-left: ', 'px;\n\n  [dir=\'rtl\'] & {\n    margin-left: 0;\n    margin-right: ', 'px;\n  }\n\n  &:first-child {\n    margin-left: 0;\n\n    [dir=\'rtl\'] & {\n      margin-right: 0;\n    }\n  }\n']),
    _templateObject5 = _taggedTemplateLiteral(['\n  display: flex;\n'], ['\n  display: flex;\n']);

import styled from 'styled-components';
import { colors, gridSize } from '@atlaskit/theme';

var ThemeColor = {
  Restricted: {
    text: colors.N100A
  }
};

export var BulletSpacer = styled.span(_templateObject, gridSize() / 2);

export var Restricted = styled.div(_templateObject2, ThemeColor.Restricted.text);

export var RestrictedIconWrapper = styled.span(_templateObject3, gridSize() / 2);
RestrictedIconWrapper.displayName = 'RestrictedIconWrapper';

export var TopItem = styled.div(_templateObject4, gridSize(), gridSize());

export var TopItemsContainer = styled.div(_templateObject5);
