// @atlaskit/editor-core/dist/es5/plugins/collab-edit/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var prosemirror_transform_1 = require("prosemirror-transform");
var prosemirror_state_1 = require("prosemirror-state");
exports.handleInit = function (initData, view) {
    var doc = initData.doc, json = initData.json;
    if (doc) {
        var state = view.state, _a = view.state, schema_1 = _a.schema, tr = _a.tr;
        var content = (doc.content || []).map(function (child) {
            return schema_1.nodeFromJSON(child);
        });
        if (content.length) {
            var newState = state.apply(tr
                .setMeta('addToHistory', false)
                .replaceWith(0, state.doc.nodeSize - 2, content)
                .scrollIntoView());
            view.updateState(newState);
        }
    }
    else if (json) {
        exports.applyRemoteSteps(json, view);
    }
};
exports.handleConnection = function (connectionData, view) {
    var tr = view.state.tr;
    view.dispatch(tr.setMeta('sessionId', connectionData));
};
exports.handlePresence = function (presenceData, view) {
    var tr = view.state.tr;
    view.dispatch(tr.setMeta('presence', presenceData));
};
exports.applyRemoteData = function (remoteData, view) {
    var json = remoteData.json, newState = remoteData.newState;
    if (json) {
        exports.applyRemoteSteps(json, view);
    }
    else if (newState) {
        view.updateState(newState);
    }
};
exports.applyRemoteSteps = function (json, view) {
    var state = view.state, schema = view.state.schema;
    var tr = state.tr;
    json.forEach(function (stepJson) {
        var step = prosemirror_transform_1.Step.fromJSON(schema, stepJson);
        tr.step(step);
    });
    tr.setMeta('addToHistory', false);
    tr.scrollIntoView();
    var newState = state.apply(tr);
    view.updateState(newState);
};
exports.handleTelePointer = function (telepointerData, view) {
    var tr = view.state.tr;
    view.dispatch(tr.setMeta('telepointer', telepointerData));
};
function isAllSelection(selection) {
    return selection instanceof prosemirror_state_1.AllSelection;
}
function isNodeSelection(selection) {
    return selection instanceof prosemirror_state_1.NodeSelection;
}
exports.getSendableSelection = function (selection) {
    /**
     * <kbd>CMD + A</kbd> triggers a AllSelection
     * <kbd>escape</kbd> triggers a NodeSelection
     */
    return {
        type: 'textSelection',
        anchor: selection.anchor,
        head: isAllSelection(selection) || isNodeSelection(selection)
            ? selection.head - 1
            : selection.head,
    };
};
