// @atlaskit/editor-common/dist/es2015/schema/nodes/bodied-extension.js
export var bodiedExtension = {
    inline: false,
    group: 'block',
    content: '(paragraph | panel | blockquote | orderedList | bulletList | rule | heading | codeBlock | mediaGroup | mediaSingle | applicationCard | decisionList | taskList | table | extension)+',
    defining: true,
    selectable: true,
    attrs: {
        extensionType: { default: '' },
        extensionKey: { default: '' },
        parameters: { default: null },
        text: { default: null },
    },
    parseDOM: [
        {
            tag: '[data-node-type="bodied-extension"]',
            getAttrs: function (dom) { return ({
                extensionType: dom.getAttribute('data-extension-type'),
                extensionKey: dom.getAttribute('data-extension-key'),
                text: dom.getAttribute('data-text'),
                parameters: JSON.parse(dom.getAttribute('data-parameters') || '{}'),
            }); },
        },
    ],
    toDOM: function (node) {
        var attrs = {
            'data-node-type': 'bodied-extension',
            'data-extension-type': node.attrs.extensionType,
            'data-extension-key': node.attrs.extensionKey,
            'data-text': node.attrs.text,
            'data-parameters': JSON.stringify(node.attrs.parameters),
        };
        return ['div', attrs, 0];
    },
};
