// @atlaskit/editor-core/dist/es2015/plugins/text-color/ui/ToolbarTextColor/index.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { PureComponent } from 'react';
import { PluginKey } from 'prosemirror-state';
import ExpandIcon from '@atlaskit/icon/glyph/chevron-down';
import TextColourIcon from '@atlaskit/icon/glyph/editor/text-color';
import { analyticsDecorator as analytics } from '../../../../analytics';
import ToolbarButton from '../../../../ui/ToolbarButton';
import ColorPalette from '../../../../ui/ColorPalette';
import Dropdown from '../../../../ui/Dropdown';
import { TriggerWrapper, Separator, Wrapper, ExpandIconWrapper, } from './styles';
export var stateKey = new PluginKey('textColorPlugin');
var ToolbarTextColor = /** @class */ (function (_super) {
    tslib_1.__extends(ToolbarTextColor, _super);
    function ToolbarTextColor() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            isOpen: false,
        };
        _this.toggleTextColor = function (color) {
            var _a = _this.props, pluginState = _a.pluginState, editorView = _a.editorView;
            if (!_this.state.disabled) {
                _this.toggleOpen();
                if (color === pluginState.defaultColor) {
                    return pluginState.removeTextColor(editorView.state, editorView.dispatch);
                }
                return pluginState.toggleTextColor(editorView.state, editorView.dispatch, color);
            }
            return false;
        };
        _this.toggleOpen = function () {
            _this.handleOpenChange({ isOpen: !_this.state.isOpen });
        };
        _this.handleOpenChange = function (_a) {
            var isOpen = _a.isOpen;
            _this.setState({ isOpen: isOpen });
        };
        _this.handlePluginStateChange = function (pluginState) {
            var color = pluginState.color, disabled = pluginState.disabled;
            _this.setState({ color: color, disabled: disabled });
        };
        _this.getIconColor = function () {
            var _a = _this.state, isOpen = _a.isOpen, color = _a.color;
            var isDefaultColor = _this.props.pluginState.defaultColor === color;
            return isOpen || isDefaultColor ? undefined : color;
        };
        return _this;
    }
    ToolbarTextColor.prototype.componentDidMount = function () {
        this.props.pluginState.subscribe(this.handlePluginStateChange);
    };
    ToolbarTextColor.prototype.componentWillUnmount = function () {
        this.props.pluginState.unsubscribe(this.handlePluginStateChange);
    };
    ToolbarTextColor.prototype.render = function () {
        var _a = this.state, disabled = _a.disabled, isOpen = _a.isOpen, color = _a.color;
        var _b = this.props, popupsMountPoint = _b.popupsMountPoint, popupsBoundariesElement = _b.popupsBoundariesElement, popupsScrollableElement = _b.popupsScrollableElement, isReducedSpacing = _b.isReducedSpacing, _c = _b.pluginState, palette = _c.palette, borderColorPalette = _c.borderColorPalette;
        return (React.createElement(Wrapper, null,
            React.createElement(Dropdown, { mountTo: popupsMountPoint, boundariesElement: popupsBoundariesElement, scrollableElement: popupsScrollableElement, isOpen: isOpen && !disabled && !this.props.disabled, onOpenChange: this.handleOpenChange, fitWidth: 242, fitHeight: 80, trigger: React.createElement(ToolbarButton, { spacing: isReducedSpacing ? 'none' : 'default', disabled: disabled || this.props.disabled, selected: isOpen, title: "Text color", onClick: this.toggleOpen, iconBefore: React.createElement(TriggerWrapper, null,
                        React.createElement(TextColourIcon, { primaryColor: this.getIconColor(), label: "Text color" }),
                        React.createElement(ExpandIconWrapper, null,
                            React.createElement(ExpandIcon, { label: "expand-dropdown-menu" }))) }) },
                React.createElement(ColorPalette, { palette: palette, onClick: this.toggleTextColor, selectedColor: color, borderColors: borderColorPalette })),
            React.createElement(Separator, null)));
    };
    tslib_1.__decorate([
        analytics('atlassian.editor.format.textcolor.button')
    ], ToolbarTextColor.prototype, "toggleTextColor", void 0);
    return ToolbarTextColor;
}(PureComponent));
export default ToolbarTextColor;
