// @atlaskit/media-picker/dist/es5/popup/reducers/fileUploadsAdd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var fileUploadsStart_1 = require("../actions/fileUploadsStart");
function fileUploadsAdd(state, action) {
    if (fileUploadsStart_1.isFileUploadsStartAction(action)) {
        var tenant_1 = state.tenant, uploads = state.uploads, selectedItems = state.selectedItems, lastUploadIndex = state.lastUploadIndex;
        var files = action.files;
        var newUploads_1 = {};
        var newLastUploadIndex_1 = lastUploadIndex;
        files.forEach(function (file) {
            return (newUploads_1[file.id] = {
                file: {
                    metadata: {
                        id: file.id,
                        name: file.name,
                        mimeType: file.type,
                        size: file.size,
                    },
                    dataURI: '',
                },
                progress: 0,
                events: [],
                index: newLastUploadIndex_1++,
                tenant: tenant_1,
            });
        });
        var newSelectedItems = files.map(function (file) { return ({
            date: 0,
            id: file.id,
            mimeType: file.type,
            name: file.name,
            parentId: '',
            size: file.size,
            serviceName: 'upload',
        }); });
        return tslib_1.__assign({}, state, { uploads: tslib_1.__assign({}, uploads, newUploads_1), selectedItems: selectedItems.concat(newSelectedItems), lastUploadIndex: newLastUploadIndex_1 });
    }
    else {
        return state;
    }
}
exports.default = fileUploadsAdd;
