// @atlaskit/renderer/dist/es5/email/nodes/panel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util_shared_styles_1 = require("@atlaskit/util-shared-styles");
var util_1 = require("../util");
var config = {
    info: {
        background: util_shared_styles_1.akColorB50,
        iconColor: util_shared_styles_1.akColorB400,
    },
    note: {
        background: util_shared_styles_1.akColorP50,
        iconColor: util_shared_styles_1.akColorP400,
    },
    tip: {
        background: util_shared_styles_1.akColorG50,
        iconColor: util_shared_styles_1.akColorG400,
    },
    success: {
        background: util_shared_styles_1.akColorG50,
        iconColor: util_shared_styles_1.akColorG400,
    },
    warning: {
        background: util_shared_styles_1.akColorY50,
        iconColor: util_shared_styles_1.akColorY400,
    },
    error: {
        background: util_shared_styles_1.akColorR50,
        iconColor: util_shared_styles_1.akColorR400,
    },
};
function panel(_a) {
    var attrs = _a.attrs, text = _a.text;
    var css = util_1.serializeStyle({
        'border-radius': util_shared_styles_1.akBorderRadius,
        margin: util_shared_styles_1.akGridSizeUnitless / 2 + "px 0",
        padding: util_shared_styles_1.akGridSizeUnitless + "px",
        background: config[attrs.panelType] && config[attrs.panelType].background,
    });
    return util_1.createTag('div', { style: css }, text);
}
exports.default = panel;
