// @atlaskit/renderer/dist/es5/email/nodes/table-header.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var util_shared_styles_1 = require("@atlaskit/util-shared-styles");
var util_1 = require("../util");
var baseStyle = {
    'background-color': util_shared_styles_1.akColorN20,
    'background-clip': 'padding-box',
    border: "1px solid " + util_shared_styles_1.akColorN50,
    'border-right-width': 0,
    'border-bottom-width': 0,
    'font-weight': 'bold',
    height: '2.5em',
    'min-width': '3em',
    padding: '6px 10px',
    'text-align': 'left',
    'vertical-align': 'top',
};
function tableHeader(_a) {
    var attrs = _a.attrs, text = _a.text;
    var colspan = attrs.colspan, rowspan = attrs.rowspan, background = attrs.background;
    var style = util_1.serializeStyle(tslib_1.__assign({}, baseStyle, { 'background-color': background ? background : baseStyle['background-color'] }));
    return util_1.createTag('th', {
        colspan: colspan,
        rowspan: rowspan,
        style: style,
    }, text);
}
exports.default = tableHeader;
