// @atlaskit/task-decision/dist/es5/api/TaskDecisionUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var type_helpers_1 = require("../type-helpers");
exports.convertServiceDecisionToDecision = function (serviceDecision) {
    var creationDate = serviceDecision.creationDate, lastUpdateDate = serviceDecision.lastUpdateDate, rawContent = serviceDecision.rawContent, contentAsFabricDocument = serviceDecision.contentAsFabricDocument, other = tslib_1.__rest(serviceDecision, ["creationDate", "lastUpdateDate", "rawContent", "contentAsFabricDocument"]);
    return tslib_1.__assign({ creationDate: new Date(creationDate), lastUpdateDate: new Date(lastUpdateDate), content: JSON.parse(contentAsFabricDocument) }, other);
};
exports.convertServiceDecisionResponseToDecisionResponse = function (serviceDecisionResponse, query) {
    var decisions = serviceDecisionResponse.decisions.map(exports.convertServiceDecisionToDecision);
    var nextQuery;
    if (query &&
        serviceDecisionResponse.meta &&
        serviceDecisionResponse.meta.cursor) {
        nextQuery = tslib_1.__assign({}, query, { cursor: serviceDecisionResponse.meta.cursor });
    }
    return {
        decisions: decisions,
        nextQuery: nextQuery,
    };
};
exports.convertServiceTaskToTask = function (serviceTask) {
    var creationDate = serviceTask.creationDate, lastUpdateDate = serviceTask.lastUpdateDate, rawContent = serviceTask.rawContent, contentAsFabricDocument = serviceTask.contentAsFabricDocument, other = tslib_1.__rest(serviceTask, ["creationDate", "lastUpdateDate", "rawContent", "contentAsFabricDocument"]);
    return tslib_1.__assign({ creationDate: new Date(creationDate), lastUpdateDate: new Date(lastUpdateDate), content: JSON.parse(contentAsFabricDocument) }, other);
};
exports.convertServiceTaskStateToBaseItem = function (serviceTaskInfo) {
    var lastUpdateDate = serviceTaskInfo.lastUpdateDate, other = tslib_1.__rest(serviceTaskInfo, ["lastUpdateDate"]);
    return tslib_1.__assign({ type: 'TASK', lastUpdateDate: new Date(lastUpdateDate) }, other);
};
exports.convertServiceTaskResponseToTaskResponse = function (serviceResponse, query) {
    var tasks = serviceResponse.tasks.map(exports.convertServiceTaskToTask);
    var nextQuery;
    if (query && serviceResponse.meta && serviceResponse.meta.cursor) {
        nextQuery = tslib_1.__assign({}, query, { cursor: serviceResponse.meta.cursor });
    }
    return {
        tasks: tasks,
        nextQuery: nextQuery,
    };
};
exports.convertServiceItemToItem = function (items, serviceItem) {
    if (type_helpers_1.isServiceDecision(serviceItem)) {
        items.push(exports.convertServiceDecisionToDecision(serviceItem));
    }
    else if (type_helpers_1.isServiceTask(serviceItem)) {
        items.push(exports.convertServiceTaskToTask(serviceItem));
    }
    return items;
};
exports.convertServiceItemResponseToItemResponse = function (serviceResponse, query) {
    var items = serviceResponse.elements.reduce(exports.convertServiceItemToItem, []);
    var nextQuery;
    if (query && serviceResponse.meta && serviceResponse.meta.cursor) {
        nextQuery = tslib_1.__assign({}, query, { cursor: serviceResponse.meta.cursor });
    }
    return {
        items: items,
        nextQuery: nextQuery,
    };
};
exports.decisionsToDocument = function (decisions) { return ({
    type: 'doc',
    version: 1,
    content: [
        {
            type: 'decisionList',
            content: decisions.map(function (decision) {
                var content = decision.content, localId = decision.localId, state = decision.state;
                return {
                    type: 'decisionItem',
                    attrs: {
                        localId: localId,
                        state: state,
                    },
                    content: content,
                };
            }),
        },
    ],
}); };
exports.tasksToDocument = function (tasks) { return ({
    type: 'doc',
    version: 1,
    content: [
        {
            type: 'taskList',
            content: tasks.map(function (task) {
                var content = task.content, localId = task.localId, state = task.state;
                return {
                    type: 'taskItem',
                    attrs: {
                        localId: localId,
                        state: state,
                    },
                    content: content,
                };
            }),
        },
    ],
}); };
exports.findIndex = function (array, predicate) {
    var index = -1;
    array.some(function (item, i) {
        if (predicate(item)) {
            index = i;
            return true;
        }
        return false;
    });
    return index;
};
