// @atlaskit/editor-common/dist/es5/schema/nodes/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extension = {
    inline: false,
    group: 'block',
    atom: true,
    selectable: true,
    attrs: {
        extensionType: { default: '' },
        extensionKey: { default: '' },
        parameters: { default: null },
        text: { default: null },
    },
    parseDOM: [
        {
            tag: '[data-node-type="extension"]',
            getAttrs: function (dom) { return ({
                extensionType: dom.getAttribute('data-extension-type'),
                extensionKey: dom.getAttribute('data-extension-key'),
                text: dom.getAttribute('data-text'),
                parameters: JSON.parse(dom.getAttribute('data-parameters') || '{}'),
            }); },
        },
    ],
    toDOM: function (node) {
        var attrs = {
            'data-node-type': 'extension',
            'data-extension-type': node.attrs.extensionType,
            'data-extension-key': node.attrs.extensionKey,
            'data-text': node.attrs.text,
            'data-parameters': JSON.stringify(node.attrs.parameters),
        };
        return ['div', attrs];
    },
};
