// @atlaskit/editor-core/dist/es5/plugins/hyperlink/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var editor_common_1 = require("@atlaskit/editor-common");
var main_1 = require("./pm-plugins/main");
exports.hyperlinkPluginKey = main_1.hyperlinkPluginKey;
var input_rule_1 = require("./pm-plugins/input-rule");
var keymap_1 = require("./pm-plugins/keymap");
var HyperlinkEdit_1 = require("./ui/HyperlinkEdit");
var hyperlinkPlugin = {
    marks: function () {
        return [{ name: 'link', mark: editor_common_1.link, rank: 100 }];
    },
    pmPlugins: function () {
        return [
            { rank: 900, plugin: function (_a) {
                    var schema = _a.schema, props = _a.props;
                    return main_1.createPlugin(schema, props);
                } },
            { rank: 910, plugin: function (_a) {
                    var schema = _a.schema;
                    return input_rule_1.createInputRulePlugin(schema);
                } },
            {
                rank: 920,
                plugin: function (_a) {
                    var schema = _a.schema, props = _a.props;
                    return keymap_1.createKeymapPlugin(schema, props);
                },
            },
        ];
    },
    contentComponent: function (_a) {
        var editorView = _a.editorView, providerFactory = _a.providerFactory, appearance = _a.appearance, popupsMountPoint = _a.popupsMountPoint, popupsBoundariesElement = _a.popupsBoundariesElement, popupsScrollableElement = _a.popupsScrollableElement;
        if (appearance === 'message') {
            return null;
        }
        var renderNode = function (providers) {
            var pluginState = main_1.hyperlinkPluginKey.getState(editorView.state);
            return (React.createElement(HyperlinkEdit_1.default, { editorView: editorView, pluginState: pluginState, activityProvider: providers.activityProvider, popupsMountPoint: popupsMountPoint, popupsBoundariesElement: popupsBoundariesElement, popupsScrollableElement: popupsScrollableElement }));
        };
        return (React.createElement(editor_common_1.WithProviders, { providerFactory: providerFactory, providers: ['activityProvider'], renderNode: renderNode }));
    },
};
exports.default = hyperlinkPlugin;
