// @atlaskit/editor-core/dist/es5/plugins/insert-block/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var editor_common_1 = require("@atlaskit/editor-common");
var main_1 = require("../block-type/pm-plugins/main");
var main_2 = require("../media/pm-plugins/main");
var hyperlink_1 = require("../hyperlink");
var main_3 = require("../mentions/pm-plugins/main");
var main_4 = require("../table/pm-plugins/main");
var main_5 = require("../image-upload/pm-plugins/main");
var placeholder_text_1 = require("../placeholder-text");
var layout_1 = require("../layout");
var macro_1 = require("../macro");
var plugin_1 = require("../date/plugin");
var main_6 = require("../emoji/pm-plugins/main");
var WithPluginState_1 = require("../../ui/WithPluginState");
var Toolbar_1 = require("../../ui/Toolbar");
var ToolbarInsertBlock_1 = require("./ui/ToolbarInsertBlock");
var toolbarSizeToButtons = function (toolbarSize) {
    switch (toolbarSize) {
        case Toolbar_1.ToolbarSize.XXL:
        case Toolbar_1.ToolbarSize.XL:
        case Toolbar_1.ToolbarSize.L:
        case Toolbar_1.ToolbarSize.M:
            return 5;
        case Toolbar_1.ToolbarSize.S:
            return 2;
        default:
            return 0;
    }
};
var insertBlockPlugin = function (options) { return ({
    primaryToolbarComponent: function (_a) {
        var editorView = _a.editorView, editorActions = _a.editorActions, eventDispatcher = _a.eventDispatcher, providerFactory = _a.providerFactory, popupsMountPoint = _a.popupsMountPoint, popupsBoundariesElement = _a.popupsBoundariesElement, popupsScrollableElement = _a.popupsScrollableElement, toolbarSize = _a.toolbarSize, disabled = _a.disabled, isToolbarReducedSpacing = _a.isToolbarReducedSpacing;
        var buttons = toolbarSizeToButtons(toolbarSize);
        var renderNode = function (providers) {
            return (React.createElement(WithPluginState_1.default, { plugins: {
                    blockTypeState: main_1.stateKey,
                    mediaState: main_2.stateKey,
                    tablesState: main_4.stateKey,
                    mentionsState: main_3.mentionPluginKey,
                    macroState: macro_1.pluginKey,
                    hyperlinkState: hyperlink_1.hyperlinkPluginKey,
                    emojiState: main_6.emojiPluginKey,
                    dateState: plugin_1.pluginKey,
                    imageUpload: main_5.stateKey,
                    placeholderTextState: placeholder_text_1.pluginKey,
                    layoutState: layout_1.pluginKey,
                }, render: function (_a) {
                    var _b = _a.blockTypeState, blockTypeState = _b === void 0 ? {} : _b, mediaState = _a.mediaState, mentionsState = _a.mentionsState, tablesState = _a.tablesState, _c = _a.macroState, macroState = _c === void 0 ? {} : _c, hyperlinkState = _a.hyperlinkState, emojiState = _a.emojiState, dateState = _a.dateState, imageUpload = _a.imageUpload, placeholderTextState = _a.placeholderTextState, layoutState = _a.layoutState;
                    return (React.createElement(ToolbarInsertBlock_1.default, { buttons: buttons, isReducedSpacing: isToolbarReducedSpacing, isDisabled: disabled, editorView: editorView, tableHidden: tablesState && tablesState.tableHidden, tableSupported: !!tablesState, mentionsEnabled: mentionsState && mentionsState.enabled, dateEnabled: !!dateState, placeholderTextEnabled: placeholderTextState && placeholderTextState.allowInserting, layoutSectionEnabled: !!layoutState, insertMentionQuery: mentionsState && mentionsState.insertMentionQuery, mentionsSupported: !!mentionsState, mediaUploadsEnabled: mediaState && mediaState.allowsUploads, onShowMediaPicker: mediaState && mediaState.showMediaPicker, mediaSupported: !!mediaState, imageUploadSupported: !!imageUpload, imageUploadEnabled: imageUpload && imageUpload.enabled, handleImageUpload: imageUpload && imageUpload.handleImageUpload.bind(imageUpload), availableWrapperBlockTypes: blockTypeState.availableWrapperBlockTypes, linkSupported: !!hyperlinkState, linkDisabled: !hyperlinkState ||
                            !hyperlinkState.linkable ||
                            hyperlinkState.active, showLinkPanel: hyperlinkState && hyperlinkState.showLinkPanel, emojiDisabled: !emojiState || !emojiState.enabled, insertEmoji: emojiState && emojiState.insertEmoji, emojiProvider: providers.emojiProvider, horizontalRuleEnabled: options.horizontalRuleEnabled, onInsertBlockType: blockTypeState.insertBlockType, onInsertMacroFromMacroBrowser: macro_1.insertMacroFromMacroBrowser, macroProvider: macroState.macroProvider, popupsMountPoint: popupsMountPoint, popupsBoundariesElement: popupsBoundariesElement, popupsScrollableElement: popupsScrollableElement, insertMenuItems: options.insertMenuItems, editorActions: editorActions }));
                } }));
        };
        return (React.createElement(editor_common_1.WithProviders, { providerFactory: providerFactory, providers: ['emojiProvider'], renderNode: renderNode }));
    },
}); };
exports.default = insertBlockPlugin;
