// @atlaskit/editor-core/dist/es2015/plugins/table/utils/selection.js
import { CellSelection } from 'prosemirror-tables';
import { TableMap } from 'prosemirror-tables';
import { isColumnSelected } from 'prosemirror-utils';
export var getCellSelection = function (state) {
    var selection = state.selection;
    if (selection instanceof CellSelection) {
        return selection;
    }
};
export var isHeaderRowSelected = function (state) {
    var cellSelection = getCellSelection(state);
    if (cellSelection && cellSelection.isRowSelection()) {
        var $from = cellSelection.$from;
        var tableHeader = state.schema.nodes.tableHeader;
        for (var i = $from.depth; i > 0; i--) {
            if ($from.node(i).type === tableHeader) {
                return true;
            }
        }
    }
    return false;
};
export var getSelectedColumn = function (state) {
    var cellSelection = getCellSelection(state);
    var _a = cellSelection, $anchorCell = _a.$anchorCell, $headCell = _a.$headCell;
    var tableNode = $anchorCell.node(-1);
    var map = TableMap.get(tableNode);
    var start = $anchorCell.start(-1);
    var anchor = map.colCount($anchorCell.pos - start);
    var head = map.colCount($headCell.pos - start);
    return { anchor: anchor, head: head };
};
export var getSelectedRow = function (state) {
    var cellSelection = getCellSelection(state);
    var _a = cellSelection, $anchorCell = _a.$anchorCell, $headCell = _a.$headCell;
    var anchor = $anchorCell.index(-1);
    var head = $headCell.index(-1);
    return { anchor: anchor, head: head };
};
export var checkIfNumberColumnSelected = function (state) {
    var cellSelection = getCellSelection(state);
    if (cellSelection) {
        var tableNode = cellSelection.$anchorCell.node(-1);
        if (tableNode.attrs.isNumberColumnEnabled &&
            isColumnSelected(0)(state.selection)) {
            return true;
        }
    }
    return false;
};
export var checkIfNumberColumnCellsSelected = function (state) {
    var cellSelection = getCellSelection(state);
    if (cellSelection) {
        var tableNode = cellSelection.$anchorCell.node(-1);
        if (tableNode.attrs.isNumberColumnEnabled) {
            var map_1 = TableMap.get(tableNode);
            var start_1 = cellSelection.$anchorCell.start(-1);
            var selected_1 = false;
            cellSelection.forEachCell(function (cell, pos) {
                var rect = map_1.findCell(pos - start_1);
                if (rect.left === 0) {
                    selected_1 = true;
                }
            });
            return selected_1;
        }
    }
    return false;
};
