// @atlaskit/editor-common/dist/es2015/styles/index.js
import { akCodeFontFamily, akColorB100, akColorB300, akColorB400, akColorB50, akColorB75, akColorN0, akColorN20, akColorN40, akColorN50, akColorN100, akColorN500, akColorN700, akColorN900, akZIndexLayer, akColorR300, akColorR50, akColorR75, } from '@atlaskit/util-shared-styles';
export var akEditorCodeFontFamily = akCodeFontFamily;
export var akEditorInactiveForeground = akColorN500;
export var akEditorFocus = akColorB100;
export var akEditorSubtleAccent = akColorN40;
export var akEditorActiveBackground = akColorN500;
export var akEditorActiveForeground = akColorN0;
export var akEditorBlockquoteBorderColor = akColorN40;
export var akEditorDropdownActiveBackground = akColorN900;
export var akEditorPopupBackground = akColorN700;
export var akEditorPopupText = akColorB50;
export var akEditorPrimaryButton = akColorB400;
export var akEditorCodeBackground = akColorN20;
export var akEditorCodeBlockPadding = '12px';
export var akEditorCodeInlinePadding = '2px 4px';
export var akEditorFloatingPanelZIndex = akZIndexLayer;
export var akEditorMentionSelected = akColorN100;
export var akEditorTableToolbarSize = 12;
export var akEditorTableBorder = akColorN50;
export var akEditorTableToolbar = akColorN20;
export var akEditorTableFloatingControls = akColorN20;
export var akEditorTableCellSelected = akColorB75;
export var akEditorTableToolbarSelected = akColorB100;
export var akEditorTableBorderSelected = akColorB300;
export var akEditorTableCellDelete = akColorR50;
export var akEditorTableBorderDelete = akColorR300;
export var akEditorTableToolbarDelete = akColorR75;
export var akEditorTableBorderRadius = '3px';
export var akEditorTableCellBackgroundOpacity = 0.5;
export var akEditorFullPageMaxWidth = 680;
export var akEditorWideLayoutWidth = 960;
export var akEditorTableNumberColumnWidth = 40;
