// @atlaskit/editor-core/dist/es2015/plugins/jira-issue/index.js
import { Plugin, PluginKey } from 'prosemirror-state';
import { confluenceJiraIssue } from '@atlaskit/editor-common';
import { nodeViewFactory } from '../../nodeviews';
import ReactJIRAIssueNode from './nodeviews/jira-issue';
export var pluginKey = new PluginKey('jiraIssuePlugin');
var createPlugin = function (schema, providerFactory) {
    return new Plugin({
        key: pluginKey,
        props: {
            nodeViews: {
                confluenceJiraIssue: nodeViewFactory(providerFactory, {
                    confluenceJiraIssue: ReactJIRAIssueNode,
                }),
            },
        },
    });
};
var jiraIssuePlugin = {
    nodes: function () {
        return [
            { rank: 1400, name: 'confluenceJiraIssue', node: confluenceJiraIssue },
        ];
    },
    pmPlugins: function () {
        return [
            {
                rank: 1410,
                plugin: function (_a) {
                    var schema = _a.schema, providerFactory = _a.providerFactory;
                    return createPlugin(schema, providerFactory);
                },
            },
        ];
    },
};
export default jiraIssuePlugin;
