// @atlaskit/editor-core/dist/es5/plugins/date/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var prosemirror_state_1 = require("prosemirror-state");
var date_1 = require("./nodeviews/date");
var nodeviews_1 = require("../../nodeviews");
exports.pluginKey = new prosemirror_state_1.PluginKey('datePlugin');
exports.default = function (dispatch, providerFactory) {
    return new prosemirror_state_1.Plugin({
        state: {
            init: function () { return ({ element: null }); },
            apply: function (tr, state) {
                var meta = tr.getMeta(exports.pluginKey);
                if (meta) {
                    var newState = tslib_1.__assign({}, state, meta);
                    dispatch(exports.pluginKey, newState);
                    return newState;
                }
                return state;
            },
        },
        key: exports.pluginKey,
        props: {
            nodeViews: {
                date: nodeviews_1.nodeViewFactory(providerFactory, { date: date_1.default }),
            },
        },
    });
};
