// @atlaskit/media-picker/dist/es2015/components/dropzone/index.js
import * as tslib_1 from "tslib";
import { LocalUploadComponent } from '../localUpload';
import { MPDropzoneLoaded } from '../../outer/analytics/events';
import { whenDomReady } from '../../util/documentReady';
import dropzoneUI from './dropzoneUI';
var toArray = function (arr) { return [].slice.call(arr, 0); };
var Dropzone = /** @class */ (function (_super) {
    tslib_1.__extends(Dropzone, _super);
    function Dropzone(analyticsContext, context, config) {
        if (config === void 0) { config = { uploadParams: {} }; }
        var _this = _super.call(this, analyticsContext, context, config) || this;
        _this.onDragOver = function (e) {
            e.preventDefault();
            if (_this.instance && Dropzone.dragContainsFiles(e)) {
                var dataTransfer = e.dataTransfer;
                var allowed = void 0;
                try {
                    allowed = dataTransfer.effectAllowed;
                }
                catch (e) { } // the error is expected in IE11
                dataTransfer.dropEffect =
                    'move' === allowed || 'linkMove' === allowed ? 'move' : 'copy';
                _this.instance.classList.add('active');
                var length_1 = _this.getDraggedItemsLength(dataTransfer);
                _this.emitDragOver({ length: length_1 });
            }
        };
        _this.onDragLeave = function (e) {
            if (_this.instance) {
                e.preventDefault();
                _this.instance.classList.remove('active');
                _this.emitDragLeave();
            }
        };
        _this.onDrop = function (e) {
            var instance = _this.instance;
            if (instance && Dropzone.dragContainsFiles(e)) {
                instance.classList.remove('active');
                _this.emit('drop', undefined);
                _this.emitDragLeave();
            }
        };
        var container = config.container, headless = config.headless;
        _this.container = container || document.body;
        _this.headless = headless || false;
        _this.uiActive = false;
        _this.analyticsContext.trackEvent(new MPDropzoneLoaded());
        return _this;
    }
    Dropzone.prototype.activate = function () {
        var _this = this;
        return whenDomReady
            .then(function () {
            _this.container = _this.container || document.body;
            if (!_this.instance) {
                return _this.createInstance();
            }
        })
            .then(function () {
            _this.deactivate(); // in case we call activate twice in a row
            _this.container.addEventListener('dragover', _this.onDragOver, false);
            _this.container.addEventListener('dragleave', _this.onDragLeave, false);
            _this.uploadService.addDropzone(_this.container);
        });
    };
    Dropzone.prototype.deactivate = function () {
        this.container.removeEventListener('dragover', this.onDragOver, false);
        this.container.removeEventListener('dragleave', this.onDragLeave, false);
        this.uploadService.removeDropzone();
    };
    // Cross-browser way of getting dragged items length, we prioritize "items" if present
    // https://www.w3.org/TR/html51/editing.html#the-datatransfer-interface
    // This method is used on 'dragover' and we have no way to retrieve FileSystemFileEntry,
    // which contains info about if the dropped item is a file or directory. That info is only
    // available on 'drop'
    Dropzone.prototype.getDraggedItemsLength = function (dataTransfer) {
        if (dataTransfer.items) {
            var items = toArray(dataTransfer.items);
            return items.filter(function (i) { return i.kind === 'file'; }).length;
        }
        // This is required for IE11
        return dataTransfer.files.length;
    };
    Dropzone.prototype.createInstance = function () {
        var element = this.getDropzoneUI();
        this.instance = element;
        this.container.appendChild(this.instance);
        this.uploadService.on('file-dropped', this.onDrop);
    };
    Dropzone.prototype.getDropzoneUI = function () {
        if (this.headless) {
            return document.createElement('DIV');
        }
        else {
            return dropzoneUI;
        }
    };
    Dropzone.prototype.emitDragOver = function (e) {
        if (!this.uiActive) {
            this.uiActive = true;
            this.emit('drag-enter', e);
        }
    };
    Dropzone.prototype.emitDragLeave = function () {
        var _this = this;
        if (this.uiActive) {
            this.uiActive = false;
            /*
             when drag over child elements, container will issue dragleave and then dragover immediately.
             The 50ms timeout will prevent from issuing that "false" dragleave event
             */
            window.setTimeout(function () {
                if (!_this.uiActive) {
                    _this.emit('drag-leave', undefined);
                }
            }, 50);
        }
    };
    Dropzone.dragContainsFiles = function (event) {
        var types = event.dataTransfer.types;
        return toArray(types).indexOf('Files') > -1;
    };
    return Dropzone;
}(LocalUploadComponent));
export { Dropzone };
