// @atlaskit/media-card/dist/es2015/links/cardGenericView/index.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { Component } from 'react';
import { ErrorCard, CardFrame, CardPreview, LinkIcon, } from '@atlaskit/media-ui';
import { getCardMinWidth, getCardMaxWidth } from '../../utils/cardDimensions';
import CardDetails from './CardDetails';
import { defaultLinkCardAppearance } from '../card';
var LinkCardGenericView = /** @class */ (function (_super) {
    tslib_1.__extends(LinkCardGenericView, _super);
    function LinkCardGenericView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(LinkCardGenericView.prototype, "isHorizontal", {
        get: function () {
            var appearance = this.props.appearance;
            return appearance === 'horizontal';
        },
        enumerable: true,
        configurable: true
    });
    LinkCardGenericView.prototype.renderIcon = function () {
        var _a = this.props, iconUrl = _a.iconUrl, isLoading = _a.isLoading;
        if (isLoading) {
            return undefined;
        }
        return React.createElement(LinkIcon, { src: iconUrl });
    };
    LinkCardGenericView.prototype.renderPreview = function () {
        var isHorizontal = this.isHorizontal;
        var _a = this.props, isLoading = _a.isLoading, thumbnailUrl = _a.thumbnailUrl;
        if (isHorizontal) {
            return null;
        }
        return (React.createElement(CardPreview, { key: "preview", isPlaceholder: isLoading, url: thumbnailUrl || '' }));
    };
    LinkCardGenericView.prototype.renderDetails = function () {
        var isHorizontal = this.isHorizontal;
        var _a = this.props, isLoading = _a.isLoading, title = _a.title, description = _a.description, thumbnailUrl = _a.thumbnailUrl;
        return (React.createElement(CardDetails, { isPlaceholder: isLoading, isThumbnailVisible: isHorizontal, title: title, description: description, thumbnail: isHorizontal ? thumbnailUrl : undefined }));
    };
    LinkCardGenericView.prototype.render = function () {
        var _a = this.props, isLoading = _a.isLoading, site = _a.site, linkUrl = _a.linkUrl, appearance = _a.appearance, errorMessage = _a.errorMessage;
        if (errorMessage) {
            return (React.createElement(ErrorCard, { hasPreview: appearance !== 'horizontal', minWidth: getCardMinWidth(appearance), maxWidth: getCardMaxWidth(appearance) }));
        }
        return (React.createElement(CardFrame, { isPlaceholder: isLoading, href: linkUrl, icon: this.renderIcon(), text: site || linkUrl, minWidth: getCardMinWidth(appearance), maxWidth: getCardMaxWidth(appearance) },
            this.renderPreview(),
            this.renderDetails()));
    };
    LinkCardGenericView.defaultProps = {
        title: '',
        description: '',
        actions: [],
        appearance: defaultLinkCardAppearance,
    };
    return LinkCardGenericView;
}(Component));
export { LinkCardGenericView };
export default LinkCardGenericView;
