// @atlaskit/flag/dist/cjs/components/FlagActions/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _v = require('uuid/v1');

var _v2 = _interopRequireDefault(_v);

var _styledFlagActions = require('./styledFlagActions');

var _styledFlagActions2 = _interopRequireDefault(_styledFlagActions);

var _Flag = require('../Flag');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FlagActions = function (_Component) {
  (0, _inherits3.default)(FlagActions, _Component);

  function FlagActions() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, FlagActions);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = FlagActions.__proto__ || (0, _getPrototypeOf2.default)(FlagActions)).call.apply(_ref, [this].concat(args))), _this), _this.getButtonFocusRingColor = function () {
      return 'focusRingColor';
    }, _this.getButtonTheme = function () {
      return 'buttonTheme';
    }, _this.getButtonAppearance = function (b) {
      return b ? 'default' : 'subtle-link';
    }, _this.getButtonSpacing = function (b) {
      return b ? 'compact' : 'none';
    }, _this.getUniqueId = function (prefix) {
      return prefix + '-' + (0, _v2.default)();
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  } // eslint-disable-line react/sort-comp

  /* eslint-disable no-undef */

  (0, _createClass3.default)(FlagActions, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          actions = _props.actions,
          appearance = _props.appearance;

      var isBold = appearance !== _Flag.DEFAULT_APPEARANCE;

      if (!actions.length) return null;

      var items = actions.map(function (action, index) {
        return _react2.default.createElement(
          _styledFlagActions.Action,
          {
            key: _this2.getUniqueId('flag-action'),
            hasDivider: !!index,
            useMidDot: !isBold
          },
          _react2.default.createElement(
            _styledFlagActions.Button,
            { onClick: action.onClick, appearance: appearance },
            action.content
          )
        );
      });

      return _react2.default.createElement(
        _styledFlagActions2.default,
        null,
        items
      );
    }
  }]);
  return FlagActions;
}(_react.Component);

FlagActions.defaultProps = {
  appearance: _Flag.DEFAULT_APPEARANCE,
  actions: []
};
exports.default = FlagActions;
