// @atlaskit/theme/dist/cjs/typography.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.h100 = exports.h200 = exports.h300 = exports.h400 = exports.h500 = exports.h600 = exports.h700 = exports.h800 = exports.h900 = undefined;

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  ', ' color: ', ';\n  font-weight: 500;\n  letter-spacing: -0.01em;\n  margin-top: ', 'px;\n'], ['\n  ', ' color: ', ';\n  font-weight: 500;\n  letter-spacing: -0.01em;\n  margin-top: ', 'px;\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  ', ' color: ', ';\n  font-weight: 600;\n  letter-spacing: -0.01em;\n  margin-top: ', 'px;\n'], ['\n  ', ' color: ', ';\n  font-weight: 600;\n  letter-spacing: -0.01em;\n  margin-top: ', 'px;\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  ', ' color: ', ';\n  font-weight: 500;\n  letter-spacing: -0.008em;\n  margin-top: ', 'px;\n'], ['\n  ', ' color: ', ';\n  font-weight: 500;\n  letter-spacing: -0.008em;\n  margin-top: ', 'px;\n']),
    _templateObject4 = (0, _taggedTemplateLiteral3.default)(['\n  ', ' color: ', ';\n  font-weight: 600;\n  letter-spacing: -0.006em;\n  margin-top: ', 'px;\n'], ['\n  ', ' color: ', ';\n  font-weight: 600;\n  letter-spacing: -0.006em;\n  margin-top: ', 'px;\n']),
    _templateObject5 = (0, _taggedTemplateLiteral3.default)(['\n  ', ' color: ', ';\n  font-weight: 600;\n  letter-spacing: -0.003em;\n  margin-top: ', 'px;\n'], ['\n  ', ' color: ', ';\n  font-weight: 600;\n  letter-spacing: -0.003em;\n  margin-top: ', 'px;\n']),
    _templateObject6 = (0, _taggedTemplateLiteral3.default)(['\n  ', ' color: ', ';\n  font-weight: 600;\n  margin-top: ', 'px;\n  text-transform: uppercase;\n'], ['\n  ', ' color: ', ';\n  font-weight: 600;\n  margin-top: ', 'px;\n  text-transform: uppercase;\n']),
    _templateObject7 = (0, _taggedTemplateLiteral3.default)(['\n  ', ' color: ', ';\n  font-weight: 600;\n  margin-top: ', 'px;\n'], ['\n  ', ' color: ', ';\n  font-weight: 600;\n  margin-top: ', 'px;\n']),
    _templateObject8 = (0, _taggedTemplateLiteral3.default)(['\n  ', ' color: ', ';\n  font-weight: normal;\n  margin-top: ', 'px;\n'], ['\n  ', ' color: ', ';\n  font-weight: normal;\n  margin-top: ', 'px;\n']);

var _styledComponents = require('styled-components');

var _colors = require('./colors');

var colors = _interopRequireWildcard(_colors);

var _index = require('./index');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var baseHeading = function baseHeading(size, lineHeight) {
  return '\n  font-size: ' + size / (0, _index.fontSize)() + 'em;\n  font-style: inherit;\n  line-height: ' + lineHeight / size + ';\n';
};

var h900 = exports.h900 = function h900() {
  return (0, _styledComponents.css)(_templateObject, baseHeading(35, 40), colors.heading, (0, _index.gridSize)() * 6.5);
};

var h800 = exports.h800 = function h800() {
  return (0, _styledComponents.css)(_templateObject2, baseHeading(29, 32), colors.heading, (0, _index.gridSize)() * 5);
};

var h700 = exports.h700 = function h700() {
  return (0, _styledComponents.css)(_templateObject, baseHeading(24, 28), colors.heading, (0, _index.gridSize)() * 5);
};

var h600 = exports.h600 = function h600() {
  return (0, _styledComponents.css)(_templateObject3, baseHeading(20, 24), colors.heading, (0, _index.gridSize)() * 3.5);
};

var h500 = exports.h500 = function h500() {
  return (0, _styledComponents.css)(_templateObject4, baseHeading(16, 20), colors.heading, (0, _index.gridSize)() * 3);
};

var h400 = exports.h400 = function h400() {
  return (0, _styledComponents.css)(_templateObject5, baseHeading(14, 16), colors.heading, (0, _index.gridSize)() * 2);
};

var h300 = exports.h300 = function h300() {
  return (0, _styledComponents.css)(_templateObject6, baseHeading(12, 16), colors.subtleHeading, (0, _index.gridSize)() * 2.5);
};

var h200 = exports.h200 = function h200() {
  return (0, _styledComponents.css)(_templateObject7, baseHeading(12, 16), colors.subtleHeading, (0, _index.gridSize)() * 2);
};

var h100 = exports.h100 = function h100() {
  return (0, _styledComponents.css)(_templateObject8, baseHeading(12, 16), colors.subtleHeading, (0, _index.gridSize)() * 2);
};
