// @atlaskit/lozenge/dist/cjs/Lozenge/theme.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.boldStyles = exports.defaultStyles = exports.boldTextColor = exports.boldBackgroundColor = exports.textColor = exports.backgroundColor = undefined;

var _theme = require('@atlaskit/theme');

/* Note:
 Lozenge does not support dark mode at the moment,
 it will be implemented later.
 Hence, color values are the same.
 */
var backgroundColor = exports.backgroundColor = (0, _theme.themed)('appearance', {
  default: { light: _theme.colors.N40, dark: _theme.colors.N40 },
  inprogress: { light: _theme.colors.B50, dark: _theme.colors.B50 },
  moved: { light: _theme.colors.Y75, dark: _theme.colors.Y75 },
  new: { light: _theme.colors.P50, dark: _theme.colors.P50 },
  removed: { light: _theme.colors.R50, dark: _theme.colors.R50 },
  success: { light: _theme.colors.G50, dark: _theme.colors.G50 }
});
var textColor = exports.textColor = (0, _theme.themed)('appearance', {
  default: { light: _theme.colors.N500, dark: _theme.colors.N500 },
  inprogress: { light: _theme.colors.B500, dark: _theme.colors.B500 },
  moved: { light: _theme.colors.N800, dark: _theme.colors.N800 },
  new: { light: _theme.colors.P500, dark: _theme.colors.P500 },
  removed: { light: _theme.colors.R500, dark: _theme.colors.R500 },
  success: { light: _theme.colors.G500, dark: _theme.colors.G500 }
});

var boldBackgroundColor = exports.boldBackgroundColor = (0, _theme.themed)('appearance', {
  default: { light: _theme.colors.N500, dark: _theme.colors.N500 },
  inprogress: { light: _theme.colors.B400, dark: _theme.colors.B400 },
  moved: { light: _theme.colors.Y500, dark: _theme.colors.Y500 },
  new: { light: _theme.colors.P400, dark: _theme.colors.P400 },
  removed: { light: _theme.colors.R400, dark: _theme.colors.R400 },
  success: { light: _theme.colors.G400, dark: _theme.colors.G400 }
});

var boldTextColor = exports.boldTextColor = (0, _theme.themed)('appearance', {
  default: { light: _theme.colors.N0, dark: _theme.colors.N0 },
  inprogress: { light: _theme.colors.N0, dark: _theme.colors.N0 },
  moved: { light: _theme.colors.N800, dark: _theme.colors.N800 },
  new: { light: _theme.colors.N0, dark: _theme.colors.N0 },
  removed: { light: _theme.colors.N0, dark: _theme.colors.N0 },
  success: { light: _theme.colors.N0, dark: _theme.colors.N0 }
});

var defaultStyles = exports.defaultStyles = { backgroundColor: backgroundColor, textColor: textColor };
var boldStyles = exports.boldStyles = {
  backgroundColor: boldBackgroundColor,
  textColor: boldTextColor
};
