// @atlaskit/editor-core/dist/es5/plugins/panel/pm-plugins/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var prosemirror_state_1 = require("prosemirror-state");
var prosemirror_utils_1 = require("prosemirror-utils");
var analytics_1 = require("../../../analytics");
var panel_1 = require("../nodeviews/panel");
exports.availablePanelType = [
    { panelType: 'info' },
    { panelType: 'note' },
    { panelType: 'success' },
    { panelType: 'warning' },
    { panelType: 'error' },
];
var PanelState = /** @class */ (function () {
    function PanelState(state) {
        this.changeHandlers = [];
        this.editorFocused = false;
        this.changeHandlers = [];
        this.state = state;
        this.toolbarVisible = false;
    }
    PanelState.prototype.updateEditorFocused = function (editorFocused) {
        this.editorFocused = editorFocused;
    };
    PanelState.prototype.changePanelType = function (view, _a) {
        var panelType = _a.panelType;
        analytics_1.analyticsService.trackEvent("atlassian.editor.format." + panelType + ".button");
        var _b = view.state, tr = _b.tr, schema = _b.schema, dispatch = view.dispatch;
        dispatch(prosemirror_utils_1.setParentNodeMarkup(schema.nodes.panel, null, { panelType: panelType })(tr));
    };
    PanelState.prototype.removePanel = function (view) {
        var _a = view.state, tr = _a.tr, schema = _a.schema, dispatch = view.dispatch;
        dispatch(prosemirror_utils_1.removeParentNodeOfType(schema.nodes.panel)(tr));
    };
    PanelState.prototype.subscribe = function (cb) {
        this.changeHandlers.push(cb);
        cb(this);
    };
    PanelState.prototype.unsubscribe = function (cb) {
        this.changeHandlers = this.changeHandlers.filter(function (ch) { return ch !== cb; });
    };
    PanelState.prototype.update = function (state, domAtPos, domEvent) {
        var _this = this;
        if (domEvent === void 0) { domEvent = false; }
        this.state = state;
        var newPanel = this.getActivePanel();
        if ((domEvent && newPanel) || this.activeNode !== newPanel) {
            var newElement = newPanel && this.getDomElement(domAtPos);
            this.activeNode = newPanel;
            this.toolbarVisible =
                this.editorFocused &&
                    !!newPanel &&
                    (domEvent || this.element !== newElement);
            this.element = newElement;
            this.activePanelType = newPanel && newPanel.attrs['panelType'];
            this.changeHandlers.forEach(function (cb) { return cb(_this); });
        }
    };
    PanelState.prototype.getActivePanel = function () {
        var _a = this.state, selection = _a.selection, panel = _a.schema.nodes.panel;
        var parent = prosemirror_utils_1.findParentNodeOfType(panel)(selection);
        if (parent) {
            return parent.node;
        }
    };
    PanelState.prototype.getDomElement = function (domAtPos) {
        var _a = this.state, selection = _a.selection, panel = _a.schema.nodes.panel;
        var node = prosemirror_utils_1.findParentDomRefOfType(panel, domAtPos)(selection);
        if (node) {
            // getting panel nodeView wrapper
            while (!node.attributes['data-panel-type']) {
                node = node.parentNode;
            }
            return node;
        }
    };
    return PanelState;
}());
exports.PanelState = PanelState;
exports.stateKey = new prosemirror_state_1.PluginKey('panelPlugin');
// TODO: Fix types (ED-2987)
exports.createPlugin = function () {
    return new prosemirror_state_1.Plugin({
        state: {
            init: function (config, state) {
                return new PanelState(state);
            },
            apply: function (tr, pluginState, oldState, newState) {
                var stored = tr.getMeta(exports.stateKey);
                if (stored) {
                    pluginState.update(newState, stored.docView, stored.domEvent);
                }
                return pluginState;
            },
        },
        key: exports.stateKey,
        view: function (view) {
            return {
                update: function (view, prevState) {
                    exports.stateKey
                        .getState(view.state)
                        .update(view.state, view.domAtPos.bind(view));
                },
            };
        },
        props: {
            nodeViews: {
                panel: panel_1.panelNodeView,
            },
            handleClick: function (view, event) {
                exports.stateKey
                    .getState(view.state)
                    .update(view.state, view.domAtPos.bind(view), true);
                return false;
            },
            handleDOMEvents: {
                focus: function (view, event) {
                    exports.stateKey.getState(view.state).updateEditorFocused(true);
                    return false;
                },
                blur: function (view, event) {
                    var pluginState = exports.stateKey.getState(view.state);
                    pluginState.updateEditorFocused(false);
                    pluginState.update(view.state, view.domAtPos.bind(view), true);
                    return false;
                },
            },
        },
    });
};
